---
title: "E3/E4 Mice Meningeal Dura scRNAseq"
subtitle: "Filtering, Normalization, Clustering"
author: "Kennedi Todd"
date: "09/29/2024"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

# Experiment notes
- Cell suspensions of meningeal dura enriched by MACS for cells expressing Cd45, Cd31, and Lyve-1 \
  + Cd45/Ptprc/B220/Ly-5/Lyt-4/T200: transmembrane protein tyrosine phosphatase, 
located on most haematopoietic cells, positive selection of leukocytes \
  + Cd31/Pecam1: adhesion and signaling receptor that is expressed on endothelial and hematopoietic cells, positive selection of endothelial cells \
  + Lyve1/1200012G08Rik/Lyve-1/Xlkd1/lymphatic vessel endothelial HA receptor-1: found primarily on lymphatic endothelial cells \
- The meningeal layer of the dura mater is a durable, dense fibrous membrane that passes through the foramen magnum and is continuous with the dura mater of the spinal cord. The meningeal layer of the dura mater creates several dural folds that divide the cranial cavity into freely communicating spaces. \
- Aimed for ~5000 cells per sample to be sequenced at least 50,000 reads per cell \
  + Might ended up with more because we sequenced with the NovaSeq S2 \
- Expecting to see a lot of leukocytes and vascular cells \
- The purpose is to check the effect of sex and Apoe isoform expression on the transcriptomes of the different cell subpopulations \
- 4 experimental groups/samples
  + E3_14M_F = APOE3 expressing; 14 months old; female \
  + E3_14M_M = APOE3 expressing; 14 months old; male \
  + E4_14M_F = APOE4 expressing; 14 months old; female \
  + E4_14M_M = APOE4 expressing; 14 months old; male \

# Setup
```{r working_directory, echo=FALSE}
knitr::opts_knit$set(root.dir = ".")
```

## Load libraries
```{r libraries, warning=FALSE, message=FALSE}
library(cowplot)     # plot_grid()
library(dotenv)      # load_dot_env()
library(dplyr)       # left_join()
library(ggplot2)     # ggplot()
library(gridExtra)   # grid.arrange()
library(Matrix)      # rowSums()
library(parallel)    # detectCores()
library(rtracklayer) # import()
library(scCustomize) # Read_CellBender_h5_Mat()
library(Seurat)      # Read10X_h5()
library(stringr)     # str_match()
```

## Variables and functions
```{r set_variables_thresholds_functions}
# user variables
out <- "../../results/pass_1/all_clusters/"
sample_order <- c("E3.M","E3.F","E4.M","E4.F")
sample_colors <- c("#26946A","#1814A1","#EAC941","#DF5F00")
sample_order2 <- c("Male E3", "Male E4", "Female E3", "Female E4")
isoform_order <- c("E4","E3")
isoform_colors <- c("darkgray","cornflowerblue")
sex_order <- c("Male","Female")
sex_colors <- c("darkgray","purple")

# environment variables
load_dot_env(file = "../../refs/.env")
counts_dir <- Sys.getenv("COUNTS_DIR")

# thresholds
nCount.min <- 500
nCount.max <- 20000
nFeature.min <- 250
complexity.cutoff <- 0.8
mt.cutoff <- 20
ribo.cutoff <- 50
hb.cutoff <- 3

# functions
source("../../../functions/Kennedi_single_cell_functions.R")
saveToPDF <- function(...) {
    d = dev.copy(pdf,...)
    dev.off(d)
}
```

# Load data
## Merge h5
```{r read_h5}
if (file.exists("../../rObjects/cellbender_h5.rds")) {
  mouse <- readRDS("../../rObjects/cellbender_h5.rds")
} else {
  # individual sample objects
  E3.F <- CreateSeuratObject(
    scCustomize::Read_CellBender_h5_Mat(paste0(counts_dir, "/E3_14M_F/outs/E3_14M_F_cellbender_filtered.h5")))
  E3.M <- CreateSeuratObject(
    scCustomize::Read_CellBender_h5_Mat(paste0(counts_dir, "/E3_14M_M/outs/E3_14M_M_cellbender_filtered.h5")))
  E4.F <- CreateSeuratObject(
    scCustomize::Read_CellBender_h5_Mat(paste0(counts_dir, "/E4_14M_F/outs/E4_14M_F_cellbender_filtered.h5")))
  E4.M <- CreateSeuratObject(
    scCustomize::Read_CellBender_h5_Mat(paste0(counts_dir, "/E4_14M_M/outs/E4_14M_M_cellbender_filtered.h5")))

  # merge objects
  mouse <- merge(x = E3.F, 
                 y = c(E3.M,E4.F,E4.M),
                 add.cell.ids = c("E3.F","E3.M","E4.F","E4.M"), 
                 project = "Mouse Meningeal Dura scRNAseq")
  
  # cleanup and save
  remove(E3.F, E3.M, E4.F, E4.M)
  saveRDS(mouse, "../../rObjects/raw_h5.rds")
}

# join layers
mouse <- JoinLayers(mouse)

# preview
mouse
```

## Annotation
- Annotation file was downloaded from 10x Genomics \
  + refdata-gex-mm10-2020-A \
  + provider: GENCODE \
  + description: evidence-based annotation of the mouse genome (GRCm38), version M23 (Ensembl 98) \
- Gm* genes are originally annotated by MGI and the *Rik genes are annotated by RIKEN \
```{r read_gtf}
# read in annotation file, GENCODE GRCm38 version M23 (Ensembl 98)
if (file.exists("../../rObjects/annotation.rds")) {
  genes <- readRDS("../../rObjects/annotation.rds")
} else {
  gtf.file <- "../../refs/mouse_genes.gtf"
  genes <- rtracklayer::import(gtf.file)
  genes <- as.data.frame(genes)
  genes <- genes[genes$type == "gene",]
  saveRDS(genes, "../../rObjects/annotation.rds")
}
```

## Metadata columns
```{r add_metadata_columns}
# create sample column
barcodes <- colnames(mouse)
pattern <- "(.+)_[ACGT]+-(\\d+)"
sample <- str_match(barcodes, pattern)[,2]
table(sample) # check
mouse$sample <- factor(sample, levels = sample_order)
Idents(mouse) <- mouse$sample

# age column
mouse$age <- "14 months"

# sex column
sex <- str_match(mouse$sample, "E\\d.([FM])")[,2]
sex <- gsub("F","Female",sex)
sex <- gsub("M","Male",sex)
mouse$sex <- factor(sex, levels = sex_order)

# Apoe isoform column
isoform <- str_match(mouse$sample, "(E\\d).[FM]")[,2]
mouse$isoform <- factor(isoform, levels = isoform_order)

# sample2 column (for title formatting)
mouse$sample2 <- paste(mouse$sex, mouse$isoform, sep = " ")

# cell.complexity
mouse$cell.complexity <- log10(mouse$nFeature_RNA) / log10(mouse$nCount_RNA)

# percent.mt
mt.genes <- genes[genes$seqnames == "chrM",13]
mouse$percent.mt <- PercentageFeatureSet(mouse, features = mt.genes)
mt.genes

# percent.ribo
# ribosomal proteins begin with 'Rps' or 'Rpl' in this annotation file
# mitochondrial ribosomes start with 'Mrps' or 'Mrpl' (not on MT chromosome)
gene.names <- genes$gene_name
ribo <- gene.names[grep("^Rp[sl]", gene.names)]
mt.ribo <- gene.names[grep("^Mrp[sl]", gene.names)]
ribo.combined <- c(mt.ribo,ribo)
mouse$percent.ribo.protein <- PercentageFeatureSet(mouse, features = ribo.combined)
ribo.combined

# percent.hb - hemoglobin proteins begin with 'Hbb' or 'Hba' for mouse
hb.genes <- gene.names[grep("^Hb[ba]-", gene.names)]
mouse$percent.hb <- PercentageFeatureSet(mouse, features = hb.genes)
hb.genes
```

# Pre-filtering QC
## Number of cells
```{r prefiltering_cells_per_sample}
# Visualize the number of cell counts per sample
data <- as.data.frame(table(mouse$sample))
colnames(data) <- c("sample","frequency")

ncells1 <- ggplot(data, aes(x = sample, y = frequency, fill = sample)) + 
  geom_col() +
  theme_classic() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            vjust=-0.25) +
  scale_fill_manual(values = sample_colors) + 
  scale_y_continuous(breaks = seq(0,15000, by = 5000), limits = c(0,15000)) +
  ggtitle("Raw: cells per sample") +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))
ncells1
```

## Density plots
```{r prefiltering_density, warning=FALSE}
# Visualize nCount_RNA
den1 <- ggplot(mouse@meta.data,
       aes(color = sample,
           x = nCount_RNA,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  scale_color_manual(values = sample_colors) +
  scale_fill_manual(values = sample_colors) +
  xlab("nCount_RNA") +
  ylab("Density") +
  theme(legend.position =  "none") +
  geom_vline(xintercept = nCount.min, linetype = "dashed", colour = "red") +
  geom_vline(xintercept = nCount.max, linetype = "dashed", colour = "red") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Visualize nFeature_RNA
den2 <- ggplot(mouse@meta.data,
       aes(color = sample,
           x = nFeature_RNA,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  scale_color_manual(values = sample_colors) +
  theme(legend.position =  "none") +
  scale_fill_manual(values = sample_colors) +
  xlab("nFeature_RNA") +
  ylab("Density") +
  geom_vline(xintercept = nFeature.min, linetype = "dashed", colour = "red") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Visualize cell complexity
# Quality cells are usually above 0.85
den3 <- ggplot(mouse@meta.data,
       aes(color = sample,
           x = cell.complexity,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_color_manual(values = sample_colors) +
  theme(legend.position =  "none") +
  scale_fill_manual(values = sample_colors) +
  xlab("Cell Complexity (log10(nFeature/nCount))") +
  ylab("Density") +
  geom_vline(xintercept = complexity.cutoff, linetype = "dashed", colour = "red") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Visualize percent.mt
den4 <- ggplot(mouse@meta.data,
       aes(color = sample,
           x = percent.mt,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_continuous(n.breaks = 4) +
  geom_vline(xintercept = mt.cutoff, linetype = "dashed", colour = "red") +
  scale_color_manual(values = sample_colors) +
  theme(legend.position =  "none") +
  scale_fill_manual(values = sample_colors) +
  xlab("% Mitochondrial Genes") +
  ylab("Density") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Visualize percent.ribo.protein
den5 <- ggplot(mouse@meta.data,
       aes(color = sample,
           x = percent.ribo.protein,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  scale_color_manual(values = sample_colors) +
  theme(legend.position =  "none") +
  scale_fill_manual(values = sample_colors) +
  xlab("% Ribosomal Protein Genes") +
  ylab("Density") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Visualize percent.hb
den6 <- ggplot(mouse@meta.data,
       aes(color = sample,
           x = percent.hb,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  geom_vline(xintercept = hb.cutoff, linetype = "dashed", colour = "red") +
  scale_color_manual(values = sample_colors) +
  theme(legend.position =  "none") +
  scale_fill_manual(values = sample_colors) +
  xlab("% Hemoglobin Genes") +
  ylab("Density") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Arrange graphs in grid
plots <- list(den1,den2,den3,den4,den5,den6)
layout <- rbind(c(1,4),c(2,5),c(3,6))
grid <- grid.arrange(grobs = plots, layout_matrix = layout)
```

```{r save_prefiltering_density, echo=FALSE, eval=FALSE}
# save
path <- paste0(out, "filtering_QC/density_curves_raw.pdf")
pdf(path, width = 8, height = 8)
grid <- grid.arrange(grobs = plots, layout_matrix = layout)

# cleanup
remove(den1,den2,den3,den4,den5,den6,layout,grid,plots,data)
```

## Violin plots
```{r prefiltering_violins}
# normalize data for violin plots
mouse <- NormalizeData(mouse, assay = "RNA")

# nFeature, nCount, and cell.complexity violins
v1 <- VlnPlot(mouse,
              features = c("nFeature_RNA", "nCount_RNA","cell.complexity"),
              ncol = 3,
              group.by = 'sample',
              cols = sample_colors,
              pt.size = 0)
v1

#  percent violins
v2 <- VlnPlot(mouse,
              features = c("percent.mt","percent.ribo.protein","percent.hb"),
              ncol = 3,
              group.by = 'sample',
              cols = sample_colors,
              pt.size = 0)
v2
```

```{r save_prefiltering_violins, echo=FALSE, eval=FALSE}
# save v1
path <- paste0(out, "filtering_QC/violins_nFeature_nCount_complexity_raw.pdf")
pdf(path, width = 12, height = 4)
v1
dev.off()

# save v2
path <- paste0(out, "filtering_QC/violins_percent_raw.pdf")
pdf(path, width = 12, height = 4)
v2
dev.off()

# cleanup
remove(v1,v2)
```

## Scatter plots
```{r prefiltering_scatter, warning=FALSE}
s1 <- ggplot(
  mouse@meta.data,
  aes(x = nCount_RNA, y = nFeature_RNA, color = percent.mt)) + 
  geom_point() + 
  geom_smooth(method = "lm") +
	scale_x_log10() +   	
  scale_y_log10() + 
  theme_classic() +
  geom_vline(xintercept = nCount.min, linetype = "dashed", colour = "red") + 
  geom_hline(yintercept = nFeature.min, linetype = "dashed", colour = "red") + 
  facet_wrap(~sample) +
  scale_colour_gradient(low = "gray90", high = "black", limits =c(0,100))
s1

s2 <- FeatureScatter(mouse,
                     feature1 = "nCount_RNA",
                     feature2 = "percent.mt",
                     group.by = 'sample',
                     cols = sample_colors,
                     shuffle = TRUE)
s2
```

```{r save_prefiltering_scatter, echo=FALSE, eval=FALSE, warning=FALSE, message=FALSE}
# save
path <- paste0(out, "filtering_QC/scatter_nFeature_vs_nCount_raw.pdf")
pdf(path, width = 8, height = 8)
s1
dev.off()

path <- paste0(out, "filtering_QC/scatter_percentMT_vs_nCount_raw.pdf")
pdf(path, width = 6, height = 6)
s2
dev.off()

# cleanup
remove(s1,s2)
```

# Filtering
## Cell-level filtering
```{r cell_filtering}
# filter
mouse.filtered <- subset(mouse,
                         subset = (nCount_RNA > nCount.min) &
                           (nCount_RNA < nCount.max) &
                           (nFeature_RNA > nFeature.min) &
                           (cell.complexity > complexity.cutoff) &
                           (percent.mt < mt.cutoff) &
                           (percent.hb < hb.cutoff))

# print cells removed
print(paste0(dim(mouse)[2] - dim(mouse.filtered)[2]," cells removed"))
```

## Gene-level filtering
Remove lowly expressed genes.  We will keep genes that have at least 1 count in 10 cells.
```{r gene_filtering}
# filter genes
mouse.filtered <- JoinLayers(mouse.filtered)
counts <- GetAssayData(object = mouse.filtered, layer = "counts")
nonzero <- counts > 0  # produces logical
keep <- Matrix::rowSums(nonzero) >= 10  # sum the true/false
counts.filtered <- counts[keep,]  # keep certain genes

# overwrite mouse.filtered
mouse.filtered <- CreateSeuratObject(counts.filtered, 
                                     meta.data = mouse.filtered@meta.data)

# print features removed
print(paste0(dim(counts)[1] - dim(counts.filtered)[1], " features removed"))
```

Remove highly abundant transcripts.
```{r remove_specific_genes}
# remove mt.genes
counts <- GetAssayData(object = mouse.filtered, layer = "counts")
keep <- !rownames(counts) %in% mt.genes # false when mt.gene
counts.filtered <- counts[keep,]

# remove ribo.genes
keep <- !rownames(counts.filtered) %in% ribo.combined
counts.filtered <- counts.filtered[keep,]

# remove Ig genes + Jchain but keep Igha + Ighd to enahnce clustering
gene.types <- c("IG_C_gene","IG_C_pseudogene","IG_D","IG_J_gene","IG_LV_gene",
                "IG_V_gene","IG_V_pseudogene")
keep <- (genes$gene_type) %in% gene.types
ig.genes <- genes[keep,]
ig.genes <- c(ig.genes$gene_name, "Jchain")
ig.genes <- ig.genes[-c(185,192)] # keep Igha and Ighd
ig.genes <- ig.genes[!ig.genes %in% c("Igha","Ighd")]
ig.genes
keep <- !rownames(counts.filtered) %in% ig.genes
counts.filtered <- counts.filtered[keep,]

# overwrite mouse.filtered
mouse.filtered <- CreateSeuratObject(counts.filtered,
                                     meta.data = mouse.filtered@meta.data)

# print features removed
print(paste0(dim(counts)[1] - dim(counts.filtered)[1], " features removed"))

# cleanup data
remove(mouse,counts,counts.filtered,nonzero)
```

```{r save_filtered_object, echo=FALSE, eval=FALSE}
saveRDS(mouse.filtered, "../../rObjects/filtered.rds")
#mouse.filtered <- readRDS("../../rObjects/filtered.rds")
```

# Post-filtering QC
## Number of cells
```{r postfiltering_cells_per_sample}
# Visualize the number of cell counts per sample
data <- as.data.frame(table(mouse.filtered$sample))
colnames(data) <- c("sample","frequency")

ncells2 <- ggplot(data, aes(x = sample, y = frequency, fill = sample)) + 
  geom_col() +
  theme_classic() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            vjust=-0.25) +
  scale_fill_manual(values = sample_colors) + 
  scale_y_continuous(breaks = seq(0,12000, by = 2000), limits = c(0,12000)) +
  ggtitle("Filtered: cells per sample") +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))

# Arrange graphs in grid
plots <- list(ncells1,ncells2)
layout <- rbind(c(1),c(2))
grid <- grid.arrange(grobs = plots, layout_matrix = layout)
```

```{r save_cells_per_sample,echo=FALSE,eval=FALSE}
# save
path <- paste0(out, "filtering_QC/cells_per_sample_before_and_after_filtering.pdf")
pdf(path, width = 8, height = 8)
grid <- grid.arrange(grobs = plots, layout_matrix = layout)
dev.off()

# cleanup
remove(ncells1,ncells2,plots,layout,grid,data)
```

## Density plots
```{r postfiltering_density, warning=FALSE}
# Visualize nCount_RNA
den1 <- ggplot(mouse.filtered@meta.data,
       aes(color = sample,
           x = nCount_RNA,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  scale_color_manual(values = sample_colors) +
  scale_fill_manual(values = sample_colors) +
  xlab("nCount_RNA") +
  ylab("Density") +
  theme(legend.position =  "none") +
  geom_vline(xintercept = nCount.min, linetype = "dashed", colour = "red") +
  geom_vline(xintercept = nCount.max, linetype = "dashed", colour = "red") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Visualize nFeature_RNA
den2 <- ggplot(mouse.filtered@meta.data,
       aes(color = sample,
           x = nFeature_RNA,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  scale_color_manual(values = sample_colors) +
  theme(legend.position =  "none") +
  scale_fill_manual(values = sample_colors) +
  xlab("nFeature_RNA") +
  ylab("Density") +
  geom_vline(xintercept = nFeature.min, linetype = "dashed", colour = "red") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Visualize cell complexity
# Quality cells are usually above 0.85
den3 <- ggplot(mouse.filtered@meta.data,
       aes(color = sample,
           x = cell.complexity,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_color_manual(values = sample_colors) +
  theme(legend.position =  "none") +
  scale_fill_manual(values = sample_colors) +
  xlab("Cell Complexity (log10(nFeature/nCount))") +
  ylab("Density") +
  geom_vline(xintercept = complexity.cutoff, linetype = "dashed", colour = "red") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Visualize percent.mt
den4 <- ggplot(mouse.filtered@meta.data,
       aes(color = sample,
           x = percent.mt,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_continuous(n.breaks = 4) +
  geom_vline(xintercept = mt.cutoff, linetype = "dashed", colour = "red") +
  scale_color_manual(values = sample_colors) +
  theme(legend.position =  "none") +
  scale_fill_manual(values = sample_colors) +
  xlab("% Mitochondrial Genes") +
  ylab("Density") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Visualize percent.ribo.protein
den5 <- ggplot(mouse.filtered@meta.data,
       aes(color = sample,
           x = percent.ribo.protein,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  scale_color_manual(values = sample_colors) +
  theme(legend.position =  "none") +
  scale_fill_manual(values = sample_colors) +
  xlab("% Ribosomal Protein Genes") +
  ylab("Density") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Visualize percent.hb
den6 <- ggplot(mouse.filtered@meta.data,
       aes(color = sample,
           x = percent.hb,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  geom_vline(xintercept = hb.cutoff, linetype = "dashed", colour = "red") +
  scale_color_manual(values = sample_colors) +
  theme(legend.position =  "none") +
  scale_fill_manual(values = sample_colors) +
  xlab("% Hemoglobin Genes") +
  ylab("Density") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Arrange graphs in grid
plots <- list(den1,den2,den3,den4,den5,den6)
layout <- rbind(c(1,4),c(2,5),c(3,6))
grid <- grid.arrange(grobs = plots, layout_matrix = layout)
```

```{r save_postfiltering_density, echo=FALSE, eval=FALSE, warning=FALSE}
# save
path <- paste0(out, "filtering_QC/density_curves_filtered.pdf")
pdf(path, width = 8, height = 8)
grid <- grid.arrange(grobs = plots, layout_matrix = layout)
dev.off()

# cleanup
remove(den1,den2,den3,den4,den5,den6,plots,layout,grid)
```

## Violin plots
```{r postfiltering_violins}
# normalize data before violin plots
mouse.filtered <- NormalizeData(mouse.filtered)

# nFeature, nCount, and cell.complexity violins
v3 <- VlnPlot(mouse.filtered,
              features = c("nFeature_RNA", "nCount_RNA","cell.complexity"),
              ncol = 3,
              group.by = 'sample',
              cols = sample_colors,
              pt.size = 0)
v3

#  percent violins
v4 <- VlnPlot(mouse.filtered,
              features = c("percent.mt","percent.ribo.protein","percent.hb"),
              ncol = 3,
              group.by = 'sample',
              cols = sample_colors,
              pt.size = 0)
v4
```

```{r save_postfiltering_violins, echo=FALSE, eval=FALSE}
# save
path <- paste0(out, "filtering_QC/violins_nFeature_nCount_complexity_filtered.pdf")
pdf(path, width = 12, height = 4)
v3
dev.off()

# save
path <- paste0(out, "filtering_QC/violins_percent_filtered.pdf")
pdf(path, width = 12, height = 4)
v4
dev.off()

# cleanup
remove(v3,v4)
```

## Scatter plots
```{r postfiltering_scatter, warning=FALSE}
s3 <- ggplot(
  mouse.filtered@meta.data,
  aes(x = nCount_RNA, y = nFeature_RNA, color = percent.mt)) + 
  geom_point() + 
  geom_smooth(method = "lm") +
	scale_x_log10() +   	
  scale_y_log10() + 
  theme_classic() +
  geom_vline(xintercept = nCount.min, linetype = "dashed", colour = "red") + 
  geom_hline(yintercept = nFeature.min, linetype = "dashed", colour = "red") + 
  facet_wrap(~sample) +
  scale_colour_gradient(low = "gray90", high = "black", limits =c(0,100))
s3

s4 <- FeatureScatter(mouse.filtered,
                     feature1 = "nCount_RNA",
                     feature2 = "percent.mt",
                     group.by = 'sample',
                     cols = sample_colors,
                     shuffle = TRUE)
s4
```

```{r save_postfiltering_scatter,echo=FALSE,eval=FALSE}
# save
path <- paste0(out, "filtering_QC/scatter_nFeature_vs_nCount_filtered.pdf")
pdf(path, width = 8, height = 8)
s3
dev.off()

path <- paste0(out, "filtering_QC/scatter_nCount_vs_percentMT_filtered.pdf")
pdf(path, width = 6, height = 6)
s4
dev.off()

# cleanup
remove(s3,s4)
```

## Box plot
```{r boxplot}
# Visualize the distribution of genes detected per cell via boxplot
b1 <- ggplot(mouse.filtered@meta.data,
       aes(x = sample, 
           y = log10(nFeature_RNA), 
           fill=sample)) + 
  geom_boxplot() + 
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust = 0.5, face="bold")) +
  ggtitle("Unique Genes / Cell / Sample") +
  scale_color_manual(values = sample_colors) +
  scale_fill_manual(values = sample_colors) +
  xlab("Sample")
b1
```

```{r save_boxplot,echo=FALSE,eval=FALSE}
# save
path <- paste0(out, "filtering_QC/boxplot_nFeature_per_sample.pdf")
pdf(path, width = 8, height = 6)
b1
dev.off()

# cleanup
remove(b1)
```

## Top transcripts
```{r top_transcripts}
df <- data.frame(gene_name = rownames(mouse.filtered))
df$rsum <- rowSums(x = mouse.filtered, slot = "counts")
df <- df[order(df$rsum, decreasing = TRUE),]
rownames(df) <- 1:nrow(df)
head(df, 10)
```

```{r save_top_transcripts,echo=FALSE,eval=FALSE}
write.table(df, 
            paste0(out, "filtering_QC/top_transcripts.tsv"),
            quote = FALSE,
            row.names = FALSE)
remove(df)
```

# Unwanted variation
## Cell cycle
- [Evaluating effects of cell cycle](https://hbctraining.github.io/scRNA-seq_online/lessons/06_SC_SCT_normalization.html)
```{r cell_cycle, eval=FALSE}
markers <- "../../../functions/cell_cycle_markers.tsv"
path <- paste0(out, "unwanted_variation")
mouse.filtered[["phase"]] <- cell_cycle_QC(obj = mouse.filtered,
                                           markersPath = markers,
                                           species = "mouse",
                                           outDir = path)
```

## Mitochondria
- [Evaluating effects of mitochondrial expression](https://hbctraining.github.io/scRNA-seq_online/lessons/06_SC_SCT_normalization.html)
```{r mitochondria_factor, eval=FALSE}
mouse.filtered[["mito.factor"]] <- mitochondria_QC(obj = mouse.filtered,
                                                   outDir = path)
```

# Cluster
## SCTransform
- SCTransform method is a more accurate method of normalizing, estimating the 
variance of the raw filtered data, and identifying the most variable genes. \
- Variation in sequencing depth (total nCount_RNA per cell) is normalized using 
a regularized negative binomial model. \
- If there are other sources of uninteresting variation it can be included. \
```{r split_object}
# split
mouse.filtered[["RNA"]] <- split(mouse.filtered[["RNA"]], 
                                 f = mouse.filtered$sample)

# transform
mouse.filtered <- SCTransform(mouse.filtered, verbose = FALSE)
```

- NOTE: By default, after normalizing, adjusting the variance, and regressing out 
uninteresting sources of variation, SCTransform will rank the genes by residual 
variance and output the 3000 most variant genes. If the data set has larger cell 
numbers, then it may be beneficial to adjust this parameter higher using the 
variable.features.n argument. \
- It is suggested to not regress out batch, and instead use a data integration method like Harmony \

## PCA
```{r run_pca}
# run PCA on the merged object
mouse.filtered <- RunPCA(object = mouse.filtered, assay = "SCT")
```

```{r plot_pca}
# Reset idents and levels
DefaultAssay(mouse.filtered) <- "SCT"
Idents(mouse.filtered) <- "sample"
mouse.filtered$sample <- factor(mouse.filtered$sample, 
                              levels = sample_order)
mouse.filtered$sex <- factor(mouse.filtered$sex, 
                           levels = sex_order)
mouse.filtered$isoform <- factor(mouse.filtered$isoform,
                               levels = isoform_order)

# Plot PCA
pca1 <- DimPlot(mouse.filtered,
                reduction = "pca",
                split.by = "sample",
                group.by = "sample",
                cols = sample_colors,
                ncol = 2)
pca1

pca2 <- DimPlot(mouse.filtered,
                reduction = "pca",
                split.by = "isoform",
                group.by = "isoform",
                cols = isoform_colors)
pca2

pca3 <- DimPlot(mouse.filtered,
                reduction = "pca",
                split.by = "sex",
                group.by = "sex",
                cols = sex_colors)
pca3

pca4 <- DimPlot(mouse.filtered,
                reduction = "pca",
                group.by = "sample",
                cols = sample_colors,
                shuffle = TRUE,
                raster = FALSE)
pca4

e1 <- ElbowPlot(mouse.filtered) +
  geom_vline(xintercept = 15, linetype = "dashed", colour = "red")
e1
```

```{r save_pca,echo=FALSE,eval=FALSE,message=FALSE,warning=FALSE}
# save pca1
path <- paste0(out, "clustering_QC/PCA_split_by_sample.pdf")
pdf(path, width = 10, height = 10)
pca1
dev.off()

# save pca2
path <- paste0(out, "clustering_QC/PCA_split_by_isoform.pdf")
pdf(path, width = 6, height = 4)
pca2
dev.off()

# save pca3
path <- paste0(out, "clustering_QC/PCA_split_by_sex.pdf")
pdf(path, width = 6, height = 4)
pca3
dev.off()

# save pca4
path <- paste0(out, "clustering_QC/PCA_colored_by_sample.pdf")
pdf(path, width = 6, height = 4)
pca4
dev.off()

# save e1
path <- paste0(out, "clustering_QC/PCA_eblow_plot.pdf")
pdf(path, width = 6, height = 4)
e1
dev.off()

# cleanup
remove(pca1,pca2,pca3,pca4,e1)
```

## Harmony
```{r run_harmony, warning=FALSE}
# integrate
mouse.filtered <- IntegrateLayers(mouse.filtered,
                                  method = "HarmonyIntegration",
                                  assay = "SCT",
                                  orig.reduction = "pca")

# re-join layers
mouse.filtered[["RNA"]] <- JoinLayers(mouse.filtered[["RNA"]])
```

```{r plot_harmony}
hrm1 <- DimPlot(mouse.filtered,
                reduction = "harmony",
                split.by = "sample",
                group.by = "sample",
                cols = sample_colors,
                ncol = 2)
hrm1

hrm2 <- DimPlot(mouse.filtered,
                reduction = "harmony",
                split.by = "isoform",
                group.by = "isoform",
                cols = isoform_colors)
hrm2

hrm3 <- DimPlot(mouse.filtered,
                reduction = "harmony",
                split.by = "sex",
                group.by = "sex",
                cols = sex_colors)
hrm3

hrm4 <- DimPlot(mouse.filtered,
                reduction = "harmony",
                group.by = "sample",
                cols = sample_colors,
                shuffle = TRUE,
                raster = FALSE)
hrm4

# Plot elbow
mouse.filtered@reductions$harmony@stdev <-
  apply(mouse.filtered@reductions$harmony@cell.embeddings, 2, sd)
e1 <- ElbowPlot(mouse.filtered,
                reduction = "harmony") +
  geom_vline(xintercept = 15, linetype = "dashed", color = "red")
e1
```

```{r save_harmony,echo=FALSE,eval=FALSE,message=FALSE,warning=FALSE}
# save hrm1
path <- paste0(out, "clustering_QC/harmony_split_by_sample.pdf")
pdf(path, width = 10, height = 10)
hrm1
dev.off()

# save hrm2
path <- paste0(out, "clustering_QC/harmony_split_by_isoform.pdf")
pdf(path, width = 6, height = 4)
hrm2
dev.off()

# save hrm3
path <- paste0(out, "clustering_QC/harmony_split_by_sex.pdf")
pdf(path, width = 6, height = 4)
hrm3
dev.off()

# save hrm4
path <- paste0(out, "clustering_QC/harmony_colored_by_sample.pdf")
pdf(path, width = 6, height = 4)
hrm4
dev.off()

# save e1
path <- paste0(out, "clustering_QC/harmony_eblow_plot.pdf")
pdf(path, width = 6, height = 4)
e1
dev.off()

# cleanup
remove(hrm1,hrm2,hrm3,hrm4,e1)
```

## UMAP
To overcome the extensive technical noise in the expression of any single gene 
for scRNA-seq data, Seurat assigns cells to clusters based on their PCA scores 
derived from the expression of the integrated most variable genes, with each PC 
essentially representing a “metagene” that combines information across a 
correlated gene set. Determining how many PCs to include in the clustering step 
is therefore important to ensure that we are capturing the majority of the 
variation, or cell types, present in our dataset.
```{r run_umap, warning=FALSE}
# run UMAP
mouse.filtered <- RunUMAP(mouse.filtered,
                          dims = 1:15,
                          reduction = "harmony",
                          assay = "SCT",
                          n.components = 3) # set to 3 to use with VR

# plot UMAP
DimPlot(mouse.filtered,
        cols = sample_colors,
        shuffle = TRUE)
```

## Clusters
- Seurat uses a graph-based clustering approach, which embeds cells in a graph 
structure, using a K-nearest neighbor (KNN) graph. By default, it uses k = 30. \
- Then, it creates a shared nearest neighbor (SNN) graph and identifies communities \
- Increased resolution leads to a greater number of clusters \
```{r find_neighbors}
# Determine the K-nearest neighbor graph
mouse.unannotated <- FindNeighbors(object = mouse.filtered,
                                   assay = "SCT", # set as default after SCTransform
                                   reduction = "harmony",
                                   dims = 1:15)

# Determine the clusters for various resolutions
mouse.unannotated <- FindClusters(object = mouse.unannotated,
                                  algorithm = 1, # 1 = Louvain
                                  resolution = seq(0.1, 0.5, by=0.1))
```

## Explore resolutions
```{r explore_resolutions}
# 0.1
DimPlot(mouse.unannotated,
        group.by = "SCT_snn_res.0.1",
        label = TRUE)

# 0.2
DimPlot(mouse.unannotated,
        group.by = "SCT_snn_res.0.2",
        label = TRUE)

# 0.3
DimPlot(mouse.unannotated,
        group.by = "SCT_snn_res.0.3",
        label = TRUE)

# 0.4
DimPlot(mouse.unannotated,
        group.by = "SCT_snn_res.0.4",
        label = TRUE)

# 0.5
DimPlot(mouse.unannotated,
        group.by = "SCT_snn_res.0.5",
        label = TRUE)
```

## Choose resolution
```{r choose_res}
mouse.unannotated$seurat_clusters <- mouse.unannotated$SCT_snn_res.0.2

DimPlot(mouse.unannotated,
        group.by = "SCT_snn_res.0.2",
        label = TRUE)

DimPlot(mouse.unannotated,
        group.by = "SCT_snn_res.0.2",
        label = TRUE,
        dims = c(2,3))
```

```{r save_final_object, echo=FALSE, eval=FALSE}
# Save seurat object
saveRDS(mouse.unannotated, "../../rObjects/unannotated_obj.rds")
```
