---
title: "Mouse Meningeal Dura scRNAseq"
subtitle: "Subset, Recluster, Filter"
author: "Kennedi Todd"
date: "04/11/2023"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

# Setup
## Working directory
```{r working_directory}
knitr::opts_knit$set(root.dir = ".")
```

## Libraries
```{r libraries, message=FALSE, warning=FALSE}
# load packages
library(dplyr)        # ungroup()
library(ggtree)       # BuildClusterTree()
library(gridExtra)    # grid.arrange()
library(gtools)       # smartbind()
library(parallel)     # detectCores()
library(plotly)       # plot_ly()
library(Seurat)       # Idents()
library(ShinyCell)    # createConfig()
library(tidyr)        # %>%

# work in parallel
options(mc.cores = detectCores() - 1)
```

## Variables and functions
```{r set_variables_and_thresholds}
# variables
sample_order <- c("E3.M","E3.F","E4.M","E4.F")
sample_colors <- c("#26946A","#1814A1","#EAC941","#DF5F00")
sample_order2 <- c("Male E3", "Male E4", "Female E3", "Female E4")
isoform_order <- c("E4","E3")
isoform_colors <- c("darkgray","cornflowerblue")
sex_order <- c("Male","Female")
sex_colors <- c("darkgray","purple")

# single cell functions
source("../../refs/Kennedi_single_cell_functions.R")

# save function
saveToPDF <- function(...) {
    d = dev.copy(pdf,...)
    dev.off(d)
}
```

## Load data
```{r read_pass1_all_clusters_annotated}
mouse.annotated <- readRDS("../../rObjects/pass1_annotated.rds")
Idents(mouse.annotated) <- "merged_clusters"
DefaultAssay(mouse.annotated) <- "RNA"

merged_colors <- c("darkred","firebrick1","gold","darkolivegreen2","darkgreen",
                   "cyan","cornflowerblue","blue","darkorchid1","deeppink",
                   "plum1","burlywood3","azure4", "black")
DimPlot(object = mouse.annotated, 
        reduction = "umap",
        repel = TRUE,
        group.by = "merged_clusters",
        cols = merged_colors)
```

## Feature plots
```{r all_clusters_annotated_feature_plots}
f1 <- FeaturePlot(mouse.annotated,
                  features = "nFeature_RNA")  + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f1

f2 <- FeaturePlot(mouse.annotated,
                  features = "nCount_RNA")  + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f2

f3 <- FeaturePlot(mouse.annotated,
                  features = "Malat1")  + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f3
```

```{r save_all_clusters_annotated_feature_plots, echo=FALSE, eval=FALSE}
path <- "../../results/pass_1/all_clusters/UMAP/"
pdf(paste0(path,"nFeature.pdf"), height = 4, width = 6)
f1
dev.off()

pdf(paste0(path,"nCount.pdf"), height = 4, width = 6)
f2
dev.off()

pdf(paste0(path,"Malat1.pdf"), height = 4, width = 6)
f3
dev.off()
remove(f1,f2,f3)
```

# Lymphoid Recluster
## Recluster
```{r subset_recluster_lymphoid, eval=FALSE}
lymphoid <- subset_recluster(obj = mouse.annotated,
                             cellIdent = "merged_clusters",
                             cellType = c("B cells","Plasma cells","T cells",
                                          "Innate lymphoid cells"),
                             projectID = "Lymphoid Recluster",
                             keepCol = 4:14)

saveRDS(lymphoid, "../../rObjects/pass1_lymphoid_recluster_unannotated.rds")
```

```{r load_reclustered_lymphoid, echo=FALSE, eval=TRUE}
lymphoid <- readRDS("../../rObjects/pass1_lymphoid_recluster_unannotated.rds")
```

## Unannotated UMAP
```{r lymphoid_unannotated_umap}
lymphoid$seurat_clusters <- lymphoid$SCT_snn_res.0.1
Idents(lymphoid) <- "seurat_clusters"
cluster_colors <- c("darkred","firebrick1","gold","darkolivegreen2","darkgreen",
                    "cyan","cornflowerblue","blue","darkorchid1","deeppink",
                    "plum1","black")

u1 <- DimPlot(lymphoid,
              label = TRUE,
              cols = cluster_colors,
              reduction = "umap")
u1

u2 <- DimPlot(lymphoid,
              label = TRUE,
              cols = cluster_colors,
              dims = c(2,3),
              reduction = "umap")
u2
```

```{r save_lymphoid_unannotated_umap, eval=FALSE, echo=FALSE}
path <- "../../results/pass_1/lymphoid_recluster/UMAP/unannotated_clusters_dim"
pdf(paste0(path,"1&2.pdf"), height = 4, width = 6)
u1
dev.off()
pdf(paste0(path,"2&3.pdf"), height = 4, width = 6)
u2
dev.off()
```

## Potential markers
```{r find_unannotated_lymphoid_markers, eval=FALSE}
all.markers <- find_cluster_markers(obj = lymphoid,
                                    cellIdent = "seurat_clusters")
saveRDS(all.markers, 
        "../../rObjects/pass1_lymphoid_recluster_unannotated_cluster_markers.rds")

all.markers <- subset(all.markers, p_val_adj < 0.01)
path <- "../../results/pass_1/lymphoid_recluster/markers/"
write.table(all.markers,
            paste0(path,"unannotated_cluster_markers_adjpval_0.01.tsv"),
            quote = FALSE, row.names = FALSE, sep = "\t")
```

```{r top20_lymphoid_markers_per_cluster}
all.markers <- readRDS(
  "../../rObjects/pass1_lymphoid_recluster_unannotated_cluster_markers.rds")

all.markers <- subset(all.markers, p_val_adj < 0.01)
top20 <- Reduce(rbind,
                by(all.markers,
                   all.markers["cluster"],
                   head,
                   n = 20))
```

```{r unannotated_lymphoid_markers}
all.markers <- readRDS(
  "../../rObjects/pass1_lymphoid_recluster_unannotated_cluster_markers.rds")
all.markers <- subset(all.markers, p_val_adj < 0.01)
top2 <- Reduce(rbind,
                by(all.markers,
                   all.markers["cluster"],
                   head,
                   n = 2))
v1 <- VlnPlot(lymphoid,
             features = top2$gene,
             flip = TRUE,
             stack = TRUE,
             cols = cluster_colors,
             split.by = "seurat_clusters")
v1

v2 <- VlnPlot(lymphoid,
             features = c("Cd3e","Cd4","Cd8a","Il1rl1","Cd19","Ighd","Fcer2a",
                           "Igha","Cr2","Sdc1","Siglech"),
             flip = TRUE,
             stack = TRUE,
             cols = cluster_colors,
             split.by = "seurat_clusters")
v2

goi <- c("Pdgfrb","Acta2","Cldn5","Vwf","Pecam1","Ptprc","Cd19","Ms4a1","Sdc1",
         "Il7r","Trbc2","Cd3e","Thy1","Gata3","Ccr2","Itgax","Ly6c2","Ly6g",
         "H2-Eb1","Lyz2","Itgam","Mrc1","Mcpt4","Col1a2","Plp1")
v3 <- VlnPlot(lymphoid,
              features = goi,
              flip = TRUE,
              stack = TRUE,
              cols = cluster_colors,
              split.by = "seurat_clusters")
v3

d1 <- DotPlot(lymphoid,
              features = c("Cd3e","Cd4","Cd8a","Il1rl1","Cd19","Ighd","Fcer2a",
                           "Igha","Cr2","Sdc1","Siglech","Mrc1")) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
d1
```

```{r save_unannotated_lymphoid_markers, eval=FALSE, echo=FALSE}
path <- "../../results/pass_1/lymphoid_recluster/markers/unannotated_top2_markers_per_cluster_violin.pdf"
pdf(path, height = 4, width = 6)
v1
dev.off()

path <- "../../results/pass_1/lymphoid_recluster/markers/unannotated_stacked_violin.pdf"
pdf(path, height = 6, width = 8)
v3
dev.off()

path <- "../../results/pass_1/lymphoid_recluster/markers/unannotated_cluster_markers_dot_plot.pdf"
pdf(path, height = 4, width = 6)
d1
dev.off()
```

## Assign identities
```{r assign_lymphoid_identities, eval=FALSE}
lymphoid <- RenameIdents(object = lymphoid,
                       "0" = "T cells",
                       "1" = "B cells",
                       "2" = "Innate lymphoid cells",
                       "3" = "T cells",
                       "4" = "B cells",
                       "5" = "Innate lymphoid cells",
                       "6" = "B cells",
                       "7" = "Plasma cells",
                       "8" = "T cells",
                       "9" = "T cells", 
                       "10" = "T cells")
lymphoid$seurat_clusters <- factor(Idents(lymphoid),
                                 levels = c("B cells",
                                            "Plasma cells",
                                            "T cells",
                                            "Innate lymphoid cells"))
Idents(lymphoid) <- "seurat_clusters"
```

## Annotated UMAP
```{r lymphoid_annotated_umap,eval=FALSE}
cluster_colors <- c("gold","red","cyan","blue","green")
u1 <- DimPlot(lymphoid,
              label = TRUE,
              cols = cluster_colors,
              reduction = "umap")
u1
u2 <- DimPlot(lymphoid,
              label = TRUE,
              cols = cluster_colors,
              dims = c(2,3),
              reduction = "umap")
u2
```

```{r save_lymphoid_unannotated_umap, eval=FALSE, echo=FALSE}
path <- "../../results/pass_1/lymphoid_recluster/UMAP/annotated_clusters_dim"
pdf(paste0(path,"1&2.pdf"), height = 4, width = 6)
u1
dev.off()
pdf(paste0(path,"2&3.pdf"), height = 4, width = 6)
u2
dev.off()
```

## Markers
- Cd19: All B-cells except plasma cell \
- Cd138/Sdc1: Plasma cell \
- Siglech: Siglec-H is a DAP12-associated receptor on plasmacytoid dendritic cells (pDCs) and microglia \
- Cd20/Ms4a1: B cell marker that plays a role in the development and differentiation of B cells to plasma cells \
```{r annotated_bcell_markers, eval=FALSE}
d1 <- DotPlot(lymphoid,
              features = c("Il1rl1","Il7r","Cd3e","Cd4","Cd8a","Igha","Sdc1",
                           "Cd19","Ighd","Fcer2a","Cr2")) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
d1
```

```{r save_annotated_lymphoid_markers, eval=FALSE, echo=FALSE}
path <- "../../results/pass_1/lymphoid_recluster/markers/annotated_cluster_markers_dot_plot.pdf"
pdf(path, height = 4, width = 6)
d1
dev.off()
```

## Feature plots
```{r lymphoid_feature_plot,eval=FALSE}
f1 <- FeaturePlot(lymphoid,
            features = "nCount_RNA")  + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f1

f2 <- FeaturePlot(lymphoid,
            features = "nFeature_RNA")  + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f2
```

```{r save_lymphoid_feature_plot, eval=FALSE, echo=FALSE}
path <- "../../results/pass_1/lymphoid_recluster/UMAP/"
pdf(paste0(path,"nCount.pdf"), height = 4, width = 6)
f1
dev.off()

pdf(paste0(path,"nFeature.pdf"), height = 4, width = 6)
f2
dev.off()
```

## Shiny app
```{r cleanup_obj,eval=FALSE}
shiny.lymphoid <- lymphoid
shiny.lymphoid@assays$RNA@var.features <- 
  shiny.lymphoid@assays$SCT@var.features
metadata <- shiny.lymphoid@meta.data
metadata <- metadata[,c(21,22,2:14)]
colnames(metadata)[1] <- "unannotated_clusters"
shiny.lymphoid@meta.data <- metadata
shiny.lymphoid@assays$SCT@meta.features <- metadata
shiny.lymphoid@assays$RNA@meta.features <- metadata
```

```{r output_shiny_folder, eval=FALSE}
# make shiny folder
DefaultAssay(shiny.lymphoid) <- "RNA"
Idents(shiny.lymphoid) <- shiny.lymphoid$seurat_clusters
sc.config <- createConfig(shiny.lymphoid)
setwd("../../results/pass_1/lymphoid_recluster")
makeShinyApp(shiny.lymphoid, sc.config, gene.mapping = TRUE,
             shiny.title = "Pass 1: Lymphoid Recluster") 
```

# Myeloid + Other Recluster
## Recluster
```{r subset_recluster_myeloid, eval=FALSE}
myeloid <- subset_recluster(obj = mouse.annotated,
                            cellIdent = "merged_clusters",
                            cellType = c("Macrophages","Ambient RNAs","Endothelial",
                                         "Fibroblasts","Monocytes and Dendritic cells",
                                         "Neutrophils","Pericytes & Smooth muscle cells",
                                         "Mast cells","Schwann cells"),
                            projectID = "Myeloid Recluster",
                            keepCol = 4:14)
myeloid$seurat_clusters <- myeloid$SCT_snn_res.0.1
Idents(myeloid) <- "seurat_clusters"
saveRDS(myeloid, "../../rObjects/pass1_myeloid_recluster_unannotated.rds")
```

```{r load_reclustered_myeloid_cells, echo=FALSE, eval=TRUE}
myeloid <- readRDS("../../rObjects/pass1_myeloid_recluster_unannotated.rds")
```

## Unannotated UMAP
```{r myeloid_unannotated_umap}
merged_colors <- c("darkred","firebrick1","gold","darkolivegreen2","darkgreen",
                   "cyan","cornflowerblue","blue","darkorchid1","deeppink",
                   "plum1","burlywood3","azure4", "black")

u1 <- DimPlot(myeloid,
              label = TRUE,
              group.by = "seurat_clusters",
              cols = merged_colors,
              reduction = "umap")
u1

u2 <- DimPlot(myeloid,
              label = TRUE,
              cols = merged_colors,
              group.by = "seurat_clusters",
              dims = c(2,3),
              reduction = "umap")
u2
```

```{r save_unannotated_myeloid_umap, eval=FALSE, eval=FALSE}
path <- "../../results/pass_1/myeloid_recluster/UMAP/unannotated_umap_dim"
pdf(paste0(path,"1&2.pdf"), height = 4, width = 6)
u1
dev.off()
pdf(paste0(path,"2&3.pdf"), height = 4, width = 6)
u2
dev.off()
```

## Potential markers
```{r find_unannotated_myeloid_markers, eval=FALSE}
all.markers <- find_cluster_markers(obj = myeloid,
                                    cellIdent = "seurat_clusters")
saveRDS(all.markers, "../../rObjects/pass1_myeloid_recluster_unannotated_cluster_markers.rds")

all.markers <- subset(all.markers, p_val_adj < 0.01)
path <- "../../results/pass_1/myeloid_recluster/markers/"
write.table(all.markers,
            paste0(path,"unannotated_cluster_markers_adjpval_0.01.tsv"),
            quote = FALSE, row.names = FALSE, sep = "\t")
```

```{r myeloid_top_markers}
all.markers <- 
  readRDS("../../rObjects/pass1_myeloid_recluster_unannotated_cluster_markers.rds")
all.markers <- subset(all.markers, p_val_adj < 0.01)

top2 <- Reduce(rbind,
               by(all.markers,
                  all.markers["cluster"],
                  head,
                  n = 2))

top20 <- Reduce(rbind,
               by(all.markers,
                  all.markers["cluster"],
                  head,
                  n = 20))

Idents(myeloid) <- "seurat_clusters"
v1 <- VlnPlot(myeloid,
              features = top2$gene,
              flip = TRUE,
              stack = TRUE,
              cols = cluster_colors,
              split.by = "seurat_clusters")
v1

v2 <- VlnPlot(myeloid,
              features = c("C1qb","Flt1","Col1a1","Itgax","Fgr","Retnlg","Rgs5",
                           "Kit","Mpz"),
              flip = TRUE,
              stack = TRUE,
              cols = cluster_colors,
              split.by = "seurat_clusters")
v2

goi <- c("Pdgfrb","Acta2","Cldn5","Vwf","Pecam1","Ptprc","Cd19","Ms4a1","Sdc1",
         "Il7r","Trbc2","Cd3e","Thy1","Gata3","Ccr2","Itgax","Ly6c2","Ly6g",
         "H2-Eb1","Lyz2","Itgam","Mrc1","Mcpt4","Col1a2","Plp1")
v3 <- VlnPlot(myeloid,
              features = goi,
              flip = TRUE,
              stack = TRUE,
              cols = merged_colors,
              split.by = "seurat_clusters")
v3

d1 <- DotPlot(myeloid,
              features = c("Flt1","Ly6c1","Mrc1","C1qb","Ttr","Igha","Col1a1",
                           "Col1a2","Itgax","Fgr","Retnlg","Kit","Rgs5","Acta2",
                           "Mpz","Mbp")) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
d1
```

```{r save_unannotated_myeloid_markers, eval=FALSE, echo=FALSE}
path <- "../../results/pass_1/myeloid_recluster/markers/unannotated_top2_markers_per_cluster_violin.pdf"
pdf(path, height = 4, width = 6)
v1
dev.off()

path <- "../../results/pass_1/myeloid_recluster/markers/unannotated_stacked_violin.pdf"
pdf(path, height = 6, width = 8)
v3
dev.off()

path <- "../../results/pass_1/myeloid_recluster/markers/unannotated_cluster_markers_dot_plot.pdf"
pdf(path, height = 4, width = 6)
d1
dev.off()
```

## Assign identities
```{r assign_myeloid_identities}
myeloid <- RenameIdents(object = myeloid,
                     "0" = "Endothelial",
                     "1" = "Macrophages",
                     "2" = "Ambient RNAs",
                     "3" = "Fibroblasts",
                     "4" = "Macrophages",
                     "5" = "Monocytes and Dendritic cells",
                     "6" = "Neutrophils",
                     "7" = "Mast cells",
                     "8" = "Pericytes & Smooth muscle cells",
                     "9" = "Schwann cells")
myeloid$seurat_clusters <- Idents(myeloid)
Idents(myeloid) <- "seurat_clusters"
```

## Annotated UMAP
```{r annotated_myeloid_umap}
u1 <- DimPlot(myeloid,
              cols = merged_colors,
              reduction = "umap")
u1

u2 <- DimPlot(myeloid,
              cols = merged_colors,
              dims = c(2,3),
              reduction = "umap")
u2
```

```{r save_myeloid_annotated_umap, eval=FALSE, echo=FALSE}
path <- "../../results/pass_1/myeloid_recluster/UMAP/annotated_umap_dim"
pdf(paste0(path,"1&2.pdf"), height = 4, width = 7)
u1
dev.off()
pdf(paste0(path,"2&3.pdf"), height = 4, width = 7)
u2
dev.off()
```

## Markers
```{r annotated_myeloid_cluster_markers}
Idents(myeloid) <- "seurat_clusters"
v <- VlnPlot(myeloid,
             features = c("Flt1","C1qb","Ttr","Col1a1","Itgax","Fgr","Retnlg",
                          "Rgs5","Mpz"),
             flip = TRUE,
             stack = TRUE,
             cols = merged_colors,
             split.by = "seurat_clusters")
v

d1 <- DotPlot(myeloid,
              features = c("Flt1","Ly6c1","Mrc1","C1qb","Ttr","Igha","Col1a1",
                           "Col1a2","Itgax","Fgr","Retnlg","Kit","Rgs5","Acta2",
                           "Mpz","Mbp")) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
d1
```

```{r save_annotated_myeloid_marks, eval=FALSE, echo=FALSE}
path <- "../../results/pass_1/myeloid_recluster/markers/annotated_cluster_markers_dot_plot"
pdf(paste0(path,".pdf"), height = 6, width = 8)
d1
dev.off()
```

## Feature plots
```{r myeloid_feature_plots}
f1 <- FeaturePlot(myeloid,
                  features = "nFeature_RNA")  + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f1
f2 <- FeaturePlot(myeloid,
                  features = "nCount_RNA")  + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f2
```

```{r save_myeloid_feature_plots,eval=FALSE,echo=FALSE}
path <- "../../results/pass_1/myeloid_recluster/UMAP/nFeature_RNA"
pdf(paste0(path,".pdf"), height = 4, width = 6)
f1
dev.off()

path <- "../../results/pass_1/myeloid_recluster/UMAP/nCount_RNA"
pdf(paste0(path,".pdf"), height = 4, width = 6)
f2
dev.off()
```

## Shiny app
```{r cleanup_obj,eval=FALSE}
shiny.myeloid <- myeloid
shiny.myeloid@assays$RNA@var.features <- 
  shiny.myeloid@assays$SCT@var.features
metadata <- shiny.myeloid@meta.data
metadata <- metadata[,c(22,17,2:14)]
shiny.myeloid@meta.data <- metadata
shiny.myeloid@assays$SCT@meta.features <- metadata
shiny.myeloid@assays$RNA@meta.features <- metadata
```

```{r output_shiny_folder, eval=FALSE}
# make shiny folder
DefaultAssay(shiny.myeloid) <- "RNA"
Idents(shiny.myeloid) <- shiny.myeloid$seurat_clusters
sc.config <- createConfig(shiny.myeloid)
setwd("../../results/pass_1/myeloid_recluster")
makeShinyApp(shiny.myeloid, sc.config, gene.mapping = TRUE,
             shiny.title = "Pass 1: Myeloid + Remaining Cell Types Recluster") 
```

## Cells to remove
```{r}
table(myeloid$seurat_clusters)
remove <- subset(myeloid, seurat_clusters == "Ambient RNAs")
table(remove$seurat_clusters)
barcodes <- colnames(remove)
saveRDS(barcodes,"../../rObjects/pass1_barcodes_to_remove.rds")
```

# New Object
```{r}
mouse.annotated <- readRDS("../../rObjects/pass1_annotated.rds")
Idents(mouse.annotated) <- "merged_clusters"
DefaultAssay(mouse.annotated) <- "RNA"

barcodes <- readRDS("../../rObjects/pass1_barcodes_to_remove.rds")

mouse.annotated$is_ambient <- colnames(mouse.annotated) %in% barcodes
table(mouse.annotated$is_ambient)

saveRDS(mouse.annotated, "../../rObjects/pass1_annotated_ambient_rnas.rds")
```
