---
title: "E3/E4 Mice Meningeal Dura scRNAseq"
subtitle: "Pass 2 Recluster"
author: "Kennedi Todd"
date: "09/17/2024"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

# Setup
```{r working_directory, echo=FALSE}
knitr::opts_knit$set(root.dir = ".")
```

## Libraries
```{r libraries, message=FALSE, warning=FALSE}
# load packages
library(ggplot2)        # geom_vline()
library(ggtree)         # BuildClusterTree()
library(Matrix)         # rowSums()
library(Seurat)         # Idents()
library(SeuratWrappers) # RunPrestoAll()
```

## Variables and functions
```{r set_variables_and_functions}
# variables
out <- "../../results/pass_2/all_clusters/"
sample_order <- c("E3.M","E3.F","E4.M","E4.F")
sample_colors <- c("#26946A","#1814A1","#EAC941","#DF5F00")
sample_order2 <- c("Male E3", "Male E4", "Female E3", "Female E4")
isoform_order <- c("E4","E3")
isoform_colors <- c("darkgray","cornflowerblue")
sex_order <- c("Male","Female")
sex_colors <- c("darkgray","purple")

# single cell functions
source("../../refs/Kennedi_single_cell_functions.R")

# save function
saveToPDF <- function(...) {
    d = dev.copy(pdf,...)
    dev.off(d)
}
```

## Load data
```{r read_pass1_annotated_ambient_RNAs}
mouse.annotated <- readRDS("../../rObjects/pass1_annotated.rds")
```

# Filter
## Cells
```{r filter_cells}
# load unwanted barcodes
barcodes_to_remove <- mouse.annotated@meta.data
barcodes_to_remove <- barcodes_to_remove[barcodes_to_remove$annotated_clusters == "Ambient RNAs" |
                       barcodes_to_remove$annotated_clusters == "Ambient RNAs?",]
barcodes_to_remove <- rownames(barcodes_to_remove)

# FALSE when cell is in barcode_to_remove
mouse.annotated$keep <- !rownames(mouse.annotated@meta.data) %in% barcodes_to_remove
table(mouse.annotated$keep)

# filter
mouse.filtered <- subset(mouse.annotated,
                         subset = (keep == TRUE))

# check
mouse.filtered

# cleanup
remove(mouse.annotated)
mouse.filtered$keep <- NULL
```

## Genes
```{r}
# filter genes
mouse.filtered <- JoinLayers(mouse.filtered)
counts <- GetAssayData(object = mouse.filtered, layer = "counts")
nonzero <- counts > 0  # produces logical
keep <- Matrix::rowSums(nonzero) >= 10  # sum the true/false
counts.filtered <- counts[keep,]  # keep certain genes

# remove hemoglobin genes (appear to be clustering by percent.hb)
keep <- !rownames(counts.filtered) %in% c("Hbb-bt", "Hbb-bs", "Hbb-bh2","Hbb-bh1", 
                                          "Hbb-y", "Hba-x", "Hba-a1","Hba-a2")
counts.filtered <- counts.filtered[keep,]

# overwrite mouse.filtered
mouse.filtered <- CreateSeuratObject(counts.filtered, 
                                     meta.data = mouse.filtered@meta.data)

# print features removed
print(paste0(dim(counts)[1] - dim(counts.filtered)[1], " features removed"))

# preview new object
mouse.filtered
```

# Recluster
## SCTransform
- When using Seurat v5 assays you can keep all the data in one object and split
the layers (unlike Seurat v4). \
- Use SCTransform for normalization, variable features, and scaling \
```{r sctransform, warning=FALSE}
# split layers by sample
mouse.filtered[["RNA"]] <- split(mouse.filtered[["RNA"]], f = mouse.filtered$sample)
# check layers are split
mouse.filtered
# SCTransform
mouse.filtered <- SCTransform(mouse.filtered, verbose = FALSE)
```

## PCA
```{r pca, warning=FALSE, message=FALSE}
mouse.filtered <- RunPCA(mouse.filtered, assay = "SCT")
```

## Harmony
- Seurat v5 enables streamlined integrative analysis using the IntegrateLayers function. \
- The method currently supports five integration methods and we will be using the Harmony method. \
- All five methods will integrate data in low-dimensional space and returns a dimensional reduction that aims to co-embed shared cell types across batches. \
```{r harmony, warning=FALSE, message=FALSE}
# Integrate samples
mouse.filtered <- IntegrateLayers(object = mouse.filtered,
                                  method = HarmonyIntegration,
                                  orig.reduction = "pca",
                                  assay = "SCT")

# Plot harmony reduction
DimPlot(mouse.filtered,
        group.by = "sample",
        reduction = "harmony",
        cols = sample_colors,
        shuffle = TRUE)

# Harmony elbow plot
mouse.filtered@reductions$harmony@stdev <-
  apply(mouse.filtered@reductions$harmony@cell.embeddings, 2, sd)
ElbowPlot(mouse.filtered,
          reduction = "harmony") +
  geom_vline(xintercept = 15, linetype = "dashed", color = "red")
```

## UMAP
```{r run_umap}
# run UMAP
mouse.filtered <- RunUMAP(mouse.filtered,
                          dims = 1:15,
                          reduction = "harmony",
                          assay = "SCT",
                          n.components = 3) # set to 3 to use with VR
DimPlot(mouse.filtered, reduction = "umap")
```

## Cluster
```{r cluster}
# k-nearest neighbor graph
mouse.unannotated <- FindNeighbors(object = mouse.filtered,
                                   assay = "SCT",
                                   reduction = "harmony",
                                   dims = 1:15)

# shared nearest neighbor graph and community detection
mouse.unannotated <- FindClusters(object = mouse.unannotated,
                                  algorithm = 1, # 1 = Louvain
                                  resolution = seq(0.1, 0.5, by=0.1))

# set desired resolution
mouse.unannotated$seurat_clusters <- mouse.unannotated$SCT_snn_res.0.3

# set default values
mouse.unannotated <- JoinLayers(mouse.unannotated, assay = "RNA")
Idents(mouse.unannotated) <- "seurat_clusters"
DefaultAssay(mouse.unannotated) <- "RNA"
mouse.unannotated <- NormalizeData(mouse.unannotated)
```


## Explore resolutions
```{r explore_resolutions}
# 0.1
DimPlot(mouse.unannotated,
        group.by = "SCT_snn_res.0.1",
        label = TRUE)

# 0.2
DimPlot(mouse.unannotated,
        group.by = "SCT_snn_res.0.2",
        label = TRUE)

# 0.3
DimPlot(mouse.unannotated,
        group.by = "SCT_snn_res.0.3",
        label = TRUE)

# 0.4
DimPlot(mouse.unannotated,
        group.by = "SCT_snn_res.0.4",
        label = TRUE)

# 0.5
DimPlot(mouse.unannotated,
        group.by = "SCT_snn_res.0.5",
        label = TRUE)
```

```{r save_unannotated_obj, eval=FALSE, echo=FALSE}
# set default params
Idents(mouse.unannotated) <- "seurat_clusters"
# save object
saveRDS(mouse.unannotated, "../../rObjects/pass2_unannotated.rds")
# cleanup
remove(counts,counts.filtered,nonzero,mouse.filtered)
```

# Unannotated QC
## UMAP
```{r unannotated_pass2_umap}
Idents(mouse.unannotated) <- "seurat_clusters"

cluster_colors <- c("red","darkorange","purple","yellow","chocolate4",
                    "lightgreen","chartreuse3","darkgreen",
                    "cyan","steelblue","blue",
                    "gold2","deeppink","pink",
                    "tan","salmon","gray","black",
                    "red4")

u1 <- DimPlot(mouse.unannotated,
              label = TRUE,
              cols = cluster_colors,
              reduction = "umap")
u1

u2 <- DimPlot(mouse.unannotated,
              label = TRUE,
              cols = cluster_colors,
              dims = c(2,3),
              reduction = "umap")
u2
```

```{r save_pass2_unannotated_umap, echo=FALSE, eval=FALSE}
u1
saveToPDF(paste0(out, "UMAP/unannotated_clusters_dim1&2.pdf"), width = 8, height = 4)
dev.off()

u2
saveToPDF(paste0(out, "UMAP/unannotated_clusters_dim2&3.pdf"), width = 8, height = 4)
dev.off()
```

## Feature plots
```{r feature_plots_unannotated, message=FALSE, warning=FALSE}
# UMAP percent.mt
FeaturePlot(mouse.unannotated,
            reduction = "umap", 
            features = "percent.mt") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))

# UMAP percent.ribo
FeaturePlot(mouse.unannotated,
            reduction = "umap", 
            features = "percent.ribo.protein") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))

# UMAP percent.hb
FeaturePlot(mouse.unannotated,
            reduction = "umap", 
            features = "percent.hb") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))

# UMAP nCount
FeaturePlot(mouse.unannotated,
            reduction = "umap", 
            features = "nCount_RNA") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))

# UMAP nFeature
FeaturePlot(mouse.unannotated,
            reduction = "umap", 
            features = "nFeature_RNA") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))

# UMAP cell.complexity
FeaturePlot(mouse.unannotated,
            reduction = "umap", 
            features = "cell.complexity") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))

# UMAP Ttr expression
FeaturePlot(mouse.unannotated,
            reduction = "umap", 
            features = "C1qb") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
```

## Violins
```{r}
VlnPlot(mouse.unannotated,
        features = "nCount_RNA",
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        features = "nFeature_RNA",
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        stack = TRUE,
        flip = TRUE,
        features = c("C1qb","C1qa","Flt1","Myh11"),
        split.by = "seurat_clusters")
```

## Cells per cluster
```{r cells_per_cluster}
# Cells per sample per cluster
sample_ncells <- FetchData(mouse.unannotated, 
                     vars = c("ident", "sample")) %>%
  dplyr::count(ident,sample) %>%
  tidyr::spread(ident, n)
sample_ncells

# Cells per isoform per cluster
isoform_ncells <- FetchData(mouse.unannotated, 
                     vars = c("ident", "isoform")) %>%
  dplyr::count(ident,isoform) %>%
  tidyr::spread(ident, n)
isoform_ncells

# Cells per sex per cluster
sex_ncells <- FetchData(mouse.unannotated, 
                     vars = c("ident", "sex")) %>%
  dplyr::count(ident,sex) %>%
  tidyr::spread(ident, n)
sex_ncells
```

# Finding markers
## Cluster tree
- Cluster trees are helpful in deciding what clusters to merge.
```{r cluster_tree_unannotated, message=FALSE, warning=FALSE}
mouse.unannotated <- BuildClusterTree(object = mouse.unannotated,
                                     dims = 1:15,
                                     reorder = FALSE,
                                     reorder.numeric = FALSE)
tree <- mouse.unannotated@tools$BuildClusterTree
tree$tip.label <- paste0(tree$tip.label)

ggtree::ggtree(tree, aes(x, y)) +
  scale_y_reverse() +
  ggtree::geom_tree() +
  ggtree::theme_tree() +
  ggtree::geom_tiplab(offset = 1) +
  ggtree::geom_tippoint(color = cluster_colors[1:length(tree$tip.label)], shape = 16, size = 5) +
  coord_cartesian(clip = 'off') +
  theme(plot.margin = unit(c(0,2.5,0,0), 'cm'))
```

## Automatically detect markers
```{r find_markers_unannotated}
Idents(mouse.unannotated) <- "seurat_clusters"
all.markers <- SeuratWrappers::RunPrestoAll(
  object = mouse.unannotated,
  assay = "RNA",
  slot = "counts",
  only.pos = TRUE
)

all.markers <- all.markers[all.markers$p_val_adj < 0.01,]

top2 <- Reduce(rbind,
               by(all.markers,
                  all.markers["cluster"],
                  head,
                  n = 2))
top20 <- Reduce(rbind,
               by(all.markers,
                  all.markers["cluster"],
                  head,
                  n = 20))

write.table(all.markers, 
            paste0(out, "markers/unannotated_auto_find_markers_adjpval_0.01.tsv"),
            quote = FALSE,
            row.names = FALSE)
```

## Violins
```{r violins_unannotated}
v1 <- VlnPlot(mouse.unannotated,
              group.by = "seurat_clusters",
              split.by = "seurat_clusters",
              stack = TRUE,
              flip = TRUE,
              cols = cluster_colors,
              features = top2$gene)
v1

goi <- c("Pdgfrb","Acta2","Vwf","Cldn5","Pecam1","Flt4","Prox1","Lyve1","Ptprc",
         "Cd19","Ms4a1","Ighd","Cd3e","Trbc2","Il7r","Nkg7","Klrb1b","Klrb1c",
         "Gata3","Rora","Itgax","H2-Eb1","Ccr2","Ly6c2","Lyz2","Ly6g","Itgam",
         "Mrc1","Csf1r","Cd38","Mki67","Mcpt4","Ms4a2","Col1a2","Plp1")
v2 <- VlnPlot(mouse.unannotated,
              group.by = "seurat_clusters",
              split.by = "seurat_clusters",
              stack = TRUE,
              flip = TRUE,
              cols = cluster_colors,
              features = goi)
v2

v3 <- VlnPlot(mouse.unannotated,
              group.by = "seurat_clusters",
              split.by = "seurat_clusters",
              stack = TRUE,
              flip = TRUE,
              cols = cluster_colors,
              features = c("C1qb","C1qa","Flt1","Cldn5"))
v3
```

```{r save_violins_unannotated, echo=FALSE, eval=FALSE}
pdf(paste0(out, "markers/auto_top2_markers_violin_unannotated.pdf"), width = 12, height = 8)
v1
dev.off()

pdf(paste0(out, "markers/sandro_markers_violin_unannotated.pdf"), width = 12, height = 10)
v2
dev.off()
```

## Assign identities
```{r assign_identities}
# Rename all identities
Idents(mouse.unannotated) <- "seurat_clusters"
mouse.annotated <- RenameIdents(object = mouse.unannotated,
                                "0" = "Macrophages",         # Aoh
                                "1" = "Endothelial",         # Dysf     
                                "2" = "Endothelial",         # Pf4 
                                "3" = "Macrophages",         # Adap2      
                                "4" = "Fibroblasts",         # Bnc2     
                                "5" = "DCs",                 # Gm15987,Fgr
                                "6" = "T cells",             # Skap1,Cd247,Cd3e
                                "7" = "Endothelial",         # Flt1,Ptprb
                                "8" = "Macro?Fibro?",
                                "9" = "B cells",             # Pax5,Ms4a1  
                                "10" = "Endothelial",        # Flt1,Ptprb 
                                "11" = "Endothelial",        # Vwf 
                                "12" = "Neutrophils",        # Cxcr2,Hdc
                                "13" = "Neutrophils",        # Cd177
                                "14" = "Pericytes and SMCs", # Notch3,Myh11
                                "15" = "ILCs",               # Il1rl1,Gata3
                                "16" = "T cells",            # Skap1,Cd3g,Cd3e
                                "17" = "Schwann cells",      # Nkain,Kcna1
                                "18" = "ILCs",               # Il1rl1,Gata3
                                "19" = "Mast cells",         # Cpa3,Hs6st2,Kit
                                "20" = "B cells",            # Pax5,Ms4a1
                                "21" = "Myeloid precurors",  # Mki67,Top2a
                                "22" = "B cells",
                                "23" = "Endothelail")        # Vwf,Cldn5
mouse.annotated$annotated_clusters <- Idents(mouse.annotated)
Idents(mouse.annotated) <- "annotated_clusters"
```
