---
title: "PLX5622 Diet Mice scRNAseq"
subtitle: "Filtering, Normalization, Clustering"
author: "Kennedi Todd"
date: "02/11/2025"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

# Experiment notes
16 scRNAseq samples of meningeal endothelial cells were sent for sequencing. In this experiment E3 and E4, male and female mice (12-month-old), were treated with a control diet or a PLX5622 diet (to ablate MHC-II high myeloid cells in the meninges). The goal is to see the effects on the transcriptome on BECs and LECs. There are 8 groups and 2 biological replicates per group (each replicate consisting of cells pooled from two animals). ~5000 cells per sample were submitted for sequencing at 50k read pairs per cell.

# Setup
## Set working directory
```{r working_directory}
knitr::opts_knit$set(root.dir = ".")
```

## Load libraries
```{r load_libraries, message=FALSE, warnings=FALSE}
library(cowplot)     # plot_grid()
library(dotenv)      # load_dot_env()
library(dplyr)       # left_join()
library(ggplot2)     # ggplot()
library(gridExtra)   # grid.arrange()
library(rtracklayer) # import()
library(Seurat)      # Read10X_h5()
library(stringr)     # str_match()
```

## Variables and functions
```{r set_variables_and_thresholds}
# levels, color, and output variables
out <- "../../results/"

sample_order <- c("E3CF1","E3CF2","E3CM1","E3CM2","E3PF1","E3PF2","E3PM1","E3PM2",
                  "E4CF1","E4CF2","E4CM1","E4CM2","E4PF1","E4PF2","E4PM1","E4PM2")
sample_colors <- c("#9e0909","#f75959","#f0b402","#f5d67a",
                   "#fbff0a","#fafaa7","#1d8c0e","#63e851",
                   "#0271f0","#9cc3f0","#6c39f7","#c589fa",
                   "#ed2af7","#f5c1f7","#755410","#b5aa82")

group_colors <- c("#f75959","#f0b402",
                   "#fbff0a","#63e851",
                   "#0271f0","#c589fa",
                   "#ed2af7","#b5aa82")
group2_colors <- c("#f75959","#fbff0a","#63e851","#0271f0")

isoform_order <- c("E4","E3")
isoform_colors <- c("darkgray","cornflowerblue")

sex_order <- c("Male","Female")
sex_colors <- c("green","purple")

diet_order <- c("control","PLX5622")
diet_color <- c("coral","cyan")

# filtering thresholds
nCount.min <- 500
nCount.max <- 25000
nFeature.min <- 100
complexity.cutoff <- 0.8
mt.cutoff <- 20
hb.cutoff <- 3

# environment variables
load_dot_env(file = "../../refs/.env")
refs_dir <- Sys.getenv("ANNOTATION_REFS")

# single cell functions
files <- list.files("../../functions", full.names = TRUE)
invisible(lapply(files, source))
```

# Load data
```{r create_seurat_obj}
seurat_obj_list <- list()

if (file.exists(paste0("../../rObjects/raw_seurat_obj.rds"))) {
  mouse <- readRDS(paste0("../../rObjects/raw_seurat_obj.rds"))
} else {
  # path info
  prefix <- "../../counts/"
  suffix <- "/outs/filtered_feature_bc_matrix.h5"
  
  # create list of individual seurat objects
  for (sample in sample_order) {
    # print sample you're on
    print(sample)
    
    # Create Seurat object
    path <- paste0(prefix, sample, suffix)
    obj <- CreateSeuratObject(Read10X_h5(path))
    
    # Add sample ID as prefix to cell names
    obj <- RenameCells(obj, add.cell.id = sample)
    
    # Add Seurat object to the list with the sample name as the key
    seurat_obj_list[[sample]] <- obj
  }
  
  # Merge all Seurat objects
  mouse <- merge(seurat_obj_list[[1]], 
                 y = seurat_obj_list[-1])
  
  # Set project name
  mouse@project.name <- "PLX5622 Diet Mice scRNAseq"
  
  # Join layers
  mouse$orig.ident <- colnames(mouse)
  mouse <- JoinLayers(mouse)
  
  # Extract sample
  mouse$sample <- str_match(colnames(mouse), "(E[34][CP][FM][12])_")[,2]
  
  # Check
  table(mouse$sample)
  
  # save
  saveRDS(mouse, "../../rObjects/raw_seurat_obj.rds")
  
  # cleanup
  remove(meta, seurat_obj_list, obj)
  gc()
}

# preview
mouse
```

## Annotation
- Annotation file was downloaded from 10x Genomics \
  + refdata-gex-mm10-2020-A \
  + provider: GENCODE \
  + description: evidence-based annotation of the mouse genome (GRCm38), version M23 (Ensembl 98) \
- Gm* genes are originally annotated by MGI and the *Rik genes are annotated by RIKEN \
```{r read_annotation}
# read in annotation file, GENCODE GRCm38 version M23 (Ensembl 98)
if (file.exists("../../rObjects/annotation.rds")) {
  genes <- readRDS("../../rObjects/annotation.rds")
} else {
  gtf.file <- paste0(refs_dir, "/mouse/refdata-gex-mm10-2020-A/genes/genes.gtf")
  genes <- rtracklayer::import(gtf.file)
  genes <- as.data.frame(genes)
  genes <- genes[genes$type == "gene",]
  saveRDS(genes, "../../rObjects/annotation.rds")
}
```

## Metadata columns
```{r add_metadata_columns}
# create sample column
mouse$sample <- factor(mouse$sample, levels = sample_order)
Idents(mouse) <- mouse$sample

# age column
mouse$age <- "12 months"

# sex column
sex <- str_match(mouse$sample, "E[34][CP]([MF])[12]")[,2]
sex <- gsub("F","Female",sex)
sex <- gsub("M","Male",sex)
mouse$sex <- factor(sex, levels = sex_order)

# Apoe isoform column
isoform <- str_match(mouse$sample, "(E[34])[CP][MF][12]")[,2]
mouse$isoform <- factor(isoform, levels = isoform_order)

# Diet
diet <- str_match(mouse$sample, "E[34]([CP])[MF][12]")[,2]
diet <- gsub("C","control",diet)
diet <- gsub("P","PLX5622",diet)
mouse$diet <- factor(diet, levels = diet_order)

# cell.complexity
mouse$cell.complexity <- log10(mouse$nFeature_RNA) / log10(mouse$nCount_RNA)

# percent.mt
mt.genes <- genes[genes$seqnames == "chrM",13]
mouse$percent.mt <- PercentageFeatureSet(mouse, features = mt.genes)
mt.genes

# percent.ribo
# ribosomal proteins begin with 'Rps' or 'Rpl' in this annotation file
# mitochondrial ribosomes start with 'Mrps' or 'Mrpl'
gene.names <- genes$gene_name
ribo <- gene.names[grep("^Rp[sl]", gene.names)]
mt.ribo <- gene.names[grep("^Mrp[sl]", gene.names)]
ribo.combined <- c(mt.ribo,ribo)
mouse$percent.ribo.protein <- PercentageFeatureSet(mouse, features = ribo.combined)
ribo.combined

# percent.hb
# percent.hb - hemoglobin proteins begin with 'Hbb' or 'Hba' for mouse
hb.genes <- gene.names[grep("^Hb[ba]-", gene.names)]
mouse$percent.hb <- PercentageFeatureSet(mouse, features = hb.genes)
hb.genes
```

# Pre-filtering QC
## Number of cells
```{r prefiltering_cells_per_sample}
# Visualize the number of cell counts per sample
data <- as.data.frame(table(mouse$sample))
colnames(data) <- c("sample","frequency")

ncells1 <- ggplot(data, aes(x = sample, y = frequency, fill = sample)) + 
  geom_col() +
  theme_classic() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            vjust=-0.25) +
  scale_fill_manual(values = sample_colors) + 
  scale_y_continuous(breaks = seq(0,10000, by = 2000), limits = c(0,10000)) +
  ggtitle("Raw: cells per sample") +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))
ncells1
```

## Density plots
```{r prefiltering_density, warning=FALSE}
# Visualize nCount_RNA
den1 <- ggplot(mouse@meta.data,
       aes(color = sample,
           x = nCount_RNA,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  scale_color_manual(values = sample_colors) +
  scale_fill_manual(values = sample_colors) +
  xlab("nCount_RNA") +
  ylab("Density") +
  theme(legend.position =  "none") +
  geom_vline(xintercept = nCount.min, lty = "dashed") +
  geom_vline(xintercept = nCount.max, lty = "dashed") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Visualize nFeature_RNA
den2 <- ggplot(mouse@meta.data,
       aes(color = sample,
           x = nFeature_RNA,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  scale_color_manual(values = sample_colors) +
  theme(legend.position =  "none") +
  scale_fill_manual(values = sample_colors) +
  xlab("nFeature_RNA") +
  ylab("Density") +
  geom_vline(xintercept = nFeature.min, lty = "dashed") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Visualize cell complexity
# Quality cells are usually above 0.85
den3 <- ggplot(mouse@meta.data,
       aes(color = sample,
           x = cell.complexity,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_color_manual(values = sample_colors) +
  theme(legend.position =  "none") +
  scale_fill_manual(values = sample_colors) +
  xlab("Cell Complexity (log10(nFeature/nCount))") +
  ylab("Density") +
  geom_vline(xintercept = complexity.cutoff, lty = "dashed") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Visualize percent.mt
den4 <- ggplot(mouse@meta.data,
       aes(color = sample,
           x = percent.mt,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_continuous(n.breaks = 4) +
  geom_vline(xintercept = mt.cutoff, lty = "dashed") +
  scale_color_manual(values = sample_colors) +
  theme(legend.position =  "none") +
  scale_fill_manual(values = sample_colors) +
  xlab("% Mitochondrial Genes") +
  ylab("Density") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Visualize percent.ribo.protein
den5 <- ggplot(mouse@meta.data,
       aes(color = sample,
           x = percent.ribo.protein,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  scale_color_manual(values = sample_colors) +
  theme(legend.position =  "none") +
  scale_fill_manual(values = sample_colors) +
  xlab("% Ribosomal Protein Genes") +
  ylab("Density") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Visualize percent.hb
den6 <- ggplot(mouse@meta.data,
       aes(color = sample,
           x = percent.hb,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  geom_vline(xintercept = hb.cutoff, lty = "dashed") +
  scale_color_manual(values = sample_colors) +
  theme(legend.position =  "none") +
  scale_fill_manual(values = sample_colors) +
  xlab("% Hemoglobin Genes") +
  ylab("Density") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Arrange graphs in grid
plots <- list(den1,den2,den3,den4,den5,den6)
layout <- rbind(c(1,4),c(2,5),c(3,6))
grid <- grid.arrange(grobs = plots, layout_matrix = layout)
```

```{r save_prefiltering_density,echo=FALSE,eval=FALSE}
# save
path <- paste0(out, "filtering_QC/density_curves_raw.pdf")
pdf(path, width = 8, height = 8)
grid <- grid.arrange(grobs = plots, layout_matrix = layout)

# cleanup
remove(den1,den2,den3,den4,den5,den6,layout,grid,plots,data)
```

## Violin plots
```{r prefiltering_violins, warning=FALSE}
# nFeature, nCount, and cell.complexity violins
v1 <- VlnPlot(mouse,
              features = c("nFeature_RNA", "nCount_RNA","cell.complexity"),
              ncol = 3,
              group.by = 'sample',
              cols = sample_colors,
              pt.size = 0)
v1

#  percent violins
v2 <- VlnPlot(mouse,
              features = c("percent.mt","percent.ribo.protein","percent.hb"),
              ncol = 3,
              group.by = 'sample',
              cols = sample_colors,
              pt.size = 0)
v2
```

```{r save_prefiltering_violins, echo=FALSE, eval=FALSE}
# save v1
path <- paste0(out, "filtering_QC/violins_nFeature_nCount_complexity_raw.pdf")
pdf(path, width = 14, height = 4)
v1
dev.off()

# save v2
path <- paste0(out, "filtering_QC/violins_percent_raw.pdf")
pdf(path, width = 14, height = 4)
v2
dev.off()

# cleanup
remove(v1,v2)
```

## Scatter plots
```{r prefiltering_scatter, warning=FALSE}
s1 <- ggplot(
  mouse@meta.data,
  aes(x = nCount_RNA, y = nFeature_RNA, color = percent.mt)) + 
  geom_point() + 
  geom_smooth(method = "lm") +
	scale_x_log10() +   	
  scale_y_log10() + 
  theme_classic() +
  geom_vline(xintercept = nCount.min, linetype = "dashed", colour = "red") + 
  geom_hline(yintercept = nFeature.min, linetype = "dashed", colour = "red") + 
  facet_wrap(~sample) +
  scale_colour_gradient(low = "gray90", high = "black", limits =c(0,100))
s1

s2 <- FeatureScatter(mouse,
                     feature1 = "nCount_RNA",
                     feature2 = "percent.mt",
                     group.by = 'sample',
                     cols = sample_colors,
                     shuffle = TRUE)
s2
```

```{r save_prefiltering_scatter, echo=FALSE, eval=FALSE}
# save
path <- paste0(out, "filtering_QC/scatter_nFeature_vs_nCount_raw.pdf")
pdf(path, width = 8, height = 8)
s1
dev.off()

path <- paste0(out, "filtering_QC/scatter_percentMT_vs_nCount_raw.pdf")
pdf(path, width = 6, height = 6)
s2
dev.off()

# cleanup
remove(s1,s2)
```

# Filtering
## Cell-level filtering
```{r cell_filtering}
# filter
mouse.filtered <- subset(mouse,
                        subset = (nCount_RNA > nCount.min) &
                          (nCount_RNA < nCount.max) &
                          (nFeature_RNA > nFeature.min) &
                          (cell.complexity > complexity.cutoff) &
                          (percent.mt < mt.cutoff) &
                          (percent.hb < hb.cutoff))

# print cells removed
print(paste0(dim(mouse)[2] - dim(mouse.filtered)[2]," cells removed"))
```

## Gene-level filtering
Remove lowly expressed genes.  We will keep genes that have at least 1 count in 10 cells.
```{r gene_filtering, warning=FALSE, message=FALSE}
# filter genes
counts <- GetAssayData(object = mouse.filtered, layer = "counts")
nonzero <- counts > 0  # produces logical
keep <- Matrix::rowSums(nonzero) >= 10  # sum the true/false
counts.filtered <- counts[keep,]  # keep certain genes

# overwrite mouse.filtered
mouse.filtered <- CreateSeuratObject(counts.filtered, 
                                     meta.data = mouse.filtered@meta.data)

# print features removed
print(paste0(dim(counts)[1] - dim(counts.filtered)[1], " features removed"))
```


```{r remove_specific_genes}
# remove mt.genes
counts <- GetAssayData(object = mouse.filtered, layer = "counts")
keep <- !rownames(counts) %in% mt.genes # false when mt.gene
counts.filtered <- counts[keep,]

# overwrite mouse.filtered
mouse.filtered <- CreateSeuratObject(counts.filtered,
                                     meta.data = mouse.filtered@meta.data)

# print features removed
print(paste0(dim(counts)[1] - dim(counts.filtered)[1], " features removed"))

# cleanup data
remove(mouse,counts,counts.filtered,nonzero)
```

```{r save_filtered_object, echo=FALSE, eval=FALSE}
saveRDS(mouse.filtered, "../../rObjects/filtered_seurat_obj.rds")
#mouse.filtered <- readRDS("../../rObjects/filtered_seurat_obj.rds")
```

# Post-filtering QC
## Number of cells
```{r postfiltering_cells_per_sample}
# Visualize the number of cell counts per sample
data <- as.data.frame(table(mouse.filtered$sample))
colnames(data) <- c("sample","frequency")

ncells2 <- ggplot(data, aes(x = sample, y = frequency, fill = sample)) + 
  geom_col() +
  theme_classic() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            vjust=-0.25) +
  scale_fill_manual(values = sample_colors) + 
  scale_y_continuous(breaks = seq(0,10000, by = 2000), limits = c(0,10000)) +
  ggtitle("Filtered: cells per sample") +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))

# Arrange graphs in grid
plots <- list(ncells1,ncells2)
layout <- rbind(c(1),c(2))
grid <- grid.arrange(grobs = plots, layout_matrix = layout)
```

```{r save_cells_per_sample,echo=FALSE,eval=FALSE}
# save
path <- paste0(out, "filtering_QC/cells_per_sample_before_and_after_filtering.pdf")
pdf(path, width = 8, height = 8)
grid <- grid.arrange(grobs = plots, layout_matrix = layout)
dev.off()

# cleanup
remove(ncells1,ncells2,plots,layout,grid,data)
```

## Density plots
```{r postfiltering_density, warning=FALSE}
# Visualize nCount_RNA
den1 <- ggplot(mouse.filtered@meta.data,
       aes(color = sample,
           x = nCount_RNA,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  scale_color_manual(values = sample_colors) +
  scale_fill_manual(values = sample_colors) +
  xlab("nCount_RNA") +
  ylab("Density") +
  theme(legend.position =  "none") +
  geom_vline(xintercept = nCount.min, lty = "dashed") +
  geom_vline(xintercept = nCount.max, lty = "dashed") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Visualize nFeature_RNA
den2 <- ggplot(mouse.filtered@meta.data,
       aes(color = sample,
           x = nFeature_RNA,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  scale_color_manual(values = sample_colors) +
  theme(legend.position =  "none") +
  scale_fill_manual(values = sample_colors) +
  xlab("nFeature_RNA") +
  ylab("Density") +
  geom_vline(xintercept = nFeature.min, lty = "dashed") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Visualize cell complexity
# Quality cells are usually above 0.85
den3 <- ggplot(mouse.filtered@meta.data,
       aes(color = sample,
           x = cell.complexity,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_color_manual(values = sample_colors) +
  theme(legend.position =  "none") +
  scale_fill_manual(values = sample_colors) +
  xlab("Cell Complexity (log10(nFeature/nCount))") +
  ylab("Density") +
  geom_vline(xintercept = complexity.cutoff, lty = "dashed") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Visualize percent.mt
den4 <- ggplot(mouse.filtered@meta.data,
       aes(color = sample,
           x = percent.mt,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_continuous(n.breaks = 4) +
  geom_vline(xintercept = mt.cutoff, lty = "dashed") +
  scale_color_manual(values = sample_colors) +
  theme(legend.position =  "none") +
  scale_fill_manual(values = sample_colors) +
  xlab("% Mitochondrial Genes") +
  ylab("Density") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Visualize percent.ribo.protein
den5 <- ggplot(mouse.filtered@meta.data,
       aes(color = sample,
           x = percent.ribo.protein,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  scale_color_manual(values = sample_colors) +
  theme(legend.position =  "none") +
  scale_fill_manual(values = sample_colors) +
  xlab("% Ribosomal Protein Genes") +
  ylab("Density") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Visualize percent.hb
den6 <- ggplot(mouse.filtered@meta.data,
       aes(color = sample,
           x = percent.hb,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  geom_vline(xintercept = hb.cutoff, lty = "dashed") +
  scale_color_manual(values = sample_colors) +
  theme(legend.position =  "none") +
  scale_fill_manual(values = sample_colors) +
  xlab("% Hemoglobin Genes") +
  ylab("Density") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Arrange graphs in grid
plots <- list(den1,den2,den3,den4,den5,den6)
layout <- rbind(c(1,4),c(2,5),c(3,6))
grid <- grid.arrange(grobs = plots, layout_matrix = layout)
```

```{r save_postfiltering_density, echo=FALSE, eval=FALSE, warning=FALSE}
# save
path <- paste0(out, "filtering_QC/density_curves_filtered.pdf")
pdf(path, width = 8, height = 8)
grid <- grid.arrange(grobs = plots, layout_matrix = layout)
dev.off()

# cleanup
remove(den1,den2,den3,den4,den5,den6,plots,layout,grid)
```

## Violin plots
```{r postfiltering_violins}
# nFeature, nCount, and cell.complexity violins
v3 <- VlnPlot(mouse.filtered,
              features = c("nFeature_RNA", "nCount_RNA","cell.complexity"),
              ncol = 3,
              group.by = 'sample',
              cols = sample_colors,
              pt.size = 0)
v3

#  percent violins
v4 <- VlnPlot(mouse.filtered,
              features = c("percent.mt","percent.ribo.protein","percent.hb"),
              ncol = 3,
              group.by = 'sample',
              cols = sample_colors,
              pt.size = 0)
v4
```

```{r save_postfiltering_violins, echo=FALSE, eval=FALSE}
# save
path <- paste0(out, "filtering_QC/violins_nFeature_nCount_complexity_filtered.pdf")
pdf(path, width = 14, height = 4)
v3
dev.off()

# save
path <- paste0(out, "filtering_QC/violins_percent_filtered.pdf")
pdf(path, width = 14, height = 4)
v4
dev.off()

# cleanup
remove(v3,v4)
```

## Scatter plots
```{r postfiltering_scatter}
s3 <- ggplot(
  mouse.filtered@meta.data,
  aes(x = nCount_RNA, y = nFeature_RNA, color = percent.mt)) + 
  geom_point() + 
  geom_smooth(method = "lm") +
	scale_x_log10() +   	
  scale_y_log10() + 
  theme_classic() +
  geom_vline(xintercept = nCount.min) + 
  geom_hline(yintercept = nFeature.min) + 
  facet_wrap(~sample) +
  scale_colour_gradient(low = "gray90", high = "black", limits =c(0,100))
s3

s4 <- FeatureScatter(mouse.filtered,
                     feature1 = "nCount_RNA",
                     feature2 = "percent.mt",
                     group.by = 'sample',
                     cols = sample_colors,
                     shuffle = TRUE)
s4
```

```{r save_postfiltering_scatter,echo=FALSE,eval=FALSE}
# save
path <- paste0(out, "filtering_QC/scatter_nFeature_vs_nCount_filtered.pdf")
pdf(path, width = 8, height = 8)
s3
dev.off()

path <- paste0(out, "filtering_QC/scatter_nCount_vs_percentMT_filtered.pdf")
pdf(path, width = 6, height = 6)
s4
dev.off()

# cleanup
remove(s3,s4)
```

## Box plot
```{r boxplot}
# Visualize the distribution of genes detected per cell via boxplot
b1 <- ggplot(mouse.filtered@meta.data,
       aes(x = sample, 
           y = log10(nFeature_RNA), 
           fill=sample)) + 
  geom_boxplot() + 
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust = 0.5, face="bold")) +
  ggtitle("Unique Genes / Cell / Sample") +
  scale_color_manual(values = sample_colors) +
  scale_fill_manual(values = sample_colors) +
  xlab("Sample")
b1
```

```{r save_boxplot,echo=FALSE,eval=FALSE}
# save
path <- paste0(out, "filtering_QC/boxplot_nFeature_per_sample.pdf")
pdf(path, width = 8, height = 6)
b1
dev.off()

# cleanup
remove(b1)
```

## Top transcripts
- Gm42418: Rn18s-rs5
```{r top_transcripts}
df <- data.frame(gene_name = rownames(mouse.filtered))
df$rsum <- rowSums(x = mouse.filtered, slot = "counts")
df <- df[order(df$rsum, decreasing = TRUE),]
rownames(df) <- 1:nrow(df)
head(df, 10)
```

```{r save_top_transcripts,echo=FALSE,eval=FALSE}
write.table(df, 
            paste0(out, "filtering_QC/top_transcripts.tsv"),
            quote = FALSE,
            row.names = FALSE)
remove(df)
```

# Unwanted variation
## Cell cycle
```{r cell_cycle, eval=FALSE}
mouse.filtered[["phase"]] <- cell_cycle_QC(
  obj = mouse.filtered,
  markersPath = "../../refs/cell_cycle_markers.tsv",
  species = "mouse",
  outDir = paste0(out, "unwanted_variation"))
```

## Mitochondria
[Evaluating effects of mitochondrial expression](https://hbctraining.github.io/scRNA-seq_online/lessons/06_SC_SCT_normalization.html)
```{r mitochondria_factor, eval=FALSE}
mouse.filtered[["mito.factor"]] <- mitochondria_QC(
  obj = mouse.filtered,
  outDir = paste0(out, "unwanted_variation"))
```

# Cluster
## SCTransform
- SCTransform method is a more accurate method of normalizing, estimating the 
variance of the raw filtered data, and identifying the most variable genes. \
- Variation in sequencing depth (total nCount_RNA per cell) is normalized using 
a regularized negative binomial model. \
- If there are other sources of uninteresting variation it can be included. \
- Keep the samples as separate objects as that is what is required for integration. \
```{r sctransform}
# split
mouse.filtered[["RNA"]] <- split(mouse.filtered[["RNA"]], 
                                 f = mouse.filtered$sample)

# transform
mouse.filtered <- SCTransform(mouse.filtered, verbose = FALSE)
```

## PCA
```{r run_pca, warning=FALSE, message=FALSE}
# run PCA on the merged object
mouse.filtered <- RunPCA(object = mouse.filtered, assay = "SCT")
```

```{r plot_pca}
# Reset idents and levels
DefaultAssay(mouse.filtered) <- "SCT"
Idents(mouse.filtered) <- "sample"
mouse.filtered$sample <- factor(mouse.filtered$sample, levels = sample_order)
mouse.filtered$sex <- factor(mouse.filtered$sex, levels = sex_order)
mouse.filtered$isoform <- factor(mouse.filtered$isoform, levels = isoform_order)
mouse.filtered$diet <- factor(mouse.filtered$diet, levels = diet_order)
mouse.filtered$group <- str_match(mouse.filtered$sample, "(E[34][CP][MF])[12]")[,2]
mouse.filtered$group2 <- str_match(mouse.filtered$sample, "(E[34][CP])[MF][12]")[,2]

# Plot PCA
pca1 <- DimPlot(mouse.filtered,
                reduction = "pca",
                group.by = "sample",
                shuffle = TRUE,
                raster = FALSE,
                cols = sample_colors)
pca1

pca2 <- DimPlot(mouse.filtered,
                reduction = "pca",
                shuffle = TRUE,
                raster = FALSE,
                group.by = "isoform",
                cols = isoform_colors)
pca2

pca3 <- DimPlot(mouse.filtered,
                reduction = "pca",
                group.by = "sex",
                shuffle = TRUE,
                raster = FALSE,
                cols = sex_colors)
pca3

pca4 <- DimPlot(mouse.filtered,
                reduction = "pca",
                group.by = "group",
                shuffle = TRUE,
                raster = FALSE,
                cols = group_colors)
pca4

pca5 <- DimPlot(mouse.filtered,
                reduction = "pca",
                group.by = "group2",
                shuffle = TRUE,
                raster = FALSE,
                cols = group2_colors)
pca5

e1 <- ElbowPlot(mouse.filtered) +
  geom_vline(xintercept = 15, linetype = "dashed", colour = "red")
e1
```

```{r save_pca, echo=FALSE, eval=FALSE, message=FALSE, warning=FALSE}
# save pca1
path <- paste0(out, "clustering_QC/PCA_sample")
pdf(paste0(path, ".pdf"), width = 6, height = 4)
pca1
dev.off()

# save pca2
path <- paste0(out, "clustering_QC/PCA_isoform")
pdf(paste0(path, ".pdf"), width = 6, height = 4)
pca2
dev.off()

# save pca3
path <- paste0(out, "clustering_QC/PCA_sex")
pdf(paste0(path, ".pdf"), width = 6, height = 4)
pca3
dev.off()

# save pca4
path <- paste0(out, "clustering_QC/PCA_group")
pdf(paste0(path, ".pdf"), width = 6, height = 4)
pca4
dev.off()

# save pca5
path <- paste0(out, "clustering_QC/PCA_group2")
pdf(paste0(path, ".pdf"), width = 6, height = 4)
pca5
dev.off()

# save e1
path <- paste0(out, "clustering_QC/PCA_eblow_plot.pdf")
pdf(path, width = 6, height = 4)
e1
dev.off()

# cleanup
remove(pca1,pca2,pca3,pca4,pca5,e1)
```

## UMAP
To overcome the extensive technical noise in the expression of any single gene 
for scRNA-seq data, Seurat assigns cells to clusters based on their PCA scores 
derived from the expression of the integrated most variable genes, with each PC 
essentially representing a “metagene” that combines information across a 
correlated gene set. Determining how many PCs to include in the clustering step 
is therefore important to ensure that we are capturing the majority of the 
variation, or cell types, present in our dataset.
```{r run_umap}
# re-join layers
mouse.filtered[["RNA"]] <- JoinLayers(mouse.filtered[["RNA"]])

# run UMAP
mouse.filtered <- RunUMAP(mouse.filtered,
                          dims = 1:15,
                          reduction = "pca",
                          n.components = 3) # set to 3 to use with VR

# plot UMAP
DimPlot(mouse.filtered,
        cols = sample_colors,
        shuffle = TRUE)
```

## Clusters
```{r find_neighbors, message=FALSE, warning=FALSE}
# Determine the K-nearest neighbor graph
mouse.unannotated <- FindNeighbors(object = mouse.filtered,
                                   assay = "SCT", # set as default after SCTransform
                                   reduction = "pca",
                                   dims = 1:15)

# Determine the clusters for various resolutions
mouse.unannotated <- FindClusters(object = mouse.unannotated,
                                  algorithm = 1, # 1 = Louvain
                                  resolution = seq(0.1, 0.8, by=0.1))
mouse.unannotated$seurat_clusters <- mouse.unannotated$SCT_snn_res.0.2
```

## Explore resolutions
```{r explore_resolutions}
# 0.1
DimPlot(mouse.unannotated,
        group.by = "SCT_snn_res.0.1",
        label = TRUE)

# 0.2
DimPlot(mouse.unannotated,
        group.by = "SCT_snn_res.0.2",
        label = TRUE)
DimPlot(mouse.unannotated,
        group.by = "SCT_snn_res.0.2",
        dims = c(2,3),
        label = TRUE)

# 0.3
DimPlot(mouse.unannotated,
        group.by = "SCT_snn_res.0.3",
        label = TRUE)

# 0.4
DimPlot(mouse.unannotated,
        group.by = "SCT_snn_res.0.4",
        label = TRUE)

# 0.5
DimPlot(mouse.unannotated,
        group.by = "SCT_snn_res.0.5",
        label = TRUE)
```

```{r save_final_object,echo=FALSE,eval=TRUE}
# Save seurat object
saveRDS(mouse.unannotated, "../../rObjects/unannotated_seurat_obj.rds")
```
