---
title: "PLX5622 Diet Mice scRNAseq"
subtitle: "Cluster Annotation"
author: "Kennedi Todd"
date: "02/12/2025"
output:
  pdf_document:
    toc: yes
  html_document:
    theme: cerulean
    toc: yes
    toc_float: yes
editor_options:
  chunk_output_type: inline
---

# Setup
## Working directory
```{r working_directory}
knitr::opts_knit$set(root.dir = ".")
```

## Libraries
```{r libraries, message=FALSE, warning=FALSE}
library(dplyr)        # ungroup()
library(gridExtra)    # grid.arrange()
library(gtools)       # smartbind()
library(parallel)     # detectCores()
library(plotly)       # plot_ly()
library(Seurat)       # Idents()
library(ShinyCell)    # createConfig()
library(tidyr)        # %>%
```

## Variables and functions
```{r set_variables_and_thresholds}
# variables
out <- "../../results/"

sample_order <- c("E3CF1","E3CF2","E3CM1","E3CM2","E3PF1","E3PF2","E3PM1","E3PM2",
                  "E4CF1","E4CF2","E4CM1","E4CM2","E4PF1","E4PF2","E4PM1","E4PM2")
sample_colors <- c("#9e0909","#f75959","#f0b402","#f5d67a",
                   "#fbff0a","#fafaa7","#1d8c0e","#63e851",
                   "#0271f0","#9cc3f0","#6c39f7","#c589fa",
                   "#ed2af7","#f5c1f7","#755410","#b5aa82")

group_order <- c("E3CM","E3PM","E4CM","E4PM",
                 "E3CF","E3PF","E4CF","E4PF")
group_colors <- c("#B89B74","#725E47","#B15928","#F7590B",
                  "#A6CEE3","#34B1E8","#ED8F47","#F4CC25")

group2_order <- c("E3C","E3P","E4C","E4P")
group2_colors <- c("#f75959","#fbff0a","#63e851","#0271f0")

isoform_order <- c("E4","E3")
isoform_colors <- c("darkgray","cornflowerblue")

sex_order <- c("Male","Female")
sex_colors <- c("green","purple")

diet_order <- c("control","PLX5622")
diet_color <- c("coral","cyan")
```

## Read data
```{r read_object}
# read obj
mouse.unannotated <- readRDS("../../rObjects/unannotated_seurat_obj.rds")

# set params
Idents(mouse.unannotated) <- "seurat_clusters"
DefaultAssay(mouse.unannotated) <- "RNA"
mouse.unannotated <- NormalizeData(mouse.unannotated)
mouse.unannotated$group <- factor(mouse.unannotated$group, levels = group_order)
mouse.unannotated$group2 <- factor(mouse.unannotated$group2, levels = group2_order)

# preview
mouse.unannotated
```

# Unannotated QC
## UMAP
```{r unannotated_umap}
cluster_colors <- c("red","blue","cyan","black","chocolate4","gray","steelblue",
                    "orange","green","magenta","yellow","purple","forestgreen",
                    "tan","salmon","plum1")
u1 <- DimPlot(object = mouse.unannotated,
              reduction = "umap",
              shuffle = TRUE,
              repel = TRUE,
              cols = cluster_colors,
              label = TRUE)
u1

u2 <- DimPlot(object = mouse.unannotated,
              reduction = "umap",
              dims = c(2,3),
              shuffle = TRUE,
              repel = TRUE,
              cols = cluster_colors,
              label = TRUE)
u2
```

```{r save_unannotated_umap, echo=FALSE, eval=FALSE}
pdf(paste0(out, "UMAP/unannotated_clusters_dim1&2.pdf"))
u1
dev.off()

pdf(paste0(out, "UMAP/unannotated_clusters_dim2&3.pdf"))
u2
dev.off()

remove(u1,u2)
```

## Feature plots
```{r feature_plots_unannotated, message=FALSE, warning=FALSE}
# UMAP percent.mt
FeaturePlot(mouse.unannotated,
            reduction = "umap", 
            features = "percent.mt") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))

# UMAP percent.ribo
FeaturePlot(mouse.unannotated,
            reduction = "umap", 
            features = "percent.ribo.protein") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))

# UMAP percent.hb
FeaturePlot(mouse.unannotated,
            reduction = "umap", 
            features = "percent.hb") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))

# UMAP nCount
FeaturePlot(mouse.unannotated,
            reduction = "umap", 
            features = "nCount_RNA") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))

# UMAP nFeature
FeaturePlot(mouse.unannotated,
            reduction = "umap", 
            features = "nFeature_RNA") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))

# UMAP cell.complexity
FeaturePlot(mouse.unannotated,
            reduction = "umap", 
            features = "cell.complexity") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))

# UMAP Ttr expression
FeaturePlot(mouse.unannotated,
            reduction = "umap", 
            features = "Ttr") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
```

## Violins
```{r qc_violins}
VlnPlot(mouse.unannotated,
        features = "nCount_RNA",
        cols = cluster_colors,
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        cols = cluster_colors,
        features = "nFeature_RNA",
        split.by = "seurat_clusters")
```

## Cells per cluster
```{r cells_per_cluster}
# Cells per sample per cluster
sample_ncells <- FetchData(mouse.unannotated, 
                     vars = c("ident", "sample")) %>%
  dplyr::count(ident,sample) %>%
  tidyr::spread(ident, n)
sample_ncells

# Cells per isoform per cluster
isoform_ncells <- FetchData(mouse.unannotated, 
                     vars = c("ident", "isoform")) %>%
  dplyr::count(ident,isoform) %>%
  tidyr::spread(ident, n)
isoform_ncells

# Cells per sex per cluster
sex_ncells <- FetchData(mouse.unannotated, 
                     vars = c("ident", "sex")) %>%
  dplyr::count(ident,sex) %>%
  tidyr::spread(ident, n)
sex_ncells

# Cells per diet per cluster
diet_ncells <- FetchData(mouse.unannotated, 
                     vars = c("ident", "diet")) %>%
  dplyr::count(ident, diet) %>%
  tidyr::spread(ident, n)
diet_ncells
```

## Gene histogram
```{r histogram}
# User params
goi <- "Malat1"
threshold <- 1

# Subset data
log2.threshold <- log2(threshold + 0.01)
counts.df <- FetchData(mouse.unannotated, vars = goi)
colnames(counts.df) <- "counts"
log2.counts.df <- log2(counts.df + 0.01)

# Histogram
title <- paste0(goi, "\nnCount_RNA > ", threshold)
hist1 <- ggplot(counts.df, aes(x = counts)) + 
  geom_histogram(bins = 100, fill = "gray", color = "black") + 
  labs(title = title, x=NULL, y=NULL) +
  xlab(paste0(goi, " nCount_RNA")) + ylab("# of Samples") + theme_bw() +
  geom_vline(xintercept = threshold, col = "blue") +
  annotate("rect", 
           xmin = -Inf,
           xmax = threshold, 
           ymin = 0, 
           ymax=Inf, 
           alpha=0.2, 
           fill="chocolate4") +
  annotate("rect", 
           xmin = threshold,
           xmax = Inf, 
           ymin = 0, 
           ymax=Inf, 
           alpha=0.2, 
           fill="deepskyblue")

# Histogram log transformed
hist2 <- ggplot(log2.counts.df, aes(x = counts)) + 
  geom_histogram(bins = 100, fill = "gray", color = "black") + 
  labs(title = title, x=NULL, y=NULL) +
  xlab(paste0("Log2(",goi, " nCount_RNA)")) + ylab("# of Samples") + theme_bw() +
  geom_vline(xintercept = log2.threshold, col = "blue") +
  annotate("rect", 
           xmin = -Inf,
           xmax = log2.threshold, 
           ymin = 0, 
           ymax=Inf, 
           alpha=0.2, 
           fill="chocolate4") +
  annotate("rect", 
           xmin = log2.threshold,
           xmax = Inf, 
           ymin = 0, 
           ymax=Inf, 
           alpha=0.2, 
           fill="deepskyblue")

# plot
plots1 <- list(hist1,hist2)
layout1 <- rbind(c(1),c(2))
grid1 <- grid.arrange(grobs = plots1, layout_matrix = layout1)
```

## Percent gene
```{r percent_gene}
# user define variable
goi <- "Malat1"

# Extract counts data
DefaultAssay(mouse.unannotated) <- "RNA"
Idents(mouse.unannotated) <- "seurat_clusters"
geneData <- FetchData(mouse.unannotated,
                      vars = goi,
                      slot = "counts")
geneData <- geneData > 1
table(geneData)
mouse.unannotated$Expression <- geneData

FetchData(mouse.unannotated,
          vars = c("ident", "Expression")) %>%
  dplyr::count(ident, Expression) %>%
  tidyr::spread(ident, n)

# Plot
mouse.unannotated@meta.data %>%
  group_by(seurat_clusters, Expression) %>%
  dplyr::count() %>%
  group_by(seurat_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=seurat_clusters,y=percent, fill=Expression)) +
  geom_col() +
  ggtitle(paste0("Percentage of cells with > 1 counts for ", goi)) +
  theme(axis.text.x = element_text(angle = 90))
```

# Potential Markers
## B-cells / Plasma cells
- Cd19: expressed in B-cells and follicular dendritic cells \
- Fcrla: a B-cell specific protein in mice \
- Cd79a and Cd79b: together form BCR complex \
- Sdc1: Plasma cells \
```{r bcells}
Idents(mouse.unannotated) <- "seurat_clusters"
VlnPlot(mouse.unannotated,
        features = "Cd19",
        cols = cluster_colors,
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        features = "Fcrla",
        cols = cluster_colors,
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        features = "Cd79a",
        cols = cluster_colors,
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        features = "Cd79b",
        cols = cluster_colors,
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        features = "Sdc1",
        cols = cluster_colors,
        split.by = "seurat_clusters")
```

## T-cells
- Trac/Cd3d/Cd3e/Cd3g are components of the TCR \
```{r tcells}
VlnPlot(mouse.unannotated,
        features = "Trac",
        cols = cluster_colors,
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        features = "Cd3e",
        cols = cluster_colors,
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        features = "Cd8a",
        cols = cluster_colors,
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        features = "Cd4",
        cols = cluster_colors,
        split.by = "seurat_clusters")
```

## Endothelial cells
```{r endothelial}
VlnPlot(mouse.unannotated,
        features = "Ly6c1",
        cols = cluster_colors,
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        features = "Flt1",
        cols = cluster_colors,
        split.by = "seurat_clusters")
# LECs
VlnPlot(mouse.unannotated,
        features = "Flt4",
        cols = cluster_colors,
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        features = "Prox1",
        cols = cluster_colors,
        split.by = "seurat_clusters")
# BECs
VlnPlot(mouse.unannotated,
        features = "Cldn5",
        cols = cluster_colors,
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        features = "Vwf",
        cols = cluster_colors,
        split.by = "seurat_clusters")
```

## Fibroblasts
```{r fibroblasts}
VlnPlot(mouse.unannotated,
        features = "Col1a1",
        cols = cluster_colors,
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        features = "Col1a2",
        cols = cluster_colors,
        split.by = "seurat_clusters")
```

## Mast cells
```{r mast_cells}
VlnPlot(mouse.unannotated,
        features = "Fcer1a",
        cols = cluster_colors,
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        features = "Kit", # aka Cd117
        cols = cluster_colors,
        split.by = "seurat_clusters")
```

## Macrophage
```{r macrophage}
VlnPlot(mouse.unannotated,
        features = "C1qa",
        cols = cluster_colors,
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        features = "C1qb",
        cols = cluster_colors,
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        features = "Mki67",
        cols = cluster_colors,
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        features = "Il7r",
        cols = cluster_colors,
        split.by = "seurat_clusters")
```

## Monocytes/DCs
```{r monocytes_DCs}
VlnPlot(mouse.unannotated,
        features = "Itgax", # Cd11c
        cols = cluster_colors,
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        features = "Cd209a",
        cols = cluster_colors,
        split.by = "seurat_clusters")
```

## Neutrophils
```{r neutrophils}
VlnPlot(mouse.unannotated,
        features = "Ly6g",
        cols = cluster_colors,
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        features = "Retnlg",
        cols = cluster_colors,
        split.by = "seurat_clusters")
```

## NK cells
```{r nk_cells}
VlnPlot(mouse.unannotated,
        features = "Nkg7",
        cols = cluster_colors,
        split.by = "seurat_clusters")
```

## Pericytes and SMCs
```{r pericytes_smcs}
VlnPlot(mouse.unannotated,
        features = "Acta2",
        cols = cluster_colors,
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        features = "Myl9",
        cols = cluster_colors,
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        features = "Rgs5",
        cols = cluster_colors,
        split.by = "seurat_clusters")
```

## Schwann cells
```{r schwann_cells}
VlnPlot(mouse.unannotated,
        features = "Cdh19",
        cols = cluster_colors,
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        features = "Mpz",
        cols = cluster_colors,
        split.by = "seurat_clusters")
VlnPlot(mouse.unannotated,
        features = "Plp1",
        cols = cluster_colors,
        split.by = "seurat_clusters")
```

## Sandro's markers
```{r sandros_markers}
Idents(mouse.unannotated) <- "seurat_clusters"
goi1 <- c("Cd3e","Trbc1","Cd4","Cd8a","Foxp3","Tbx21","Gata3","Thy1",
         "Cd19","Ms4a1","Cd27","Ighg1","Ptprc","Ly6g","Itgam","Aif1","Klrb1")
goi2 <- c("Adgre1","Ms4a3","Ly6c2","Mrc1","Lyz2","Cd74","Cd83","Cd14","H2-Aa",
          "H2-Ab1","Sirpa","Xcr1","Siglech","Itgax","Kit","Mcpt4")
goi3 <- c("Pdgfra","Col1a1","Lum","Pdgfrb","Rgs5","Cspg4","Acta2","Tagln",
          "Pecam1","Cd34","Plvap","Stmn2","Slc38a5","Vwf","Mfsd2a","Cldn5")
goi4 <- c("Prox1","Flt4","Pdpn","Lyve1","Sox10","Mbp","Fgf13","Kcnab2","Tubb3",
          "Slc17a6","Shank2","Erbb4","Park7","Kif5b","Slc4a1","Hmbs")
goi5 <- c("Pecam1","Flt4","Itgam","Mrc1","Cd3e","Gata3","Cd19","Ly6g","Pdgfrb",
          "Kit","Col1a1","Pmp22","Hmbs")

v1 <- VlnPlot(mouse.unannotated,
              features = goi1,
              cols = cluster_colors,
              split.by = "seurat_clusters",
              flip = TRUE,
              stack = TRUE)
v1

v2 <- VlnPlot(mouse.unannotated,
              features = goi2,
              cols = cluster_colors,
              split.by = "seurat_clusters",
              flip = TRUE,
              stack = TRUE)
v2

v3 <- VlnPlot(mouse.unannotated,
              features = goi3,
              cols = cluster_colors,
              split.by = "seurat_clusters",
              flip = TRUE,
              stack = TRUE)
v3

v4 <- VlnPlot(mouse.unannotated,
              features = goi4,
              cols = cluster_colors,
              split.by = "seurat_clusters",
              flip = TRUE,
              stack = TRUE)
v4

v5 <- VlnPlot(mouse.unannotated,
              features = goi5,
              cols = cluster_colors,
              split.by = "seurat_clusters",
              flip = TRUE,
              stack = TRUE)
v5
```

## Find markers
```{r find_markers_unannotated}
# Find markers for each cluster
markers <- SeuratWrappers::RunPrestoAll(
  object = mouse.unannotated,
  assay = "RNA",
  slot = "counts",
  only.pos = TRUE
)

# copy
all.markers <- markers

# more stringent filtering
all.markers <- all.markers[all.markers$p_val_adj < 0.01,]

# compare 
table(all.markers$cluster)

# subset
cluster0 <- all.markers[all.markers$cluster == 0,]
cluster1 <- all.markers[all.markers$cluster == 1,]
cluster2 <- all.markers[all.markers$cluster == 2,]
cluster3 <- all.markers[all.markers$cluster == 3,]
cluster4 <- all.markers[all.markers$cluster == 4,]
cluster5 <- all.markers[all.markers$cluster == 5,]
cluster6 <- all.markers[all.markers$cluster == 6,]
cluster7 <- all.markers[all.markers$cluster == 7,]
cluster8 <- all.markers[all.markers$cluster == 8,]
cluster9 <- all.markers[all.markers$cluster == 9,]
cluster10 <- all.markers[all.markers$cluster == 10,]
cluster11 <- all.markers[all.markers$cluster == 11,]
cluster12 <- all.markers[all.markers$cluster == 12,]
cluster13 <- all.markers[all.markers$cluster == 13,]
cluster14 <- all.markers[all.markers$cluster == 14,]
cluster15 <- all.markers[all.markers$cluster == 15,]

# save
write.table(all.markers,
            paste0(out, "markers/unannotated_auto_find_markers_adjpval_0.01.tsv"), 
            sep = "\t", quote = FALSE, row.names = FALSE)
```

# Cluster Annotations
## Cluster 0: BECs
- Flt1, Ptprb, Ly6c1
```{r cluster0}
VlnPlot(mouse.unannotated,
        features = cluster0$gene[1:20],
        cols = cluster_colors,
        stack = TRUE,
        flip = TRUE,
        split.by = "seurat_clusters")
```

## Cluster 1: Macrophages
- Csf1r, Mrc1, C1qa, C1qb, C1qc
```{r cluster1}
VlnPlot(mouse.unannotated,
        features = cluster1$gene[1:20],
        cols = cluster_colors,
        stack = TRUE,
        flip = TRUE,
        split.by = "seurat_clusters")
```

## Cluster 2: B cells
- Fcrla, Ms4a1, Cd79a
```{r cluster2}
VlnPlot(mouse.unannotated,
        features = cluster2$gene[1:20],
        cols = cluster_colors,
        stack = TRUE,
        flip = TRUE,
        split.by = "seurat_clusters")
```

## Cluster 3: T and NK cells
- Ms4a4b, Cd3g, Cd3d, Cd3e
```{r cluster3}
VlnPlot(mouse.unannotated,
        features = cluster3$gene[1:20],
        cols = cluster_colors,
        stack = TRUE,
        flip = TRUE,
        split.by = "seurat_clusters")
```

## Cluster 4: ILCs
- Gata3, Il7r
```{r cluster4}
VlnPlot(mouse.unannotated,
        features = cluster4$gene[1:20],
        cols = cluster_colors,
        stack = TRUE,
        flip = TRUE,
        split.by = "seurat_clusters")
```

## Cluster 5: Fibroblasts
- Col1a1, Col1a2
```{r cluster5, echo=FALSE, eval=FALSE}
VlnPlot(mouse.unannotated,
        features = cluster5$gene[1:20],
        cols = cluster_colors,
        stack = TRUE,
        flip = TRUE,
        split.by = "seurat_clusters")
```

## Cluster 6: Macrophages
- C1qb, C1qa
- coexpressed with cluster 1
```{r cluster6, message=FALSE}
VlnPlot(mouse.unannotated,
        features = cluster6$gene[1:20],
        cols = cluster_colors,
        stack = TRUE,
        flip = TRUE,
        split.by = "seurat_clusters")
```

## Cluster 7: BECs
- Vwf, Flt1
```{r cluster7}
VlnPlot(mouse.unannotated,
        features = cluster7$gene[1:20],
        cols = cluster_colors,
        stack = TRUE,
        flip = TRUE,
        split.by = "seurat_clusters")
```

## Cluster 8: Dendritic cells
- Ccr2, Fgr
```{r cluster8}
VlnPlot(mouse.unannotated,
        features = cluster8$gene[1:20],
        cols = cluster_colors,
        stack = TRUE,
        flip = TRUE,
        split.by = "seurat_clusters")
```

## Cluster 9: Neutrophils
- Retnlg, S100a9, S100a8, Hp
```{r cluster9}
VlnPlot(mouse.unannotated,
        features = cluster9$gene[1:20],
        cols = cluster_colors,
        stack = TRUE,
        flip = TRUE,
        split.by = "seurat_clusters")
```

## Cluster 10: Pericytes & SMCs
- Notch3, Rgs5, Myh11
```{r cluster10}
VlnPlot(mouse.unannotated,
        features = cluster10$gene[1:20],
        cols = cluster_colors,
        stack = TRUE,
        flip = TRUE,
        split.by = "seurat_clusters")
```

## Cluster 11: BECs
```{r cluster11}
VlnPlot(mouse.unannotated,
        features = cluster11$gene[1:20],
        cols = cluster_colors,
        stack = TRUE,
        flip = TRUE,
        split.by = "seurat_clusters")
```

## Cluster 12: Myeloid precursors
```{r cluster12}
VlnPlot(mouse.unannotated,
        features = cluster12$gene[1:20],
        cols = cluster_colors,
        stack = TRUE,
        flip = TRUE,
        split.by = "seurat_clusters")
```

## Cluster 13: Mast cells
- Kit, Fcer1a
```{r cluster13}
VlnPlot(mouse.unannotated,
        features = cluster13$gene[1:20],
        cols = cluster_colors,
        stack = TRUE,
        flip = TRUE,
        split.by = "seurat_clusters")
```

## Cluster 14: LECs
- Prox1, Flt4
```{r cluster14}
VlnPlot(mouse.unannotated,
        features = cluster14$gene[1:20],
        cols = cluster_colors,
        stack = TRUE,
        flip = TRUE,
        split.by = "seurat_clusters")
```

## Cluster 15: Schwann cells
- Plp1, Cdh19
```{r cluster15}
VlnPlot(mouse.unannotated,
        features = cluster15$gene[1:20],
        cols = cluster_colors,
        stack = TRUE,
        flip = TRUE,
        split.by = "seurat_clusters")
```

# Assign identities
```{r assign_identities, eval=FALSE, echo=FALSE}
# rename all identities
mouse.annotated <- RenameIdents(object = mouse.unannotated,
                                "0" = "BECs",                # Flt1, Ptprb, Ly6c1
                                "1" = "Macrophages",         # Csf1r, Mrc1, C1qa, C1qb, C1qc
                                "2" = "B cells",             # Fcrla, Ms4a1, Cd79a
                                "3" = "T and NK cells",      # Ms4a4b, Cd3g, Cd3d, Cd3e
                                "4" = "ILCs",                # Gata3, Il7r
                                "5" = "Fibroblasts",         # Col1a1, Col1a2
                                "6" = "Macrophages",         # C1qa, C1qb
                                "7" = "BECs",                # Vwf, Flt1
                                "8" = "Dendritic cells",     # Ccr2, Fgr
                                "9" = "Neutrophils",         # Retnlg, S100a9, S100a8, Hp
                                "10" = "Pericytes and SMCs", # Notch3, Rgs5, Myh11
                                "11" = "BECs",               # Cldn5, Flt1
                                "12" = "Myeloid precursors", # Stmn1, Mki67
                                "13" = "Mast cells",         # Kit, Fcer1a
                                "14" = "LECs",               # Prox1, Flt4
                                "15" = "Schwann cells")      # Plp1, Cdh19
mouse.annotated$annotated_clusters <- Idents(mouse.annotated)

mouse.annotated$annotated_clusters <- factor(mouse.annotated$annotated_clusters,
                                             levels = c("Pericytes and SMCs",
                                                        "BECs",
                                                        "LECs",
                                                        "B cells",
                                                        "T and NK cells",
                                                        "ILCs",
                                                        "Dendritic cells",
                                                        "Neutrophils",
                                                        "Macrophages",
                                                        "Myeloid precursors",
                                                        "Mast cells",
                                                        "Fibroblasts",
                                                        "Schwann cells"))

# set ident
Idents(mouse.annotated) <- "annotated_clusters"
```

## Annotated UMAP
```{r annotated_umap, eval=FALSE, echo=FALSE}
# set colors
cluster_colors <- c("#B5B9BA", # Pericytes and SMCs
                    "#3385BB", # BECs
                    "#40BBFF", # LECs
                    "#A5D5A9", # B cells
                    "#1C7E24", # T and NK cells
                    "#F57C7C", # ILCs
                    "#E42622", # Dendritic cells
                    "#FBB268", # Neutrophils
                    "#FE8D19", # Macrophages
                    "#DE9E83", # Myeloid precursors
                    "#A6CEE3", # Mast cells
                    "#9D7BBA", # Fibroblasts
                    "#977899") # Schwann cells

# umap
umap1 <- DimPlot(object = mouse.annotated, 
        reduction = "umap",
        repel = TRUE,
        group.by = "annotated_clusters",
        cols = cluster_colors)
umap1
umap2 <- DimPlot(object = mouse.annotated, 
        reduction = "umap",
        repel = TRUE,
        dims = c(2,3),
        group.by = "annotated_clusters",
        cols = cluster_colors)
umap2
```

```{r save_annotated_object, echo=FALSE, eval=FALSE}
# save object
saveRDS(mouse.annotated, "../../rObjects/annotated_seurat_obj.rds")
```

# Downsample
```{r downsample}
# check cells per sample
table(mouse.annotated$sample)

# downsample the number of cells per identity class
Idents(mouse.annotated) <- "sample"
mouse.annotated <- subset(x = mouse.annotated, downsample = 2000)

# check
table(mouse.annotated$sample)
```

```{r annotated_umap_downsampled, eval=FALSE, echo=FALSE}
# umap
umap1 <- DimPlot(object = mouse.annotated, 
        reduction = "umap",
        repel = TRUE,
        group.by = "annotated_clusters",
        cols = cluster_colors)
umap1
umap2 <- DimPlot(object = mouse.annotated, 
        reduction = "umap",
        repel = TRUE,
        dims = c(2,3),
        group.by = "annotated_clusters",
        cols = cluster_colors)
umap2
```

```{r save_annotated_downsampled_umap}
pdf(paste0(out, "UMAP/annotated_clusters_dim1&2.pdf"), width = 6, height = 4)
umap1
dev.off()

pdf(paste0(out, "UMAP/annotated_clusters_dim2&3.pdf"), width = 6, height = 4)
umap2
dev.off()

remove(umap1,umap2)
```

```{r save_downsampled_obj, eval=FALSE, echo=FALSE}
saveRDS(mouse.annotated, "../../rObjects/annotated_downsampled_seurat_obj.rds")
```

# Markers
## Automatically detect markers
```{r run_presto_annotated}
# auto find markers
Idents(mouse.annotated) <- "annotated_clusters"
all.markers <- SeuratWrappers::RunPrestoAll(
  object = mouse.annotated,
  assay = "RNA",
  slot = "counts",
  only.pos = TRUE
)

# filter based on p_val_adj
all.markers <- all.markers[all.markers$p_val_adj < 0.01,]

# subset table to top 2 and top 20
top2 <- Reduce(rbind,
               by(all.markers,
                  all.markers["cluster"],
                  head,
                  n = 2))
top20 <- Reduce(rbind,
               by(all.markers,
                  all.markers["cluster"],
                  head,
                  n = 20))

# save
write.table(all.markers, 
            paste0(out, "markers/annotated_auto_find_markers_adjpval_0.01.tsv"),
            quote = FALSE,
            row.names = FALSE)
```

## Violin markers
```{r annotated_cluster_markers}
v1 <- VlnPlot(mouse.annotated,
              features = top2$gene,
              split.by = "annotated_clusters",
              flip = TRUE,
              stack = TRUE,
              cols = cluster_colors)
v1
```

```{r save_annotated_cluster_markers, eval=FALSE, echo=FALSE}
pdf(paste0(out, "markers/auto_markers_violin_annotated.pdf"), width = 8, height = 8)
v1
dev.off()

remove(v1)
```

## Pseudobulk PCA
- Each sample will have data points equal to the number of annotated_clusters
```{r}
# Step 1: Pseudo-bulk the counts based on sample and cell type
pseudo <- AggregateExpression(
    object = mouse.annotated, 
    assays = "RNA", 
    features = rownames(mouse.annotated),
    return.seurat = TRUE, 
    group.by = c("sample", "annotated_clusters")
)

# Step 2: Normalize the data
pseudo <- NormalizeData(pseudo,
                        normalization.method = "LogNormalize",
                        scale.factor = 10000)

# Step 3: Find variable features
pseudo <- FindVariableFeatures(pseudo,
                               selection.method = "vst", 
                               nfeatures = 2000)

# Step 4: Scale the data
pseudo <- ScaleData(pseudo, 
                    features = rownames(pseudo))

# Step 5: Run PCA
pseudo <- RunPCA(pseudo, 
                 features = VariableFeatures(pseudo),
                 npcs = 10)

# Step 6: Visualize PCA
pca_colors <- c("firebrick1","cyan","gold","blue","black","forestgreen",
                    "darkorchid1","green","gray","deeppink","chocolate4",
                    "steelblue","pink","orange")
pca <- DimPlot(pseudo, 
        reduction = "pca",
        group.by = "annotated_clusters",
        cols = pca_colors,
        pt.size = 3)
pca
```

```{r}
pdf(paste0(out, "clustering_QC/pseudobulk_pca.pdf"), width = 6, height = 4)
pca
dev.off()
```
