---
title: "PLX5622 Diet Mice scRNAseq"
subtitle: "QC and DE"
author: "Kennedi Todd"
date: "02/12/2025"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

# Setup
## Working directory
```{r working_directory}
knitr::opts_knit$set(root.dir = ".")
```

## Libraries
```{r libraries, message=FALSE, warning=FALSE}
library(dplyr)        # ungroup()
library(Seurat)       # DimPlot()
library(ShinyCell)    # createConfig()
library(tidyr)        # %>%
```

## Variables and functions
```{r set_variables_and_functions}
# variables
out <- "../../results/"

sample_order <- c("E3CF1","E3CF2","E3CM1","E3CM2","E3PF1","E3PF2","E3PM1","E3PM2",
                  "E4CF1","E4CF2","E4CM1","E4CM2","E4PF1","E4PF2","E4PM1","E4PM2")
sample_colors <- c("#9e0909","#f75959","#f0b402","#f5d67a",
                   "#fbff0a","#fafaa7","#1d8c0e","#63e851",
                   "#0271f0","#9cc3f0","#6c39f7","#c589fa",
                   "#ed2af7","#f5c1f7","#755410","#b5aa82")

group_order <- c("E3CM","E3PM","E4CM","E4PM",
                 "E3CF","E3PF","E4CF","E4PF")
group_colors <- c("#B89B74","#725E47","#B15928","#F7590B",
                  "#A6CEE3","#34B1E8","#ED8F47","#F4CC25")

group2_order <- c("E3C","E3P","E4C","E4P")
group2_colors <- c("#f75959","#fbff0a","#63e851","#0271f0")

isoform_order <- c("E4","E3")
isoform_colors <- c("darkgray","cornflowerblue")

sex_order <- c("Male","Female")
sex_colors <- c("green","purple")

diet_order <- c("control","PLX5622")
diet_color <- c("coral","cyan")

# single cell functions
files <- list.files("../../functions", full.names = TRUE)
invisible(lapply(files, source))
```

## Load data
```{r read_annotated_obj}
mouse.annotated <- readRDS("../../rObjects/annotated_downsampled_seurat_obj.rds")
```

# QC
## Annotated UMAP
```{r annotated_umap}
# set colors
cluster_colors <- c("#B5B9BA", # Pericytes and SMCs
                    "#3385BB", # BECs
                    "#40BBFF", # LECs
                    "#A5D5A9", # B cells
                    "#1C7E24", # T and NK cells
                    "#F57C7C", # ILCs
                    "#E42622", # Dendritic cells
                    "#FBB268", # Neutrophils
                    "#FE8D19", # Macrophages
                    "#DE9E83", # Myeloid precursors
                    "#A6CEE3", # Mast cells
                    "#9D7BBA", # Fibroblasts
                    "#977899") # Schwann cells

# umap
Idents(mouse.annotated) <- "annotated_clusters"
umap1 <- DimPlot(object = mouse.annotated,
                 reduction = "umap",
                 repel = TRUE,
                 group.by = "annotated_clusters",
                 cols = cluster_colors)
umap1

umap2 <- DimPlot(object = mouse.annotated,
                 reduction = "umap",
                 repel = TRUE,
                 dims = c(2,3),
                 group.by = "annotated_clusters",
                 cols = cluster_colors)
umap2
```

## Cluster markers
```{r cluster_markers_violin}
goi <- c("Pdgfrb","Acta2","Vwf","Cldn5","Pecam1","Flt4","Prox1","Lyve1","Ptprc",
         "Cd19","Ms4a1","Ighd","Igha","Sdc1","Cd3e","Trbc2","Il7r","Nkg7","Klrb1b",
         "Klrb1c","Gata3","Rora","Itgax","H2-Eb1","Ccr2","Ly6c2","Lyz2","Ly6g","Itgam",
         "Mrc1","Csf1r","Cd38","Mki67","Mcpt4","Ms4a2","Col1a2","Plp1")

v <- VlnPlot(mouse.annotated,
             features = goi,
             split.by = "annotated_clusters",
             flip = TRUE,
             stack = TRUE,
             cols = cluster_colors)
v
```

```{r save_ann_markers, echo=FALSE}
pdf(paste0(out, "markers/sandro_markers_violin_annotated.pdf"), width = 12, height = 12)
v
dev.off()
remove(v)
```

## UMAP meta
```{r umap_meta}
# Apoe isoform
umap1 <- DimPlot(object = mouse.annotated,
                 reduction = "umap",
                 group.by = "isoform",
                 repel = TRUE,
                 shuffle = TRUE,
                 cols = isoform_colors)
umap1

# sex
umap2 <- DimPlot(object = mouse.annotated, 
                 reduction = "umap",
                 group.by = "sex",
                 shuffle = TRUE,
                 repel = TRUE,
                 cols = sex_colors)
umap2

# sample
umap3 <- DimPlot(object = mouse.annotated, 
                 reduction = "umap",
                 group.by = "sample",
                 shuffle = TRUE,
                 repel = TRUE,
                 cols = sample_colors)
umap3

# phase
umap4 <- DimPlot(object = mouse.annotated, 
                 reduction = "umap",
                 shuffle = TRUE,
                 group.by = "phase",
                 repel = TRUE)
umap4

# mito.factor
umap5 <- DimPlot(object = mouse.annotated,
                 reduction = "umap",
                 shuffle = TRUE,
                 group.by = "mito.factor",
                 repel = TRUE)
umap5

# group
umap6 <- DimPlot(object = mouse.annotated,
                 reduction = "umap",
                 shuffle = TRUE,
                 group.by = "group",
                 cols = group_colors,
                 repel = TRUE)
umap6

# group2
umap7 <- DimPlot(object = mouse.annotated, 
                 reduction = "umap",
                 shuffle = TRUE,
                 group.by = "group2",
                 cols = group2_colors,
                 repel = TRUE)
umap7

# diet
umap8 <- DimPlot(object = mouse.annotated, 
                 reduction = "umap",
                 shuffle = TRUE,
                 group.by = "diet",
                 cols = diet_color,
                 repel = TRUE)
umap8
```

```{r save_umap_meta, echo=FALSE, eval=FALSE}
path <- paste0(out, "clustering_QC/umap_colored_by_")
pdf(paste0(path, "isoform.pdf"), width = 6, height = 4)
umap1
dev.off()

pdf(paste0(path, "sex.pdf"), width = 6, height = 4)
umap2
dev.off()

pdf(paste0(path, "sample.pdf"), width = 6, height = 4)
umap3
dev.off()

pdf(paste0(path, "cell_cycle_phase.pdf"), width = 6, height = 4)
umap4
dev.off()

pdf(paste0(path, "mito_factor.pdf"), width = 6, height = 4)
umap5
dev.off()

pdf(paste0(path, "group.pdf"), width = 6, height = 4)
umap6
dev.off()

pdf(paste0(path, "group2.pdf"), width = 6, height = 4)
umap7
dev.off()

pdf(paste0(path, "diet.pdf"), width = 6, height = 4)
umap8
dev.off()

remove(umap1,umap2,umap3,umap4,umap5,umap6,umap7,umap8)
```

## Heatmap UMAP
```{r feature_plots_annotated, warning=FALSE, message=FALSE}
# UMAP percent.mt
f1 <- FeaturePlot(mouse.annotated, 
            reduction = "umap", 
            features = "percent.mt")  + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f1

# UMAP nCount
f2 <- FeaturePlot(mouse.annotated, 
            reduction = "umap",
            features = "nCount_RNA") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f2

# UMAP nFeature
f3 <- FeaturePlot(mouse.annotated, 
            reduction = "umap", 
            features = "nFeature_RNA") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f3

# UMAP percent.ribo
f4 <- FeaturePlot(mouse.annotated, 
            reduction = "umap", 
            features = "percent.ribo.protein") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f4

# UMAP cell.complexity
f5 <- FeaturePlot(mouse.annotated, 
            reduction = "umap", 
            features = "cell.complexity") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f5
```

```{r, echo=FALSE, eval=FALSE}
path <- paste0(out, "clustering_QC/umap_heatmap_")
pdf(paste0(path, "percent_mito.pdf"), width = 8, height = 6)
f1
dev.off()

pdf(paste0(path, "nCount.pdf"), width = 8, height = 6)
f2
dev.off()

pdf(paste0(path, "nFeature.pdf"), width = 8, height = 6)
f3
dev.off()

pdf(paste0(path, "percent_ribo.pdf"), width = 8, height = 6)
f4
dev.off()

pdf(paste0(path, "cell_complexity.pdf"), width = 8, height = 6)
f5
dev.off()

remove(f1,f2,f3,f4,f5)
```

## Percent cells per cluster
```{r percent_cells}
# isoform
b1 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, isoform) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=isoform)) +
  theme_classic() +
  geom_col() +
  scale_fill_manual(values = isoform_colors) +
  ggtitle("Percentage of isoform per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b1

# sex
b2 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, sex) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=sex)) +
  theme_classic() +
  geom_col() +
  scale_fill_manual(values = sex_colors) +
  ggtitle("Percentage of sex per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b2

# sample
b3 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, sample) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=sample)) +
  theme_classic() +
  geom_col() +
  scale_fill_manual(values = sample_colors) +
  ggtitle("Percentage of sample per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b3

# phase
b4 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, phase) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=phase)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of phase per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b4

# mito.factor
b5 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, mito.factor) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=mito.factor)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of mito.factor per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b5

# group
b6 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, group) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=group)) +
  theme_classic() +
  geom_col() +
  scale_fill_manual(values = group_colors) +
  ggtitle("Percentage of group per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b6

# group2
b7 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, group2) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=group2)) +
  theme_classic() +
  geom_col() +
  scale_fill_manual(values = group2_colors) +
  ggtitle("Percentage of group per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b7

# diet
b8 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, diet) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=diet)) +
  theme_classic() +
  geom_col() +
  scale_fill_manual(values = diet_color) +
  ggtitle("Percentage of diet per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b8
```

```{r save_percent_cells, echo=FALSE, eval=FALSE}
# save
path <- paste0(out, "clustering_QC/percent_")
pdf(paste0(path, "isoform_per_cluster.pdf"), width = 8, height = 6)
b1
dev.off()

# save
pdf(paste0(path, "sex_per_cluster.pdf"), width = 8, height = 6)
b2
dev.off()

# save
pdf(paste0(path, "sample_per_cluster.pdf"), width = 8, height = 6)
b3
dev.off()

# save
pdf(paste0(path, "phase_per_cluster.pdf"), width = 8, height = 6)
b4
dev.off()

# save
pdf(paste0(path, "mito_factor_per_cluster.pdf"), width = 8, height = 6)
b5
dev.off()

# save
pdf(paste0(path, "group_per_cluster.pdf"), width = 8, height = 6)
b6
dev.off()

# save
pdf(paste0(path, "group2_per_cluster.pdf"), width = 8, height = 6)
b7
dev.off()

# save
pdf(paste0(path, "diet_per_cluster.pdf"), width = 8, height = 6)
b8
dev.off()

# cleanup
remove(b1,b2,b3,b4,b5,b6,b7,b8)
```

## Cells per sample
```{r cells_per_sample}
data <- as.data.frame(table(mouse.annotated$sample))
colnames(data) <- c("sample","frequency")

ncells <- ggplot(data, aes(x = sample, y = frequency, fill = sample)) + 
  geom_col() +
  theme_classic() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            vjust=-0.25) +
  scale_fill_manual(values = sample_colors) + 
  scale_y_continuous(breaks = seq(0,3000, by = 1000), limits = c(0,3000)) +
  ggtitle("Filtered: cells per sample") +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))
ncells
```

```{r save_cells_per_sample, echo=FALSE, eval=FALSE}
pdf(paste0(out, "clustering_QC/cells_per_sample.pdf"), width = 6, height = 4)
ncells
dev.off()
```

## Cells per cluster
```{r cells_per_cluster}
data <- as.data.frame(table(mouse.annotated$annotated_clusters))
colnames(data) <- c("cluster","frequency")

ncells <- ggplot(data, aes(x = cluster, y = frequency, fill = cluster)) + 
  geom_col() +
  theme_classic() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            vjust=-0.25) +
  scale_fill_manual(values = cluster_colors) + 
  scale_y_continuous(breaks = seq(0,8000, by = 2000), limits = c(0,8000)) +
  ggtitle("Cells per Cluster") +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))
ncells
```

```{r save_cells_per_cluster, echo=FALSE, eval=FALSE}
pdf(paste0(out, "clustering_QC/cells_per_cluster.pdf"), width = 8, height = 6)
ncells
dev.off()
```

## Cells per cluster per sample
```{r cells_per_cluster_per_sample}
# extract data
data <- data.frame(cluster = mouse.annotated$annotated_clusters,
                   sample = mouse.annotated$sample)
data_summary <- data %>%
  group_by(cluster, sample) %>%
  summarise(count = n(), .groups = 'drop') %>%
  ungroup()

# Create the stacked bar plot
ncells <- ggplot(data_summary, aes(x = cluster, y = count, fill = sample)) +
  geom_bar(stat = "identity") +
  labs(y = "Frequency", x = "Cluster", fill = "Sample") +
  theme_minimal() +
  ggtitle("Frequency of Cells in Each Cluster with Sample Distribution") +
  scale_fill_manual(values = sample_colors) + 
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust=1)) +
  ylim(c(0,7000))
ncells

# reformat for table output
data_summary <- data_summary %>% pivot_wider(names_from = cluster, values_from = count)
```

```{r save_cells_per_cluster_per_sample, echo=FALSE, eval=FALSE}
pdf(paste0(out, "clustering_QC/cells_per_cluster_with_sample_distribution.pdf"),
    width = 8, height = 6)
ncells
dev.off()

write.table(data_summary, 
            file = paste0(out, "clustering_QC/cells_per_cluster_with_sample_distribution.tsv"), 
            sep = "\t",quote = FALSE, row.names = FALSE)
```

## Cells per cluster per group
```{r cells_per_cluster_per_group}
# extract data
data <- data.frame(cluster = mouse.annotated$annotated_clusters,
                   group = mouse.annotated$group)
data_summary <- data %>%
  group_by(cluster, group) %>%
  summarise(count = n(), .groups = 'drop') %>%
  ungroup()

# Create the stacked bar plot
ncells <- ggplot(data_summary, aes(x = cluster, y = count, fill = group)) +
  geom_bar(stat = "identity") +
  labs(y = "Frequency", x = "Cluster", fill = "Group") +
  theme_minimal() +
  ggtitle("Frequency of Cells in Each Cluster with Group Distribution") +
  scale_fill_manual(values = group_colors) + 
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust=1)) +
  ylim(c(0,7000))
ncells

# reformat for table output
data_summary <- data_summary %>% pivot_wider(names_from = cluster, values_from = count)
```

```{r save_cells_per_cluster_per_group, echo=FALSE, eval=FALSE}
pdf(paste0(out, "clustering_QC/cells_per_cluster_with_group_distribution.pdf"),
    width = 8, height = 6)
ncells
dev.off()

write.table(data_summary, 
            file = paste0(out, "clustering_QC/cells_per_cluster_with_group_distribution.tsv"), 
            sep = "\t",quote = FALSE, row.names = FALSE)
```

# Differential expression
## E3PM vs E3CM
```{r E3PM_vs_E3CM, warning=FALSE, message=FALSE, eval=FALSE}
DE_within_each_cluster(obj = mouse.annotated,
                       outDir = paste0(out, "DEGs/DEG_tables"),
                       clusterCol = "annotated_clusters",
                       groupCol = "group",
                       group1 = "E3PM",
                       group2 = "E3CM")
gc()
```

## E4PM vs E4CM
```{r E4PM_vs_E4CM, warning=FALSE, message=FALSE, eval=FALSE}
DE_within_each_cluster(obj = mouse.annotated,
                       outDir = paste0(out, "DEGs/DEG_tables"),
                       clusterCol = "annotated_clusters",
                       groupCol = "group",
                       group1 = "E4PM",
                       group2 = "E4CM")
gc()
```

## E3PF vs E3CF
```{r E3PF_vs_E3CF, warning=FALSE, message=FALSE, eval=FALSE}
DE_within_each_cluster(obj = mouse.annotated,
                       outDir = paste0(out, "DEGs/DEG_tables"),
                       clusterCol = "annotated_clusters",
                       groupCol = "group",
                       group1 = "E3PF",
                       group2 = "E3CF")
gc()
```

## E4PF vs E4CF
```{r E4PF_vs_E4CF, warning=FALSE, message=FALSE, eval=FALSE}
DE_within_each_cluster(obj = mouse.annotated,
                       outDir = paste0(out, "DEGs/DEG_tables"),
                       clusterCol = "annotated_clusters",
                       groupCol = "group",
                       group1 = "E4PF",
                       group2 = "E4CF")
gc()
```

## E4CM vs E3CM
```{r E4CM_vs_E3CM, warning=FALSE, message=FALSE, eval=FALSE}
DE_within_each_cluster(obj = mouse.annotated,
                       outDir = paste0(out, "DEGs/DEG_tables"),
                       clusterCol = "annotated_clusters",
                       groupCol = "group",
                       group1 = "E4CM",
                       group2 = "E3CM")
gc()
```

## E4CF vs E3CF
```{r E4CF_vs_E3CF, warning=FALSE, message=FALSE, eval=FALSE}
DE_within_each_cluster(obj = mouse.annotated,
                       outDir = paste0(out, "DEGs/DEG_tables"),
                       clusterCol = "annotated_clusters",
                       groupCol = "group",
                       group1 = "E4CF",
                       group2 = "E3CF")
gc()
```

## E4PM vs E3PM
```{r E4PM_vs_E3PM, warning=FALSE, message=FALSE, eval=FALSE}
DE_within_each_cluster(obj = mouse.annotated,
                       outDir = paste0(out, "DEGs/DEG_tables"),
                       clusterCol = "annotated_clusters",
                       groupCol = "group",
                       group1 = "E4PM",
                       group2 = "E3PM")
gc()
```

## E4PF vs E3PF
```{r E4PF_vs_E3PF}
DE_within_each_cluster(obj = mouse.annotated,
                       outDir = paste0(out, "DEGs/DEG_tables"),
                       clusterCol = "annotated_clusters",
                       groupCol = "group",
                       group1 = "E4PF",
                       group2 = "E3PF")
gc()
```

## E4PM vs E3CM
```{r E4PM_vs_E3CM}
DE_within_each_cluster(obj = mouse.annotated,
                       outDir = paste0(out, "DEGs/DEG_tables"),
                       clusterCol = "annotated_clusters",
                       groupCol = "group",
                       group1 = "E4PM",
                       group2 = "E3CM")
gc()
```

## E4PF vs E3CF
```{r E4PF_vs_E3CF}
DE_within_each_cluster(obj = mouse.annotated,
                       outDir = paste0(out, "DEGs/DEG_tables"),
                       clusterCol = "annotated_clusters",
                       groupCol = "group",
                       group1 = "E4PF",
                       group2 = "E3CF")
gc()
```

# Shiny App
- scDRcell is the function used to plot the first UMAP
- inpConf is an argument of scDRcell, inpConf = sc1conf
- infConf gets subset by UI column based on inpdrX which is the user's input for x-axis
- inpdrX is used to set the x-axis label
```{r shiny_app, eval=FALSE}
# create new object
shiny.obj <- mouse.annotated
shiny.obj <- JoinLayers(shiny.obj)

# set default params
VariableFeatures(shiny.obj) <- shiny.obj@assays$SCT@var.features
DefaultAssay(shiny.obj) <- "RNA"
Idents(shiny.obj) <- "annotated_clusters"

# create config
names <- colnames(shiny.obj@meta.data)
names <- names[c(28,29,2:12,14:27)]
sc.config <- createConfig(obj = shiny.obj,
                          meta.to.include = names)

# change wd
setwd("../../results/")

# output shiny app folder
makeShinyApp(obj = shiny.obj, 
             scConf = sc.config, 
             gene.mapping = TRUE,
             shiny.title = "Meningeal Dura scRNAseq")

# manual config edits
sc1conf <- readRDS("shinyApp/sc1conf.rds")
cluster_hex <- "#B5B9BA|#3385BB|#40BBFF|#A5D5A9|#1C7E24|#F57C7C|#E42622|#FBB268|#FE8D19|#DE9E83|#A6CEE3|#9D7BBA|#977899"
sc1conf[2,4] <- cluster_hex
group_hex <- "#B89B74|#725E47|#B15928|#F7590B|#A6CEE3|#34B1E8|#ED8F47|#F4CC25"
sc1conf[18,4] <- group_hex
saveRDS(sc1conf,"shinyApp/sc1conf.rds")
```
