---
title: "PLX5622 Diet Mice scRNAseq"
subtitle: "All Clusters Pass 1, Downsampled, QC and DE"
author: "Kennedi Todd"
date: "11/25/2024"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

# Setup
## Working directory
```{r working_directory}
knitr::opts_knit$set(root.dir = ".")
```

## Libraries
```{r libraries, message=FALSE, warning=FALSE}
library(ComplexUpset) # instersection_size()
library(dplyr)        # ungroup()
library(ggrepel)      # geom_text_repel()
library(ggtree)       # BuildClusterTree()
library(gridExtra)    # grid.arrange()
library(gtools)       # smartbind()
library(parallel)     # detectCores()
library(plotly)       # plot_ly()
library(Seurat)       # DimPlot()
library(tidyr)        # %>%
library(UpSetR)       # fromList()
```

## Variables and functions
```{r set_variables_and_functions}
# variables
out <- "../../results/all_clusters/downsampled/"

sample_order <- c("E3CF1","E3CF2","E3CM1","E3CM2","E3PF1","E3PF2","E3PM1","E3PM2",
                  "E4CF1","E4CF2","E4CM1","E4CM2","E4PF1","E4PF2","E4PM1","E4PM2")
sample_colors <- c("#9e0909","#f75959","#f0b402","#f5d67a",
                   "#fbff0a","#fafaa7","#1d8c0e","#63e851",
                   "#0271f0","#9cc3f0","#6c39f7","#c589fa",
                   "#ed2af7","#f5c1f7","#755410","#b5aa82")

group_order <- c("E3CM","E3PM","E4CM","E4PM",
                 "E3CF","E3PF","E4CF","E4PF")
group_colors <- c("#B89B74","#725E47","#B15928","#F7590B",
                  "#A6CEE3","#34B1E8","#ED8F47","#F4CC25")

group2_order <- c("E3C","E3P","E4C","E4P")
group2_colors <- c("#f75959","#fbff0a","#63e851","#0271f0")

isoform_order <- c("E4","E3")
isoform_colors <- c("darkgray","cornflowerblue")

sex_order <- c("Male","Female")
sex_colors <- c("green","purple")

diet_order <- c("control","PLX5622")
diet_color <- c("coral","cyan")

# single cell functions
files <- list.files("../../functions", full.names = TRUE)
invisible(lapply(files, source))
```

## Load data
```{r read_pass1_annotated_obj}
mouse.annotated <- readRDS("../../rObjects/annotated_seurat_obj.rds")
```

# Downsample
```{r downsample}
# check cells per sample
table(mouse.annotated$sample)

# downsample the number of cells per identity class
Idents(mouse.annotated) <- "sample"
mouse.downsample <- subset(x = mouse.annotated, downsample = 2000)

# check
table(mouse.downsample$sample)

# cleanup
remove(mouse.annotated)
```

```{r save_downsampled_obj, eval=FALSE, echo=FALSE}
saveRDS(mouse.downsample, "../../rObjects/annotated_downsampled_seurat_obj.rds")
```

# QC
## 2D UMAP
```{r annotated_umap}
# set colors
cluster_colors <- c("#B5B9BA", # Pericytes and SMCs
                    "#3385BB", # BECs
                    "#40BBFF", # LECs
                    "#A5D5A9", # B cells
                    "#1C7E24", # T and NK cells
                    "#F57C7C", # ILCs
                    "#E42622", # Dendritic cells
                    "#FBB268", # Neutrophils
                    "#FE8D19", # Macrophages
                    "#DE9E83", # Myeloid precursors
                    "#A6CEE3", # Mast cells
                    "#9D7BBA", # Fibroblasts
                    "#977899") # Schwann cells

# umap
Idents(mouse.downsample) <- "annotated_clusters"
umap1 <- DimPlot(object = mouse.downsample,
                 reduction = "umap",
                 repel = TRUE,
                 group.by = "annotated_clusters",
                 cols = cluster_colors)
umap1

umap2 <- DimPlot(object = mouse.downsample,
                 reduction = "umap",
                 repel = TRUE,
                 dims = c(2,3),
                 group.by = "annotated_clusters",
                 cols = cluster_colors)
umap2
```

```{r save_annotated_umap}
pdf(paste0(out, "UMAP/annotated_clusters_dim1&2.pdf"), width = 7, height = 4)
umap1
dev.off()

pdf(paste0(out, "UMAP/annotated_clusters_dim2&3.pdf"), width = 7, height = 4)
umap2
dev.off()

remove(umap1,umap2)
```

## Cluster markers
```{r cluster_markers_violin}
goi <- c("Pdgfrb","Acta2","Vwf","Cldn5","Pecam1","Flt4","Prox1","Lyve1","Ptprc",
         "Cd19","Ms4a1","Ighd","Igha","Sdc1","Cd3e","Trbc2","Il7r","Nkg7","Klrb1b",
         "Klrb1c","Gata3","Rora","Itgax","H2-Eb1","Ccr2","Ly6c2","Lyz2","Ly6g","Itgam",
         "Mrc1","Csf1r","Cd38","Mki67","Mcpt4","Ms4a2","Col1a2","Plp1")

v <- VlnPlot(mouse.downsample,
             features = goi,
             split.by = "annotated_clusters",
             flip = TRUE,
             stack = TRUE,
             cols = cluster_colors)
v
```

```{r save_ann_markers, echo=FALSE}
pdf(paste0(out, "markers/sandro_markers_violin_annotated.pdf"), width = 10, height = 8)
v
dev.off()
remove(v)
```

## Split UMAP
```{r split_umaps}
# Apoe isoform
umap1 <- DimPlot(object = mouse.downsample, 
                 group.by = "annotated_clusters",
                 reduction = "umap",
                 split.by = "isoform",
                 repel = TRUE,
                 cols = cluster_colors)
umap1

# sex
umap2 <- DimPlot(object = mouse.downsample, 
                 group.by = "annotated_clusters",
                 reduction = "umap",
                 split.by = "sex",
                 repel = TRUE,
                 cols = cluster_colors)
umap2

# sample
umap3 <- DimPlot(object = mouse.downsample, 
                 group.by = "annotated_clusters",
                 reduction = "umap",
                 split.by = "sample",
                 ncol = 4,
                 repel = TRUE,
                 cols = cluster_colors)
umap3

# phase
umap4 <- DimPlot(object = mouse.downsample, 
                 group.by = "annotated_clusters",
                 reduction = "umap",
                 split.by = "phase",
                 cols = cluster_colors,
                 repel = TRUE)
umap4

# mito.factor
umap5 <- DimPlot(object = mouse.downsample, 
                 group.by = "annotated_clusters",
                 reduction = "umap",
                 split.by = "mito.factor",
                 cols = cluster_colors,
                 ncol = 2,
                 repel = TRUE)
umap5

# group
umap6 <- DimPlot(object = mouse.downsample, 
                 group.by = "annotated_clusters",
                 reduction = "umap",
                 split.by = "group",
                 cols = cluster_colors,
                 ncol = 4,
                 repel = TRUE)
umap6

# group2
umap7 <- DimPlot(object = mouse.downsample, 
                 group.by = "annotated_clusters",
                 reduction = "umap",
                 split.by = "group2",
                 cols = cluster_colors,
                 ncol = 2,
                 repel = TRUE)
umap7

# diet
umap8 <- DimPlot(object = mouse.downsample, 
                 group.by = "annotated_clusters",
                 reduction = "umap",
                 split.by = "diet",
                 cols = cluster_colors,
                 ncol = 2,
                 repel = TRUE)
umap8
```

```{r, echo=FALSE, eval=FALSE}
path <- paste0(out, "clustering_QC/")
pdf(paste0(path, "umap_split_isoform.pdf"), width = 12, height = 6)
umap1
dev.off()

pdf(paste0(path, "umap_split_sex.pdf"), width = 12, height = 6)
umap2
dev.off()

pdf(paste0(path, "umap_split_sample.pdf"), width = 14, height = 10)
umap3
dev.off()

pdf(paste0(path, "umap_split_cell_cycle_phase.pdf"), width = 14, height = 6)
umap4
dev.off()

pdf(paste0(path, "umap_split_mito_factor.pdf"), width = 12, height = 8)
umap5
dev.off()

pdf(paste0(path, "umap_split_group.pdf"), width = 16, height = 8)
umap6
dev.off()

pdf(paste0(path, "umap_split_group2.pdf"), width = 12, height = 8)
umap7
dev.off()

pdf(paste0(path, "umap_split_diet.pdf"), width = 12, height = 6)
umap8
dev.off()

remove(umap1,umap2,umap3,umap4,umap5,umap6,umap7,umap8)
```

## Heatmap UMAP
```{r feature_plots_annotated, warning=FALSE, message=FALSE}
# UMAP percent.mt
f1 <- FeaturePlot(mouse.downsample, 
            reduction = "umap", 
            features = "percent.mt")  + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f1

# UMAP nCount
f2 <- FeaturePlot(mouse.downsample, 
            reduction = "umap",
            features = "nCount_RNA") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f2

# UMAP nFeature
f3 <- FeaturePlot(mouse.downsample, 
            reduction = "umap", 
            features = "nFeature_RNA") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f3

# UMAP percent.ribo
f4 <- FeaturePlot(mouse.downsample, 
            reduction = "umap", 
            features = "percent.ribo.protein") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f4

# UMAP cell.complexity
f5 <- FeaturePlot(mouse.downsample, 
            reduction = "umap", 
            features = "cell.complexity") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f5
```

```{r, echo=FALSE, eval=FALSE}
path <- paste0(out, "clustering_QC/umap_heatmap_")
pdf(paste0(path, "percent_mito.pdf"), width = 8, height = 6)
f1
dev.off()

pdf(paste0(path, "nCount.pdf"), width = 8, height = 6)
f2
dev.off()

pdf(paste0(path, "nFeature.pdf"), width = 8, height = 6)
f3
dev.off()

pdf(paste0(path, "percent_ribo.pdf"), width = 8, height = 6)
f4
dev.off()

pdf(paste0(path, "cell_complexity.pdf"), width = 8, height = 6)
f5
dev.off()

remove(f1,f2,f3,f4,f5)
```

## Percent cells per cluster
```{r percent_cells}
# isoform
b1 <- mouse.downsample@meta.data %>%
  group_by(annotated_clusters, isoform) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=isoform)) +
  theme_classic() +
  geom_col() +
  scale_fill_manual(values = isoform_colors) +
  ggtitle("Percentage of isoform per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b1

# sex
b2 <- mouse.downsample@meta.data %>%
  group_by(annotated_clusters, sex) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=sex)) +
  theme_classic() +
  geom_col() +
  scale_fill_manual(values = sex_colors) +
  ggtitle("Percentage of sex per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b2

# sample
b3 <- mouse.downsample@meta.data %>%
  group_by(annotated_clusters, sample) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=sample)) +
  theme_classic() +
  geom_col() +
  scale_fill_manual(values = sample_colors) +
  ggtitle("Percentage of sample per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b3

# phase
b4 <- mouse.downsample@meta.data %>%
  group_by(annotated_clusters, phase) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=phase)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of phase per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b4

# mito.factor
b5 <- mouse.downsample@meta.data %>%
  group_by(annotated_clusters, mito.factor) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=mito.factor)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of mito.factor per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b5

# group
b6 <- mouse.downsample@meta.data %>%
  group_by(annotated_clusters, group) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=group)) +
  theme_classic() +
  geom_col() +
  scale_fill_manual(values = group_colors) +
  ggtitle("Percentage of group per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b6

# group2
b7 <- mouse.downsample@meta.data %>%
  group_by(annotated_clusters, group2) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=group2)) +
  theme_classic() +
  geom_col() +
  scale_fill_manual(values = group2_colors) +
  ggtitle("Percentage of group per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b7

# diet
b8 <- mouse.downsample@meta.data %>%
  group_by(annotated_clusters, diet) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=diet)) +
  theme_classic() +
  geom_col() +
  scale_fill_manual(values = diet_color) +
  ggtitle("Percentage of diet per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b8
```

```{r,echo=FALSE,eval=FALSE}
# save
path <- paste0(out, "clustering_QC/percent_")
pdf(paste0(path, "isoform_per_cluster.pdf"), width = 8, height = 6)
b1
dev.off()

# save
pdf(paste0(path, "sex_per_cluster.pdf"), width = 8, height = 6)
b2
dev.off()

# save
pdf(paste0(path, "sample_per_cluster.pdf"), width = 8, height = 6)
b3
dev.off()

# save
pdf(paste0(path, "phase_per_cluster.pdf"), width = 8, height = 6)
b4
dev.off()

# save
pdf(paste0(path, "mito_factor_per_cluster.pdf"), width = 8, height = 6)
b5
dev.off()

# save
pdf(paste0(path, "group_per_cluster.pdf"), width = 8, height = 6)
b6
dev.off()

# save
pdf(paste0(path, "group2_per_cluster.pdf"), width = 8, height = 6)
b7
dev.off()

# save
pdf(paste0(path, "diet_per_cluster.pdf"), width = 8, height = 6)
b8
dev.off()

# cleanup
remove(b1,b2,b3,b4,b5,b6,b7,b8)
```

## Cluster tree
```{r cluster_tree_indv_annotated, message=FALSE, warning=FALSE}
Idents(mouse.downsample) <- mouse.downsample$annotated_clusters
mouse.downsample <- BuildClusterTree(object = mouse.downsample,
                                     dims = 1:15,
                                     reorder = FALSE,
                                     reorder.numeric = FALSE)

tree <- mouse.downsample@tools$BuildClusterTree
tree$tip.label <- paste0(tree$tip.label)

p <- ggtree::ggtree(tree, aes(x, y)) +
  scale_y_reverse() +
  ggtree::geom_tree() +
  ggtree::theme_tree() +
  ggtree::geom_tiplab(offset = 1) +
  ggtree::geom_tippoint(color = cluster_colors[1:length(tree$tip.label)], shape = 16, size = 5) +
  coord_cartesian(clip = 'off') +
  theme(plot.margin = unit(c(0,2.5,0,0), 'cm'))
p
```

```{r save_annotated_cluster_tree, echo=FALSE, eval=FALSE}
pdf(paste0(out, "clustering_QC/cluster_tree_annotated.pdf"), width = 20, height = 10)
p
dev.off()
```

## Cells per sample
```{r}
data <- as.data.frame(table(mouse.downsample$sample))
colnames(data) <- c("sample","frequency")

ncells <- ggplot(data, aes(x = sample, y = frequency, fill = sample)) + 
  geom_col() +
  theme_classic() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            vjust=-0.25) +
  scale_fill_manual(values = sample_colors) + 
  scale_y_continuous(breaks = seq(0,3000, by = 1000), limits = c(0,3000)) +
  ggtitle("Filtered: cells per sample") +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))
ncells
```

```{r,echo=FALSE,eval=FALSE}
pdf(paste0(out, "clustering_QC/cells_per_sample.pdf"), width = 6, height = 4)
ncells
dev.off()
```

## Cells per cluster
```{r}
data <- as.data.frame(table(mouse.downsample$annotated_clusters))
colnames(data) <- c("cluster","frequency")

ncells <- ggplot(data, aes(x = cluster, y = frequency, fill = cluster)) + 
  geom_col() +
  theme_classic() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            vjust=-0.25) +
  scale_fill_manual(values = cluster_colors) + 
  scale_y_continuous(breaks = seq(0,8000, by = 2000), limits = c(0,8000)) +
  ggtitle("Cells per Cluster") +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))
ncells
```

```{r,echo=FALSE,eval=FALSE}
pdf(paste0(out, "clustering_QC/cells_per_cluster.pdf"), width = 8, height = 6)
ncells
dev.off()
```

## Cells per cluster per sample
```{r cells_per_cluster}
# extract data
data <- data.frame(cluster = mouse.downsample$annotated_clusters,
                   sample = mouse.downsample$sample)
data_summary <- data %>%
  group_by(cluster, sample) %>%
  summarise(count = n(), .groups = 'drop') %>%
  ungroup()

# Create the stacked bar plot
ncells <- ggplot(data_summary, aes(x = cluster, y = count, fill = sample)) +
  geom_bar(stat = "identity") +
  labs(y = "Frequency", x = "Cluster", fill = "Sample") +
  theme_minimal() +
  ggtitle("Frequency of Cells in Each Cluster with Sample Distribution") +
  scale_fill_manual(values = sample_colors) + 
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust=1)) +
  ylim(c(0,7000))
ncells

# reformat for table output
data_summary <- data_summary %>% pivot_wider(names_from = cluster, values_from = count)
```

```{r, echo=FALSE, eval=FALSE}
pdf(paste0(out, "clustering_QC/cells_per_cluster_with_sample_distribution.pdf"),
    width = 8, height = 6)
ncells
dev.off()

write.table(data_summary, 
            file = paste0(out, "clustering_QC/cells_per_cluster_with_sample_distribution.tsv"), 
            sep = "\t",quote = FALSE, row.names = FALSE)
```

## Cells per cluster per group
```{r cells_per_cluster}
# extract data
data <- data.frame(cluster = mouse.downsample$annotated_clusters,
                   group = mouse.downsample$group)
data_summary <- data %>%
  group_by(cluster, group) %>%
  summarise(count = n(), .groups = 'drop') %>%
  ungroup()

# Create the stacked bar plot
ncells <- ggplot(data_summary, aes(x = cluster, y = count, fill = group)) +
  geom_bar(stat = "identity") +
  labs(y = "Frequency", x = "Cluster", fill = "Group") +
  theme_minimal() +
  ggtitle("Frequency of Cells in Each Cluster with Group Distribution") +
  scale_fill_manual(values = group_colors) + 
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust=1)) +
  ylim(c(0,7000))
ncells

# reformat for table output
data_summary <- data_summary %>% pivot_wider(names_from = cluster, values_from = count)
```

```{r, echo=FALSE, eval=FALSE}
pdf(paste0(out, "clustering_QC/cells_per_cluster_with_group_distribution.pdf"),
    width = 8, height = 6)
ncells
dev.off()

write.table(data_summary, 
            file = paste0(out, "clustering_QC/cells_per_cluster_with_group_distribution.tsv"), 
            sep = "\t",quote = FALSE, row.names = FALSE)
```

# Differential expression
## E3PM vs E3CM
```{r E3PM_vs_E3CM, warning=FALSE, message=FALSE, eval=FALSE}
DE_within_each_cluster(obj = mouse.downsample,
                       outDir = paste0(out, "DEGs/DEG_tables"),
                       clusterCol = "annotated_clusters",
                       groupCol = "group",
                       group1 = "E3PM",
                       group2 = "E3CM")
gc()
```

## E4PM vs E4CM
```{r E4PM_vs_E4CM, warning=FALSE, message=FALSE, eval=FALSE}
DE_within_each_cluster(obj = mouse.downsample,
                       outDir = paste0(out, "DEGs/DEG_tables"),
                       clusterCol = "annotated_clusters",
                       groupCol = "group",
                       group1 = "E4PM",
                       group2 = "E4CM")
gc()
```

## E3PF vs E3CF
```{r E3PF_vs_E3CF, warning=FALSE, message=FALSE, eval=FALSE}
DE_within_each_cluster(obj = mouse.downsample,
                       outDir = paste0(out, "DEGs/DEG_tables"),
                       clusterCol = "annotated_clusters",
                       groupCol = "group",
                       group1 = "E3PF",
                       group2 = "E3CF")
gc()
```

## E4PF vs E4CF
```{r E4PF_vs_E4CF, warning=FALSE, message=FALSE, eval=FALSE}
DE_within_each_cluster(obj = mouse.downsample,
                       outDir = paste0(out, "DEGs/DEG_tables"),
                       clusterCol = "annotated_clusters",
                       groupCol = "group",
                       group1 = "E4PF",
                       group2 = "E4CF")
gc()
```

## E4CM vs E3CM
```{r E4CM_vs_E3CM, warning=FALSE, message=FALSE, eval=FALSE}
DE_within_each_cluster(obj = mouse.downsample,
                       outDir = paste0(out, "DEGs/DEG_tables"),
                       clusterCol = "annotated_clusters",
                       groupCol = "group",
                       group1 = "E4CM",
                       group2 = "E3CM")
gc()
```

## E4CF vs E3CF
```{r E4CF_vs_E3CF, warning=FALSE, message=FALSE, eval=FALSE}
DE_within_each_cluster(obj = mouse.downsample,
                       outDir = paste0(out, "DEGs/DEG_tables"),
                       clusterCol = "annotated_clusters",
                       groupCol = "group",
                       group1 = "E4CF",
                       group2 = "E3CF")
gc()
```

## E4PM vs E3PM
```{r E4PM_vs_E3PM, warning=FALSE, message=FALSE, eval=FALSE}
DE_within_each_cluster(obj = mouse.downsample,
                       outDir = paste0(out, "DEGs/DEG_tables"),
                       clusterCol = "annotated_clusters",
                       groupCol = "group",
                       group1 = "E4PM",
                       group2 = "E3PM")
gc()
```

## E4PF vs E3PF
```{r E4PF_vs_E3PF}
DE_within_each_cluster(obj = mouse.downsample,
                       outDir = paste0(out, "DEGs/DEG_tables"),
                       clusterCol = "annotated_clusters",
                       groupCol = "group",
                       group1 = "E4PF",
                       group2 = "E3PF")
gc()
```

## E4PM vs E3CM
```{r E3PM_vs_E3CM}
DE_within_each_cluster(obj = mouse.downsample,
                       outDir = paste0(out, "DEGs/DEG_tables"),
                       clusterCol = "annotated_clusters",
                       groupCol = "group",
                       group1 = "E4PM",
                       group2 = "E3CM")
gc()
```

## E4PF vs E3CF
```{r E3PF_vs_E3CF}
DE_within_each_cluster(obj = mouse.downsample,
                       outDir = paste0(out, "DEGs/DEG_tables"),
                       clusterCol = "annotated_clusters",
                       groupCol = "group",
                       group1 = "E4PF",
                       group2 = "E3CF")
gc()
```

# DEG tables and figures
## Compare DEGs
```{r,eval=FALSE}
# get file paths
files <- list.files(paste0(out, "DEGs/DEG_tables"))
files <- files[grepl("_DEGs.tsv", files)]

# init data.frame
deg.df <- data.frame()

# loop through files
for (i in files) {
  
  # read table
  df <- read.table(paste0(out, "DEGs/DEG_tables/", i),
                   sep = "\t", 
                   header = TRUE)
  
  # filter
  df <- df[df$p_val_adj < 0.05,]
  
  # add direction column
  comparisonName <- gsub("_DEGs.tsv", "", i)
  direction <- df$avg_log2FC > 0
  direction <- gsub(TRUE, paste0(comparisonName, "_up"), direction)
  direction <- gsub(FALSE, paste0(comparisonName, "_down"), direction)
  df$direction <- direction
  
  # reformat table
  df <- df %>% 
    dplyr::count(cluster, direction) %>%
    tidyr::spread(cluster, n)
  
  # add to master table
  deg.df <- smartbind(deg.df, df)
}

# reformat
deg.df[is.na(deg.df)] <- 0
rownames(deg.df) <- deg.df$direction
deg.df$direction <- NULL
```

```{r,echo=FALSE,eval=FALSE}
# save
write.table(deg.df, 
            paste0(prefix,"DEG_comparison_FDRq_0.05_LFC_0.00.tsv"),
            sep = "\t", quote = FALSE, row.names = TRUE)
```

## DEG heatmap
```{r}
# set heatmap colors and names
#meta <- data.frame(diet = c(rep("PLX5622 vs Control", 8),
#                            rep("Control vs Control", 4)),
#                   allele = c(rep("E3 vs E3", 2),
#                              rep("E4 vs E4", 2),
#                              rep("E3 vs E3", 2),
#                              rep("E4 vs E4", 2),
#                              rep("E4 vs E3", 4)),
#                   sex = c(rep("Male", 4),
#                           rep("Female", 4),
#                           rep("Male", 2),
#                           rep("Female", 2)))
#rownames(meta) <- rownames(df)
paletteLength <- 100
myColor <- colorRampPalette(c("white","#f0eb9e","darkgreen"))(paletteLength)
#ann_colors = list(diet = c(`PLX5622 vs Control` = "cyan",
#                           `Control vs Control` = "coral"),
#                  allele = c(`E3 vs E3` = "#f75959",
#                             `E4 vs E4` = "#0271f0",
#                             `E4 vs E3` = "#fbff0a"),
#                  sex = c(Male = "#63e851",
#                          Female = "gray"))

# save
path <- paste0(prefix, "/DEG_comparison_heatmap_FDRq_0.05_LFC_0.00.pdf")
pdf(path, width = 10, height = 6)

# plot
pheatmap::pheatmap(deg.df,
                   main = "FDRq < 0.05, |LFC| > 0",
                   treeheight_row = 0,
                   treeheight_col = 0,
                   color = myColor,
                   cluster_rows = FALSE,
                   #annotation_row = meta,
                   #annotation_colors = ann_colors,
                   display_numbers = round(deg.df, digits = 0),
                   fontsize_number = 12,
                   number_color = "black")
```

## Volcano
```{r volcano_plot, message=FALSE, eval=FALSE}
variables <- c("E3PM_vs_E3CM","E4PM_vs_E4CM",
               "E3PF_vs_E3CF","E4PF_vs_E4CF",
               "E4CM_vs_E3CM","E4CF_vs_E3CF")
all_clusters <- levels(mouse.downsample$annotated_clusters)
prefix <- paste0(out, "DEGs/")

for (i in variables) {
  
  # read DEG file
  treatment_vs_control <- read.delim(paste0(prefix,"DEG_tables/",i,"_DEGs.tsv"))
  
  # assign colors
  color_values <- vector()
  max <- nrow(treatment_vs_control)
  for(row in 1:max){
    if (treatment_vs_control$p_val_adj[row] < 0.05){
      if (treatment_vs_control$avg_log2FC [row] > 0){
        color_values <- c(color_values, 1) # 1 when logFC > 0 and FDRq < 0.05
      }
      else if (treatment_vs_control$avg_log2FC[row] < 0){
        color_values <- c(color_values, 2) # 2 when logFC < 0 and FDRq < 0.05
      }
    }
    else{
      color_values <- c(color_values, 3) # 3 when FDRq >= 0.05
    }
  }
  treatment_vs_control$color_adjpval_0.05 <- factor(color_values)
  
  # loop through clusters
  for (j in all_clusters) {
    
    # subset cluster
    data <- subset(treatment_vs_control, cluster == j)
    
    # plot only if there are DEGs with p_val_adj < 0.05
    num <- subset(data, p_val_adj < 0.05)
    num <- nrow(num)
    if(num != 0) {
        
      # subset genes to label
      up <- data[data$color_adjpval_0.05 == 1,]
      up.sig <- up[order(up$p_val_adj),][1:15,]
      up.lfc <- up[order(up$avg_log2FC, decreasing = TRUE),][1:15,]
      up30 <- rbind(up.sig,up.lfc)
      up30 <- unique(up30)
      up30 <- up30[rowSums(is.na(up30)) != ncol(up30), ]
      down <- data[data$color_adjpval_0.05 == 2,]
      down.sig <- down[order(down$p_val_adj),][1:15,]
      down.lfc <- down[order(down$avg_log2FC, decreasing = FALSE),][1:15,]
      down30 <- rbind(down.sig,down.lfc)
      down30 <- unique(down30)
      down30 <- down30[rowSums(is.na(down30)) != ncol(down30), ]
      
      # set manual colors
      if (!1 %in% unique(data$color_adjpval_0.05)) {
        my_colors <- c("blue","gray")
      } else if (!2 %in% unique(data$color_adjpval_0.05)) {
        my_colors <- c("red","gray")
      } else if (!1 %in% unique(data$color_adjpval_0.05) && !2 %in% unique(data$color_adjpval_0.05)) {
        my_colors <- c("gray")
      } else {
        my_colors <- c("red","blue","gray")
      }
      
      # set significance threshold
      hadjpval <- (-log10(max(
        data$p_val[data$p_val_adj < 0.05], 
        na.rm=TRUE)))

      # plot
      p <-
        ggplot(data = data, 
               aes(x = avg_log2FC,  # x-axis is logFC
                   y = -log10(p_val),  # y-axis will be -log10 of P.Value
                   color = color_adjpval_0.05)) +  # color is based on factored color column
        geom_point(alpha = 0.8, size = 2) +  # create scatterplot, alpha makes points transparent
        theme_bw() +  # set color theme
        theme(legend.position = "none") +  # no legend
        scale_color_manual(values = my_colors) +  # set factor colors
        labs(
          title = "", # no main title
          x = expression(log[2](FC)), # x-axis title
          y = expression(-log[10] ~ "(" ~ italic("p") ~ "-value)") # y-axis title
        ) +
        theme(axis.title.x = element_text(size = 10),
              axis.text.x = element_text(size = 10)) +
        theme(axis.title.y = element_text(size = 10),
              axis.text.y = element_text(size = 10)) +
        geom_hline(yintercept = hadjpval,  #  horizontal line
                           colour = "#000000",
                           linetype = "dashed") +
        ggtitle(paste0(j,"\n",i,", p_val_adj < 0.05")) +
        geom_text_repel(data = up30,
                        aes(x = avg_log2FC, y= -log10(p_val), label = gene), 
                        color = "maroon", 
                        fontface="italic",
                        max.overlaps = getOption("ggrepel.max.overlaps", default = 30)
                        ) +
        geom_text_repel(data = down30,
                        aes(x = avg_log2FC, y= -log10(p_val), label = gene), 
                        color = "navyblue", 
                        fontface="italic",
                        max.overlaps = getOption("ggrepel.max.overlaps", default = 30)
                        )
      p
      
      # save
      clus <- j
      clus <- gsub(" ","_",clus)
      path <- paste0(prefix,"volcano/",i,"/",i,"_",clus,"_volcano.pdf")
      pdf(paste(path,".pdf"), height = 8, width = 8)
      print(p)
      dev.off()
      
      print(paste("i =",i,", j =",j))
      
    }
  } # end loop through clusters
} # end loop through variables
```

## Metascape input
```{r}
# set variables
thresh <- 0.05

# get file list
files <- list.files(paste0(out, "DEGs/DEG_tables/"))
keep <- grep("_DEGs.tsv", files)
files <- files[keep]
files <- paste0(out, "DEGs/DEG_tables/", files)

# get cell types
cell_types <- levels(mouse.downsample$annotated_clusters)

# loop through files
for (i in 1:length(files)) {
  
  # read table
  data <- read.table(files[i], header = TRUE, sep = "\t")
  data <- data[data$p_val_adj < thresh,]
  up.df <- data[data$avg_log2FC > 0,]
  down.df <- data[data$avg_log2FC < 0,]
  
  # loop through cluster
  for (j in cell_types) {
    
    # subset based on cluster
    print(paste0(i, ": ", j))
    up <- up.df[up.df$cluster == j,]
    up <- up$gene
    down <- down.df[down.df$cluster == j,]
    down <- down$gene
    
    # file name
    clus <- j
    clus <- gsub(" ","_",clus)
    file <- gsub(paste0(out, "DEGs/DEG_tables/"), "", files[i])
    file <- gsub("_DEGs.tsv", "", file)
    file <- paste0(out, "DEGs/metascape_input/", file, "/", clus)
     
    # save
    if(length(up) > 10) {
      write.table(x = up,
                  file = paste0(file, "_up_FDRq_", format(thresh, nsmall = 2), ".tsv"),
                  quote = FALSE,
                  row.names = FALSE,
                  col.names = FALSE)      
    }
    if (length(down) > 10) {
       write.table(x = down,
                file = paste0(file, "_down_FDRq_", format(thresh, nsmall = 2), ".tsv"),
                quote = FALSE,
                row.names = FALSE,
                col.names = FALSE)   
    }
  } # end j for loop
} # end i for loop
```

## Upset plots
### Function
```{r upset_function}
get_exclusive_intersection_genes <- function(upset_data, list_input) {
  # Initialize a list to store the results
  results <- list()
  
  # Get the names of the columns (gene sets)
  set_names <- colnames(upset_data)
  
  # For each row of the UpSet matrix, extract the exclusive intersection
  for (i in 1:nrow(upset_data)) {
    # Check which sets (columns) are "active" (i.e., 1) for the current row
    active_sets <- set_names[upset_data[i, ] == 1]
    
    if (length(active_sets) > 0) {
      # Find the intersection of genes in the active sets
      intersected_genes <- Reduce(intersect, list_input[active_sets])
      
      # Find genes that are present in any of the other sets (not part of this combination)
      other_sets <- set_names[upset_data[i, ] == 0]
      other_genes <- Reduce(union, list_input[other_sets])
      
      # Subtract the genes that are in the other sets to make the intersection exclusive
      exclusive_genes <- setdiff(intersected_genes, other_genes)
      
      # Create a label for the current intersection (e.g., "Set1 & Set2")
      intersection_label <- paste(active_sets, collapse = " & ")
      
      # Store the number of genes and the exclusive gene list
      results[[intersection_label]] <- list(
        total_genes = length(exclusive_genes),
        gene_list = paste(exclusive_genes, collapse = ", ")
      )
    }
  }
  
  return(results)
}
```

### P vs C Up-regulated
```{r E4_vs_E3_upreg_tables}
# read tables
prefix <- paste0(out, "DEGs/DEG_tables/")
E3PM_vs_E3CM <- read.table(paste0(prefix,"E3PM_vs_E3CM_DEGs.tsv"),
                         sep = "\t", 
                         header = TRUE)
E4PM_vs_E4CM <- read.table(paste0(prefix,"E4PM_vs_E4CM_DEGs.tsv"),
                         sep = "\t", 
                         header = TRUE)
E3PF_vs_E3CF <- read.table(paste0(prefix,"E3PF_vs_E3CF_DEGs.tsv"),
                           sep = "\t", 
                           header = TRUE)

E4PF_vs_E4CF <- read.table(paste0(prefix,"E4PF_vs_E4CF_DEGs.tsv"),
                           sep = "\t", 
                           header = TRUE)

# filter
E3PM_vs_E3CM <- E3PM_vs_E3CM[E3PM_vs_E3CM$p_val_adj < 0.05,]
E4PM_vs_E4CM <- E4PM_vs_E4CM[E4PM_vs_E4CM$p_val_adj < 0.05,]
E3PF_vs_E3CF <- E3PF_vs_E3CF[E3PF_vs_E3CF$p_val_adj < 0.05,]
E4PF_vs_E4CF <- E4PF_vs_E4CF[E4PF_vs_E4CF$p_val_adj < 0.05,]
```

```{r E4_vs_E3_upreg_upset, eval=FALSE}
# loop through cell types
clusters <- unique(c(E3PM_vs_E3CM$cluster, E4PM_vs_E4CM$cluster,
                     E3PF_vs_E3CF$cluster, E4PF_vs_E4CF$cluster))

for (i in clusters) {
  
  # Skip cluster with not enough DEGs
  if (i == "LECs") {
    next
  }
  
  # Subset df by cluster
  E3PM_vs_E3CM_ct <- E3PM_vs_E3CM[E3PM_vs_E3CM$cluster == i,]
  E4PM_vs_E4CM_ct <- E4PM_vs_E4CM[E4PM_vs_E4CM$cluster == i,]
  E3PF_vs_E3CF_ct <- E3PF_vs_E3CF[E3PF_vs_E3CF$cluster == i,]
  E4PF_vs_E4CF_ct <- E4PF_vs_E4CF[E4PF_vs_E4CF$cluster == i,]
  
  # Subset by log2FC
  E3PM_vs_E3CM_ct_up <- E3PM_vs_E3CM_ct[E3PM_vs_E3CM_ct$avg_log2FC > 0,][,"gene"]
  E4PM_vs_E4CM_ct_up <- E4PM_vs_E4CM_ct[E4PM_vs_E4CM_ct$avg_log2FC > 0,][,"gene"]
  E3PF_vs_E3CF_ct_up <- E3PF_vs_E3CF_ct[E3PF_vs_E3CF_ct$avg_log2FC > 0,][,"gene"]
  E4PF_vs_E4CF_ct_up <- E4PF_vs_E4CF_ct[E4PF_vs_E4CF_ct$avg_log2FC > 0,][,"gene"]
  axis_max <- max(length(E3PM_vs_E3CM_ct_up), length(E4PM_vs_E4CM_ct_up),
                  length(E3PF_vs_E3CF_ct_up), length(E4PF_vs_E4CF_ct_up)) + 300
  axis_max <- ceiling(axis_max / 100) * 100
 
  # format in a list
  list_input <- list(
    "E3PM vs E3CM Up-regulated" = E3PM_vs_E3CM_ct_up,
    "E4PM vs E4CM Up-regulated" = E4PM_vs_E4CM_ct_up,
    "E3PF vs E3CF Up-regulated" = E3PF_vs_E3CF_ct_up,
    "E4PF vs E4CF Up-regulated" = E4PF_vs_E4CF_ct_up
  )
  data <- UpSetR::fromList(list_input)
  
  # You will specify the male and female groups by their group number
  # Print this to help assign shapes/colors below
  group_mapping <- data.frame(
    Group_Number = 1:4,  # Group numbers (1-8)
    Gene_Set_Name = sort(colnames(data))  # Sorted gene set names
  )
  
  # Assign shapes based on group numbers
  male_groups <- c("2","4")  # Male group numbers
  female_groups <- c("1","3")  # Female group numbers
  
  # Assign colors and shapes based on specific group numbers
  upregulated_groups <- c("1","2","3","4")
  downregulated_groups <- c()
  
  # Create the basic UpSet plot (no dynamic colors or shapes) to extract n_points
  dummy_upset <- ComplexUpset::upset(
    data, 
    colnames(data),  # Use column names (gene sets)
    set_sizes = ComplexUpset::upset_set_size(),  # Display set sizes
    sort_sets = FALSE,
    
    # Plot the intersection matrix with points (default shapes)
    matrix = ComplexUpset::intersection_matrix(
      ggplot2::geom_point(size = 3)  # Simple points without dynamic aesthetics
    )
  )
  
  # Ensure n_points matches the actual data
  n_points <- nrow(dummy_upset$data)
  
  # Create empty vectors for colors and shapes
  color_vector <- rep(NA, n_points)  # To store colors
  shape_vector <- rep(NA, n_points)  # To store shapes
  
  # Assign shapes and colors based on group numbers
  for (j in 1:n_points) {
    group_value <- as.character(dummy_upset$data$group[j])
    
    # Assign shapes
    if (group_value %in% male_groups) {
      shape_vector[j] <- 22  # Filled square for Male comparisons
    } else if (group_value %in% female_groups) {
      shape_vector[j] <- 21  # Circle for Female comparisons
    }
    
    # Assign colors based on group numbers, only if the point is involved in an intersection
    if (dummy_upset$data$value[j]) {  # If the point is part of an intersection
      if (group_value %in% upregulated_groups) {
        color_vector[j] <- "red"  # Red for Up-regulated
      } else if (group_value %in% downregulated_groups) {
        color_vector[j] <- "blue"  # Blue for Down-regulated
      }
    } else {
      color_vector[j] <- NA  # No fill for points not involved in intersections
    }
  }
  
  # Ensure both vectors are of correct length
  if (length(shape_vector) != n_points || length(color_vector) != n_points) {
    stop("Error: shape_vector or color_vector length does not match the number of plotted points.")
  }
  
  # Reintroduce dynamic fill color for filled shapes
  upset_gene <- ComplexUpset::upset(
    data, 
    colnames(data),
    set_sizes = (
      ComplexUpset::upset_set_size() +
      geom_text(aes(label = ..count..), hjust = 1.1, stat = 'count') +
      expand_limits(y = axis_max)
    ),
    
    # Prevent automatic sorting of sets
    sort_sets = FALSE,
    
    matrix = ComplexUpset::intersection_matrix(
      ggplot2::geom_point(
        ggplot2::aes(shape = factor(shape_vector), fill = color_vector),  # Fill shapes with dynamic colors
        size = 3,
        stroke = 0.45  # Keeps the outline black
      )
    ) + ggplot2::scale_shape_manual(
      name = "Sex",
      labels = c("Male", "Female"),  # Legend labels
      values = c(21, 22) # Circle=21, Filled square=22
      ) + ggplot2::scale_fill_identity(na.translate = FALSE),  # Use the exact colors in color_vector and remove NA fill
    
      base_annotations = list(
      'Intersection size' = (
        intersection_size(bar_number_threshold = 1, width = 0.5) +
        scale_y_continuous(expand = expansion(mult = c(0, 0.05)), limits = c(0, axis_max)) +
        theme(
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_line(colour = 'black')
        )
      )
    ),
  )
  
  # give title
  upset_gene <- upset_gene + ggtitle(paste0(i,", adj_p_val < 0.05"))
  
  # remove legend
  upset_gene <- upset_gene + theme(legend.position = "None")
  
  # save
  i <- gsub(" ","_",i)
  pdf(paste0(out, "DEGs/upset/P_vs_C_upreg/",tolower(i),"_upset.pdf"), height = 6, width = 12)
  print(upset_gene)
  dev.off()
}
```

### P vs C Down-regulated
```{r E4_vs_E3_downreg_tables}
# read tables
prefix <- paste0(out, "DEGs/DEG_tables/")
E3PM_vs_E3CM <- read.table(paste0(prefix,"E3PM_vs_E3CM_DEGs.tsv"),
                         sep = "\t", 
                         header = TRUE)
E4PM_vs_E4CM <- read.table(paste0(prefix,"E4PM_vs_E4CM_DEGs.tsv"),
                         sep = "\t", 
                         header = TRUE)
E3PF_vs_E3CF <- read.table(paste0(prefix,"E3PF_vs_E3CF_DEGs.tsv"),
                           sep = "\t", 
                           header = TRUE)

E4PF_vs_E4CF <- read.table(paste0(prefix,"E4PF_vs_E4CF_DEGs.tsv"),
                           sep = "\t", 
                           header = TRUE)

# filter
E3PM_vs_E3CM <- E3PM_vs_E3CM[E3PM_vs_E3CM$p_val_adj < 0.05,]
E4PM_vs_E4CM <- E4PM_vs_E4CM[E4PM_vs_E4CM$p_val_adj < 0.05,]
E3PF_vs_E3CF <- E3PF_vs_E3CF[E3PF_vs_E3CF$p_val_adj < 0.05,]
E4PF_vs_E4CF <- E4PF_vs_E4CF[E4PF_vs_E4CF$p_val_adj < 0.05,]
```

```{r E4_vs_E3_downreg_upset, eval=FALSE}
# loop through cell types
clusters <- unique(c(E3PM_vs_E3CM$cluster, E4PM_vs_E4CM$cluster,
                     E3PF_vs_E3CF$cluster, E4PF_vs_E4CF$cluster))

for (i in clusters) {
  
  # Subset df by cluster
  E3PM_vs_E3CM_ct <- E3PM_vs_E3CM[E3PM_vs_E3CM$cluster == i,]
  E4PM_vs_E4CM_ct <- E4PM_vs_E4CM[E4PM_vs_E4CM$cluster == i,]
  E3PF_vs_E3CF_ct <- E3PF_vs_E3CF[E3PF_vs_E3CF$cluster == i,]
  E4PF_vs_E4CF_ct <- E4PF_vs_E4CF[E4PF_vs_E4CF$cluster == i,]
  
  # Subset by log2FC
  E3PM_vs_E3CM_ct_down <- E3PM_vs_E3CM_ct[E3PM_vs_E3CM_ct$avg_log2FC < 0,][,"gene"]
  E4PM_vs_E4CM_ct_down <- E4PM_vs_E4CM_ct[E4PM_vs_E4CM_ct$avg_log2FC < 0,][,"gene"]
  E3PF_vs_E3CF_ct_down <- E3PF_vs_E3CF_ct[E3PF_vs_E3CF_ct$avg_log2FC < 0,][,"gene"]
  E4PF_vs_E4CF_ct_down <- E4PF_vs_E4CF_ct[E4PF_vs_E4CF_ct$avg_log2FC < 0,][,"gene"]
  axis_max <- max(length(E3PM_vs_E3CM_ct_down), length(E4PM_vs_E4CM_ct_down),
                  length(E3PF_vs_E3CF_ct_down), length(E4PF_vs_E4CF_ct_down)) + 200
  axis_max <- ceiling(axis_max / 100) * 100
 
  # format in a list
  list_input <- list(
    "E3PM vs E3CM down-regulated" = E3PM_vs_E3CM_ct_down,
    "E4PM vs E4CM down-regulated" = E4PM_vs_E4CM_ct_down,
    "E3PF vs E3CF down-regulated" = E3PF_vs_E3CF_ct_down,
    "E4PF vs E4CF down-regulated" = E4PF_vs_E4CF_ct_down
  )
  data <- UpSetR::fromList(list_input)
  
  # You will specify the male and female group by their group number
  # Print this to help assign shapes/colors below
  group_mapping <- data.frame(
    Group_Number = 1:4,  # Group numbers (1-4)
    Gene_Set_Name = sort(colnames(data))  # Sorted gene set names
  )
  
  # Assign shapes based on group numbers
  male_groups <- c("2","4")  # Male group numbers
  female_groups <- c("1","3")  # Female group numbers
  
  # Assign colors and shapes based on specific group numbers
  upregulated_groups <- c()
  downregulated_groups <- c("1","2","3","4")
  
  # Create the basic upset plot (no dynamic colors or shapes) to extract n_points
  dummy_upset <- ComplexUpset::upset(
    data, 
    colnames(data),  # Use column names (gene sets)
    set_sizes = ComplexUpset::upset_set_size(),  # Display set sizes
    sort_sets = FALSE,
    
    # Plot the intersection matrix with points (default shapes)
    matrix = ComplexUpset::intersection_matrix(
      ggplot2::geom_point(size = 3)  # Simple points without dynamic aesthetics
    )
  )
  
  # Ensure n_points matches the actual data
  n_points <- nrow(dummy_upset$data)
  
  # Create empty vectors for colors and shapes
  color_vector <- rep(NA, n_points)  # To store colors
  shape_vector <- rep(NA, n_points)  # To store shapes
  
  # Assign shapes and colors based on group numbers
  for (j in 1:n_points) {
    group_value <- as.character(dummy_upset$data$group[j])
    
    # Assign shapes
    if (group_value %in% male_groups) {
      shape_vector[j] <- 22  # Filled square for Male comparisons
    } else if (group_value %in% female_groups) {
      shape_vector[j] <- 21  # Circle for Female comparisons
    }
    
    # Assign colors based on group numbers, only if the point is involved in an intersection
    if (dummy_upset$data$value[j]) {  # If the point is part of an intersection
      if (group_value %in% upregulated_groups) {
        color_vector[j] <- "red"  # Red for Up-regulated
      } else if (group_value %in% downregulated_groups) {
        color_vector[j] <- "blue"  # Blue for Down-regulated
      }
    } else {
      color_vector[j] <- NA  # No fill for points not involved in intersections
    }
  }
  
  # Ensure both vectors are of correct length
  if (length(shape_vector) != n_points || length(color_vector) != n_points) {
    stop("Error: shape_vector or color_vector length does not match the number of plotted points.")
  }
  
  # Reintroduce dynamic fill color for filled shapes
  upset_gene <- ComplexUpset::upset(
    data, 
    colnames(data),
    set_sizes = (
      ComplexUpset::upset_set_size() +
      geom_text(aes(label = ..count..), hjust = 1.1, stat = 'count') +
      expand_limits(y = axis_max)
    ),
    
    # Prevent automatic sorting of sets
    sort_sets = FALSE,
    
    matrix = ComplexUpset::intersection_matrix(
      ggplot2::geom_point(
        ggplot2::aes(shape = factor(shape_vector), fill = color_vector),  # Fill shapes with dynamic colors
        size = 3,
        stroke = 0.45  # Keeps the outline black
      )
    ) + ggplot2::scale_shape_manual(
      name = "Sex",
      labels = c("Male", "Female"),  # Legend labels
      values = c(21, 22) # Circle=21, Filled square=22
      ) + ggplot2::scale_fill_identity(na.translate = FALSE),  # Use the exact colors in color_vector and remove NA fill
    
      base_annotations = list(
      'Intersection size' = (
        intersection_size(bar_number_threshold = 1, width = 0.5) +
        scale_y_continuous(expand = expansion(mult = c(0, 0.05)), limits = c(0, axis_max)) +
        theme(
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_line(colour = 'black')
        )
      )
    ),
  )
  
  # give title
  upset_gene <- upset_gene + ggtitle(paste0(i,", adj_p_val < 0.05"))
  
  # remove legend
  upset_gene <- upset_gene + theme(legend.position = "None")
  
  # save
  i <- gsub(" ","_",i)
  pdf(paste0(out, "DEGs/upset/P_vs_C_downreg/",tolower(i),"_upset.pdf"), height = 6, width = 12)
  print(upset_gene)
  dev.off()
}
```

### E4C vs E3C
Desired order
- E4CF_vs_E3CF Up
- E4CM_vs_E3CM Up 
- E4CF_vs_E3CF Down
- E4CM_vs_E3CM Down
```{r E4C_vs_E3C_tables}
# set thresh
thresh <- 0.05

# read tables
prefix <- paste0(out, "DEGs/DEG_tables/")
E4CF_vs_E3CF <- read.table(paste0(prefix,"E4CF_vs_E3CF_DEGs.tsv"),
                           sep = "\t",
                           header = TRUE)
E4CM_vs_E3CM <- read.table(paste0(prefix,"E4CM_vs_E3CM_DEGs.tsv"),
                           sep = "\t",
                           header = TRUE)

# filter
E4CF_vs_E3CF <- E4CF_vs_E3CF[E4CF_vs_E3CF$p_val_adj < thresh,]
E4CM_vs_E3CM <- E4CM_vs_E3CM[E4CM_vs_E3CM$p_val_adj < thresh,]
```

```{r E4C_vs_E3C_upset, eval=FALSE}
# loop through cell types
clusters <- unique(c(E4CF_vs_E3CF$cluster, E4CM_vs_E3CM$cluster))

for (i in clusters) {
  
  # skip cluster with not enough DEGs
  #if (i == "LECs") {
  #  next
  #}
  # Subset df by cluster
  E4CF_vs_E3CF_ct <- E4CF_vs_E3CF[E4CF_vs_E3CF$cluster == i,]
  E4CM_vs_E3CM_ct <- E4CM_vs_E3CM[E4CM_vs_E3CM$cluster == i,]
  
  # Subset by log2FC
  E4CF_vs_E3CF_ct_up <- E4CF_vs_E3CF_ct[E4CF_vs_E3CF_ct$avg_log2FC > 0,][,"gene"]
  E4CM_vs_E3CM_ct_up <- E4CM_vs_E3CM_ct[E4CM_vs_E3CM_ct$avg_log2FC > 0,][,"gene"]
  E4CF_vs_E3CF_ct_down <- E4CF_vs_E3CF_ct[E4CF_vs_E3CF_ct$avg_log2FC < 0,][,"gene"]
  E4CM_vs_E3CM_ct_down <- E4CM_vs_E3CM_ct[E4CM_vs_E3CM_ct$avg_log2FC < 0,][,"gene"]
  axis_max <- max(length(E4CF_vs_E3CF_ct_up), length(E4CM_vs_E3CM_ct_up),
                  length(E4CF_vs_E3CF_ct_down), length(E4CM_vs_E3CM_ct_down)) + 300
  axis_max <- ceiling(axis_max / 100) * 100
 
  # format in a list
  list_input <- list(
    "E4CM vs E3CM Down-regulated" = E4CM_vs_E3CM_ct_down,
    "E4CF vs E3CF Down-regulated" = E4CF_vs_E3CF_ct_down,
    "E4CM vs E3CM Up-regulated" = E4CM_vs_E3CM_ct_up,
    "E4CF vs E3CF Up-regulated" = E4CF_vs_E3CF_ct_up
  )
  data <- UpSetR::fromList(list_input)
  
  # You will specify the male and female groups by their group number
  # Print this to help assign shapes/colors below
  group_mapping <- data.frame(
    Group_Number = 1:4,  # Group numbers (1-8)
    Gene_Set_Name = sort(colnames(data))  # Sorted gene set names
  )
  
  # Assign shapes based on group numbers
  male_groups <- c("3","4")  # Male group numbers
  female_groups <- c("1","2")  # Female group numbers
  
  # Assign colors and shapes based on specific group numbers
  upregulated_groups <- c("2","4")
  downregulated_groups <- c("1","3")
  
  # Create the basic UpSet plot (no dynamic colors or shapes) to extract n_points
  dummy_upset <- ComplexUpset::upset(
    data, 
    colnames(data),  # Use column names (gene sets)
    set_sizes = ComplexUpset::upset_set_size(),  # Display set sizes
    sort_sets = FALSE,
    
    # Plot the intersection matrix with points (default shapes)
    matrix = ComplexUpset::intersection_matrix(
      ggplot2::geom_point(size = 3)  # Simple points without dynamic aesthetics
    )
  )
  
  # Ensure n_points matches the actual data
  n_points <- nrow(dummy_upset$data)
  
  # Create empty vectors for colors and shapes
  color_vector <- rep(NA, n_points)  # To store colors
  shape_vector <- rep(NA, n_points)  # To store shapes
  
  # Assign shapes and colors based on group numbers
  for (j in 1:n_points) {
    group_value <- as.character(dummy_upset$data$group[j])
    
    # Assign shapes
    if (group_value %in% male_groups) {
      shape_vector[j] <- 22  # Filled square for Male comparisons
    } else if (group_value %in% female_groups) {
      shape_vector[j] <- 21  # Circle for Female comparisons
    }
    
    # Assign colors based on group numbers, only if the point is involved in an intersection
    if (dummy_upset$data$value[j]) {  # If the point is part of an intersection
      if (group_value %in% upregulated_groups) {
        color_vector[j] <- "red"  # Red for Up-regulated
      } else if (group_value %in% downregulated_groups) {
        color_vector[j] <- "blue"  # Blue for Down-regulated
      }
    } else {
      color_vector[j] <- NA  # No fill for points not involved in intersections
    }
  }
  
  # Ensure both vectors are of correct length
  if (length(shape_vector) != n_points || length(color_vector) != n_points) {
    stop("Error: shape_vector or color_vector length does not match the number of plotted points.")
  }
  
  # Reintroduce dynamic fill color for filled shapes
  upset_gene <- ComplexUpset::upset(
    data, 
    colnames(data),
    set_sizes = (
      ComplexUpset::upset_set_size() +
      geom_text(aes(label = ..count..), hjust = 1.1, stat = 'count') +
      expand_limits(y = axis_max)
    ),
    
    # Prevent automatic sorting of sets
    sort_sets = FALSE,
    
    matrix = ComplexUpset::intersection_matrix(
      ggplot2::geom_point(
        ggplot2::aes(shape = factor(shape_vector), fill = color_vector),  # Fill shapes with dynamic colors
        size = 3,
        stroke = 0.45  # Keeps the outline black
      )
    ) + ggplot2::scale_shape_manual(
      name = "Sex",
      labels = c("Male", "Female"),  # Legend labels
      values = c(21, 22) # Circle=21, Filled square=22
      ) + ggplot2::scale_fill_identity(na.translate = FALSE),  # Use the exact colors in color_vector and remove NA fill
    
      base_annotations = list(
      'Intersection size' = (
        intersection_size(bar_number_threshold = 1, width = 0.5) +
        scale_y_continuous(expand = expansion(mult = c(0, 0.05)), limits = c(0, axis_max)) +
        theme(
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_line(colour = 'black')
        )
      )
    ),
  )
  
  # give title
  upset_gene <- upset_gene + ggtitle(paste0(i,", adj_p_val < 0.05"))
  
  # remove legend
  upset_gene <- upset_gene + theme(legend.position = "None")
  
  # save
  i <- gsub(" ","_",i)
  pdf(paste0(out, "DEGs/upset/E4C_vs_E3C/", tolower(i), "_upset.pdf"), height = 6, width = 12)
  print(upset_gene)
  dev.off()
}
```

### E4 vs E3 Up
```{r E4_vs_E3_up_read_tables}
# set thresh
thresh <- 0.05

# read tables
prefix <- paste0(out, "DEGs/DEG_tables/")
E4CF_vs_E3CF <- read.table(paste0(prefix,"E4CF_vs_E3CF_DEGs.tsv"),
                           sep = "\t",
                           header = TRUE)
E4PF_vs_E4CF <- read.table(paste0(prefix,"E4PF_vs_E4CF_DEGs.tsv"),
                           sep = "\t",
                           header = TRUE)
E4CM_vs_E3CM <- read.table(paste0(prefix,"E4CM_vs_E3CM_DEGs.tsv"),
                           sep = "\t",
                           header = TRUE)
E4PM_vs_E4CM <- read.table(paste0(prefix,"E4PM_vs_E4CM_DEGs.tsv"),
                           sep = "\t",
                           header = TRUE)


# filter
E4CF_vs_E3CF <- E4CF_vs_E3CF[E4CF_vs_E3CF$p_val_adj < thresh,]
E4PF_vs_E4CF <- E4PF_vs_E4CF[E4PF_vs_E4CF$p_val_adj < thresh,]
E4CM_vs_E3CM <- E4CM_vs_E3CM[E4CM_vs_E3CM$p_val_adj < thresh,]
E4PM_vs_E4CM <- E4PM_vs_E4CM[E4PM_vs_E4CM$p_val_adj < thresh,]
```

```{r E4_vs_E3_down_upset, eval=FALSE}
# loop through cell types
clusters <- unique(c(E4CF_vs_E3CF$cluster, E4PF_vs_E4CF$cluster,
                     E4CM_vs_E3CM$cluster, E4PM_vs_E4CM$cluster))

for (i in clusters) {
  
  # skip cluster with not enough DEGs
  if (i == "LECs") {
    next
  }
  
  # Subset df by cluster
  E4CF_vs_E3CF_ct <- E4CF_vs_E3CF[E4CF_vs_E3CF$cluster == i,]
  E4PF_vs_E4CF_ct <- E4PF_vs_E4CF[E4PF_vs_E4CF$cluster == i,]
  E4CM_vs_E3CM_ct <- E4CM_vs_E3CM[E4CM_vs_E3CM$cluster == i,]
  E4PM_vs_E4CM_ct <- E4PM_vs_E4CM[E4PM_vs_E4CM$cluster == i,]

  # Subset by log2FC
  E4CF_vs_E3CF_ct_up <- E4CF_vs_E3CF_ct[E4CF_vs_E3CF_ct$avg_log2FC > 0,][,"gene"]
  E4PF_vs_E4CF_ct_up <- E4PF_vs_E4CF_ct[E4PF_vs_E4CF_ct$avg_log2FC > 0,][,"gene"]
  E4CM_vs_E3CM_ct_up <- E4CM_vs_E3CM_ct[E4CM_vs_E3CM_ct$avg_log2FC > 0,][,"gene"]
  E4PM_vs_E4CM_ct_up <- E4PM_vs_E4CM_ct[E4PM_vs_E4CM_ct$avg_log2FC > 0,][,"gene"]
  axis_max <- max(length(E4CF_vs_E3CF_ct_up), length(E4PF_vs_E4CF_ct_up),
                  length(E4CM_vs_E3CM_ct_up), length(E4PM_vs_E4CM_ct_up)) + 300
  axis_max <- ceiling(axis_max / 100) * 100
 
  # format in a list
  list_input <- list(
    "E4PM vs E4CM Up-regulated" = E4PM_vs_E4CM_ct_up,
    "E4CM vs E3CM Up-regulated" = E4CM_vs_E3CM_ct_up,
    "E4PF vs E4CF Up-regulated" = E4PF_vs_E4CF_ct_up,
    "E4CF vs E3CF Up-regulated" = E4CF_vs_E3CF_ct_up
  )
  data <- UpSetR::fromList(list_input)
  
  # You will specify the male and female groups by their group number
  # Print this to help assign shapes/colors below
  group_mapping <- data.frame(
    Group_Number = 1:4,  # Group numbers (1-8)
    Gene_Set_Name = sort(colnames(data))  # Sorted gene set names
  )
  
  # Assign shapes based on group numbers
  male_groups <- c("2","4")  # Male group numbers
  female_groups <- c("1","3")  # Female group numbers
  
  # Assign colors and shapes based on specific group numbers
  upregulated_groups <- c("1","2","3","4")
  downregulated_groups <- c()
  
  # Create the basic UpSet plot (no dynamic colors or shapes) to extract n_points
  dummy_upset <- ComplexUpset::upset(
    data, 
    colnames(data),  # Use column names (gene sets)
    set_sizes = ComplexUpset::upset_set_size(),  # Display set sizes
    sort_sets = FALSE,
    
    # Plot the intersection matrix with points (default shapes)
    matrix = ComplexUpset::intersection_matrix(
      ggplot2::geom_point(size = 3)  # Simple points without dynamic aesthetics
    )
  )
  
  # Ensure n_points matches the actual data
  n_points <- nrow(dummy_upset$data)
  
  # Create empty vectors for colors and shapes
  color_vector <- rep(NA, n_points)  # To store colors
  shape_vector <- rep(NA, n_points)  # To store shapes
  
  # Assign shapes and colors based on group numbers
  for (j in 1:n_points) {
    group_value <- as.character(dummy_upset$data$group[j])
    
    # Assign shapes
    if (group_value %in% male_groups) {
      shape_vector[j] <- 22  # Filled square for Male comparisons
    } else if (group_value %in% female_groups) {
      shape_vector[j] <- 21  # Circle for Female comparisons
    }
    
    # Assign colors based on group numbers, only if the point is involved in an intersection
    if (dummy_upset$data$value[j]) {  # If the point is part of an intersection
      if (group_value %in% upregulated_groups) {
        color_vector[j] <- "red"  # Red for Up-regulated
      } else if (group_value %in% downregulated_groups) {
        color_vector[j] <- "blue"  # Blue for Down-regulated
      }
    } else {
      color_vector[j] <- NA  # No fill for points not involved in intersections
    }
  }
  
  # Ensure both vectors are of correct length
  if (length(shape_vector) != n_points || length(color_vector) != n_points) {
    stop("Error: shape_vector or color_vector length does not match the number of plotted points.")
  }
  
  # Reintroduce dynamic fill color for filled shapes
  upset_gene <- ComplexUpset::upset(
    data, 
    colnames(data),
    set_sizes = (
      ComplexUpset::upset_set_size() +
      geom_text(aes(label = ..count..), hjust = 1.1, stat = 'count') +
      expand_limits(y = axis_max)
    ),
    
    # Prevent automatic sorting of sets
    sort_sets = FALSE,
    
    matrix = ComplexUpset::intersection_matrix(
      ggplot2::geom_point(
        ggplot2::aes(shape = factor(shape_vector), fill = color_vector),  # Fill shapes with dynamic colors
        size = 3,
        stroke = 0.45  # Keeps the outline black
      )
    ) + ggplot2::scale_shape_manual(
      name = "Sex",
      labels = c("Male", "Female"),  # Legend labels
      values = c(21, 22) # Circle=21, Filled square=22
      ) + ggplot2::scale_fill_identity(na.translate = FALSE),  # Use the exact colors in color_vector and remove NA fill
    
      base_annotations = list(
      'Intersection size' = (
        intersection_size(bar_number_threshold = 1, width = 0.5) +
        scale_y_continuous(expand = expansion(mult = c(0, 0.05)), limits = c(0, axis_max)) +
        theme(
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_line(colour = 'black')
        )
      )
    ),
  )
  
  # give title
  upset_gene <- upset_gene + ggtitle(paste0(i,", adj_p_val < 0.05"))
  
  # remove legend
  upset_gene <- upset_gene + theme(legend.position = "None")
  
  # save
  i <- gsub(" ","_",i)
  pdf(paste0(out, "DEGs/upset/E4_vs_E3_upreg/", tolower(i), "_upset.pdf"), height = 6, width = 12)
  print(upset_gene)
  dev.off()
}
```

### E4 vs E3 Down
```{r E4_vs_E3_down_read_tables}
# set thresh
thresh <- 0.05

# read tables
prefix <- paste0(out, "DEGs/DEG_tables/")
E4CF_vs_E3CF <- read.table(paste0(prefix,"E4CF_vs_E3CF_DEGs.tsv"),
                           sep = "\t",
                           header = TRUE)
E4PF_vs_E4CF <- read.table(paste0(prefix,"E4PF_vs_E4CF_DEGs.tsv"),
                           sep = "\t",
                           header = TRUE)
E4CM_vs_E3CM <- read.table(paste0(prefix,"E4CM_vs_E3CM_DEGs.tsv"),
                           sep = "\t",
                           header = TRUE)
E4PM_vs_E4CM <- read.table(paste0(prefix,"E4PM_vs_E4CM_DEGs.tsv"),
                           sep = "\t",
                           header = TRUE)


# filter
E4CF_vs_E3CF <- E4CF_vs_E3CF[E4CF_vs_E3CF$p_val_adj < thresh,]
E4PF_vs_E4CF <- E4PF_vs_E4CF[E4PF_vs_E4CF$p_val_adj < thresh,]
E4CM_vs_E3CM <- E4CM_vs_E3CM[E4CM_vs_E3CM$p_val_adj < thresh,]
E4PM_vs_E4CM <- E4PM_vs_E4CM[E4PM_vs_E4CM$p_val_adj < thresh,]
```

```{r E4_vs_E3_down_upset, eval=FALSE}
# loop through cell types
clusters <- unique(c(E4CF_vs_E3CF$cluster, E4PF_vs_E4CF$cluster,
                     E4CM_vs_E3CM$cluster, E4PM_vs_E4CM$cluster))

for (i in clusters) {
  
  # skip cluster with not enough DEGs
  if (i == "LECs") {
    next
  }
  
  # Subset df by cluster
  E4CF_vs_E3CF_ct <- E4CF_vs_E3CF[E4CF_vs_E3CF$cluster == i,]
  E4PF_vs_E4CF_ct <- E4PF_vs_E4CF[E4PF_vs_E4CF$cluster == i,]
  E4CM_vs_E3CM_ct <- E4CM_vs_E3CM[E4CM_vs_E3CM$cluster == i,]
  E4PM_vs_E4CM_ct <- E4PM_vs_E4CM[E4PM_vs_E4CM$cluster == i,]

  # Subset by log2FC
  E4CF_vs_E3CF_ct_down <- E4CF_vs_E3CF_ct[E4CF_vs_E3CF_ct$avg_log2FC < 0,][,"gene"]
  E4PF_vs_E4CF_ct_down <- E4PF_vs_E4CF_ct[E4PF_vs_E4CF_ct$avg_log2FC < 0,][,"gene"]
  E4CM_vs_E3CM_ct_down <- E4CM_vs_E3CM_ct[E4CM_vs_E3CM_ct$avg_log2FC < 0,][,"gene"]
  E4PM_vs_E4CM_ct_down <- E4PM_vs_E4CM_ct[E4PM_vs_E4CM_ct$avg_log2FC < 0,][,"gene"]
  axis_max <- max(length(E4CF_vs_E3CF_ct_down), length(E4PF_vs_E4CF_ct_down),
                  length(E4CM_vs_E3CM_ct_down), length(E4PM_vs_E4CM_ct_down)) + 300
  axis_max <- ceiling(axis_max / 100) * 100
 
  # format in a list
  list_input <- list(
    "E4PM vs E4CM Down-regulated" = E4PM_vs_E4CM_ct_down,
    "E4CM vs E3CM Down-regulated" = E4CM_vs_E3CM_ct_down,
    "E4PF vs E4CF Down-regulated" = E4PF_vs_E4CF_ct_down,
    "E4CF vs E3CF Down-regulated" = E4CF_vs_E3CF_ct_down
  )
  data <- UpSetR::fromList(list_input)
  
  # You will specify the male and female groups by their group number
  # Print this to help assign shapes/colors below
  group_mapping <- data.frame(
    Group_Number = 1:4,  # Group numbers (1-8)
    Gene_Set_Name = sort(colnames(data))  # Sorted gene set names
  )
  
  # Assign shapes based on group numbers
  male_groups <- c("2","4")  # Male group numbers
  female_groups <- c("1","3")  # Female group numbers
  
  # Assign colors and shapes based on specific group numbers
  upregulated_groups <- c()
  downregulated_groups <- c("1","2","3","4")
  
  # Create the basic UpSet plot (no dynamic colors or shapes) to extract n_points
  dummy_upset <- ComplexUpset::upset(
    data, 
    colnames(data),  # Use column names (gene sets)
    set_sizes = ComplexUpset::upset_set_size(),  # Display set sizes
    sort_sets = FALSE,
    
    # Plot the intersection matrix with points (default shapes)
    matrix = ComplexUpset::intersection_matrix(
      ggplot2::geom_point(size = 3)  # Simple points without dynamic aesthetics
    )
  )
  
  # Ensure n_points matches the actual data
  n_points <- nrow(dummy_upset$data)
  
  # Create empty vectors for colors and shapes
  color_vector <- rep(NA, n_points)  # To store colors
  shape_vector <- rep(NA, n_points)  # To store shapes
  
  # Assign shapes and colors based on group numbers
  for (j in 1:n_points) {
    group_value <- as.character(dummy_upset$data$group[j])
    
    # Assign shapes
    if (group_value %in% male_groups) {
      shape_vector[j] <- 22  # Filled square for Male comparisons
    } else if (group_value %in% female_groups) {
      shape_vector[j] <- 21  # Circle for Female comparisons
    }
    
    # Assign colors based on group numbers, only if the point is involved in an intersection
    if (dummy_upset$data$value[j]) {  # If the point is part of an intersection
      if (group_value %in% upregulated_groups) {
        color_vector[j] <- "red"  # Red for Up-regulated
      } else if (group_value %in% downregulated_groups) {
        color_vector[j] <- "blue"  # Blue for Down-regulated
      }
    } else {
      color_vector[j] <- NA  # No fill for points not involved in intersections
    }
  }
  
  # Ensure both vectors are of correct length
  if (length(shape_vector) != n_points || length(color_vector) != n_points) {
    stop("Error: shape_vector or color_vector length does not match the number of plotted points.")
  }
  
  # Reintroduce dynamic fill color for filled shapes
  upset_gene <- ComplexUpset::upset(
    data, 
    colnames(data),
    set_sizes = (
      ComplexUpset::upset_set_size() +
      geom_text(aes(label = ..count..), hjust = 1.1, stat = 'count') +
      expand_limits(y = axis_max)
    ),
    
    # Prevent automatic sorting of sets
    sort_sets = FALSE,
    
    matrix = ComplexUpset::intersection_matrix(
      ggplot2::geom_point(
        ggplot2::aes(shape = factor(shape_vector), fill = color_vector),  # Fill shapes with dynamic colors
        size = 3,
        stroke = 0.45  # Keeps the outline black
      )
    ) + ggplot2::scale_shape_manual(
      name = "Sex",
      labels = c("Male", "Female"),  # Legend labels
      values = c(21, 22) # Circle=21, Filled square=22
      ) + ggplot2::scale_fill_identity(na.translate = FALSE),  # Use the exact colors in color_vector and remove NA fill
    
      base_annotations = list(
      'Intersection size' = (
        intersection_size(bar_number_threshold = 1, width = 0.5) +
        scale_y_continuous(expand = expansion(mult = c(0, 0.05)), limits = c(0, axis_max)) +
        theme(
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_line(colour = 'black')
        )
      )
    ),
  )
  
  # give title
  upset_gene <- upset_gene + ggtitle(paste0(i,", adj_p_val < 0.05"))
  
  # remove legend
  upset_gene <- upset_gene + theme(legend.position = "None")
  
  # save
  i <- gsub(" ","_",i)
  pdf(paste0(out, "DEGs/upset/E4_vs_E3_downreg/", tolower(i), "_upset.pdf"), height = 6, width = 12)
  print(upset_gene)
  dev.off()
}
```

# Shiny App
- scDRcell is the function used to plot the first UMAP
- inpConf is an argument of scDRcell, inpConf = sc1conf
- infConf gets subset by UI column based on inpdrX which is the user's input for x-axis
- inpdrX is used to set the x-axis label
```{r shiny_app, eval=FALSE}
# create new object
shiny.obj <- mouse.downsample
shiny.obj <- JoinLayers(shiny.obj)

# set default params
VariableFeatures(shiny.obj) <- shiny.obj@assays$SCT@var.features
DefaultAssay(shiny.obj) <- "RNA"
Idents(shiny.obj) <- "annotated_clusters"

# create config
names <- colnames(shiny.obj@meta.data)
names <- names[c(27,28,2:26)]
sc.config <- createConfig(obj = shiny.obj,
                          meta.to.include = names)

# change wd
setwd("../../results/all_clusters/downsampled")

# output shiny app folder
makeShinyApp(obj = shiny.obj, 
             scConf = sc.config, 
             gene.mapping = TRUE,
             shiny.title = "PLX5622 Downsampled")

# manual config edits
setwd("../../results/all_clusters/downsampled")
sc1conf <- readRDS("shinyApp/sc1conf.rds")
cluster_hex <- "#B5B9BA|#3385BB|#40BBFF|#A5D5A9|#1C7E24|#F57C7C|#E42622|#FBB268|#FE8D19|#DE9E83|#A6CEE3|#9D7BBA|#977899"
sc1conf[2,4] <- cluster_hex
group_hex <- "#B89B74|#725E47|#B15928|#F7590B|#A6CEE3|#34B1E8|#ED8F47|#F4CC25"
sc1conf[18,4] <- group_hex
saveRDS(sc1conf,"shinyApp/sc1conf.rds")
```
